<?xml version="1.0" encoding="utf-8" ?> 
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1 plus MathML 2.0//EN" 
"http://www.w3.org/Math/DTD/mathml2/xhtml-math11-f.dtd" > 
<html xmlns="http://www.w3.org/1999/xhtml"  
> 
<head><title>Introduction to Biological Statistics
 Imperial College London</title> 
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" /> 
<meta name="generator" content="TeX4ht (http://www.cse.ohio-state.edu/~gurari/TeX4ht/)" /> 
<meta name="originator" content="TeX4ht (http://www.cse.ohio-state.edu/~gurari/TeX4ht/)" /> 
<!-- xhtml,mathml,charset=utf-8,html --> 
<meta name="src" content="ICUGBioStats.tex" /> 
<meta name="date" content="2018-01-25 21:02:00" /> 
<link rel="stylesheet" type="text/css" href="ICUGBioStats.css" /> 
</head><body 
>
<div class="maketitle">
                                                                                 
                                                                                 
                                                                                 
                                                                                 
                                                                                 
                                                                                 

<h2 class="titleHead">Introduction to Biological Statistics<br />
Imperial College London</h2>
                  <div class="author" ><span 
class="ptmr8t-x-x-120">Samraat Pawar (s.pawar@imperial.ac.uk),</span>
<br /> <span 
class="ptmri8t-x-x-120">with many inputs from Steve Cook and major contributions from David Orme! </span></div>
<br />
<div class="date" ><span 
class="ptmr8t-x-x-120">January 25, 2018</span></div>
                                                                                 
                                                                                 
</div>
                                                                                 
                                                                                 
<h2 class="likechapterHead"><a 
 id="x1-1000"></a>Contents</h2> <div class="tableofcontents">
<span class="chapterToc" >0 <a 
href="#x1-20000" id="QQ2-1-2">Introduction</a></span>
<br /> <span class="sectionToc" >0.1 <a 
href="#x1-30000.1" id="QQ2-1-3">What is this document about?</a></span>
<br /> <span class="sectionToc" >0.2 <a 
href="#x1-40000.2" id="QQ2-1-4">Conventions used in this document</a></span>
<br /> <span class="sectionToc" >0.3 <a 
href="#x1-50000.3" id="QQ2-1-5">Readings</a></span>
<br /><span class="chapterToc" >1 <a 
href="#x1-60001" id="QQ2-1-6">Experimental design and Data exploration</a></span>
<br /> <span class="sectionToc" >1.1 <a 
href="#x1-70001.1" id="QQ2-1-7">Some statistical parlance</a></span>
<br /> <span class="sectionToc" >1.2 <a 
href="#x1-80001.2" id="QQ2-1-8">Descriptive Statistics</a></span>
<br />  <span class="subsectionToc" >1.2.1 <a 
href="#x1-90001.2.1" id="QQ2-1-9">Descriptive statistic functions in R</a></span>
<br /> <span class="sectionToc" >1.3 <a 
href="#x1-100001.3" id="QQ2-1-10">Data types and distributions</a></span>
<br />  <span class="subsectionToc" >1.3.1 <a 
href="#x1-110001.3.1" id="QQ2-1-11">Sampling from distributions in R</a></span>
<br /> <span class="sectionToc" >1.4 <a 
href="#x1-120001.4" id="QQ2-1-12">Two basic rules of experimental design and sampling</a></span>
<br /> <span class="sectionToc" >1.5 <a 
href="#x1-130001.5" id="QQ2-1-13">A data exploration case study</a></span>
<br />  <span class="subsectionToc" >1.5.1 <a 
href="#x1-140001.5.1" id="QQ2-1-14">Visualise distributions of the variables</a></span>
<br />  <span class="subsectionToc" >1.5.2 <a 
href="#x1-150001.5.2" id="QQ2-1-15">Take a quick look at eﬀects of certain factors</a></span>
</div>
                                                                                 
                                                                                 
<!--l. 114--><p class="noindent" >
                                                                                 
                                                                                 
                                                                                 
                                                                                 
</p><!--l. 1--><p class="noindent" >
                                                                                 
                                                                                 
</p>
<h2 class="chapterHead"><span class="titlemark">Chapter 0</span><br /><a 
 id="x1-20000"></a>Introduction</h2>
     <div class="quote">
     <!--l. 3--><p class="noindent" >Donald Knuth, 1995:   <span 
class="ptmri8t-x-x-109">Science is what we understand well enough to explain to a</span>
     <span 
class="ptmri8t-x-x-109">computer. Art is everything else we do.</span></p></div>
<h3 class="sectionHead"><span class="titlemark">0.1    </span> <a 
 id="x1-30000.1"></a>What is this document about?</h3>
<!--l. 10--><p class="noindent" >This document contains the content of the Imperial College London Department of Life Sciences BSc
Year 1 (Y1) and 2 (Y2) statistics modules, including workshops and practicals. Diﬀerent subsets of the
content of this document will be covered in Y1 and Y2. You will be given instructions about which
sections will be covered in your respective (Y1 or Y2) modules. You will learn statistics with R in a
hands-on on way in this module. A few “pure” lectures on statistical concepts will also be
delivered.
</p><!--l. 19--><p class="noindent" >This document is accompanied by data and code on which you can practice your skills in your own time
and during the Practical sessions. These materials are available (and will be updated regularly) at
bitbucket (it&#x2019;s like GitHub): <a 
href="https://bitbucket.org/mhasoba/iccompbiostat" class="url" ><span 
class="pcrr8t-x-x-109">https://bitbucket.org/mhasoba/iccompbiostat</span></a>.
</p><!--l. 24--><p class="noindent" ><span 
class="ptmb8t-x-x-109">Note that you can download the code, data and everything from the bitbucket repository at one go,</span>
<span 
class="ptmb8t-x-x-109">by going to </span><a 
href="https://bitbucket.org/mhasoba/iccompbiostat" class="url" ><span 
class="pcrb8t-x-x-109">https://bitbucket.org/mhasoba/iccompbiostat</span></a><span 
class="ptmb8t-x-x-109">and then clicking on the</span>
<span 
class="ptmb8t-x-x-109">“Download repository” link. You can then unzip the downloaded .zip and grab the ﬁles you</span>
<span 
class="ptmb8t-x-x-109">need.</span>
</p><!--l. 30--><p class="noindent" >These notes and the scripts that you create form a valuable reference for using and interpreting statistics.
You will be expected to be able to use your experience and these notes to complete analyses for statistical
components of other courses, in the second and third year in particular.
</p><!--l. 36--><p class="noindent" >The topics covered here assume that you have already worked through at least the basic R sections of the
<span 
class="pcrr8t-x-x-109">SilBioComp.pdf </span>document (also available in the same bitbucket repository)
</p><!--l. 41--><p class="noindent" >Finally, it is important that you work through the problems in each chapter, particularly as some
of the questions ask you to ﬁnd out about commands and functions not introduced in the
chapter&#x2019;s text itself, but which will be relied on in later chapters. By the time you have attended
the lectures and workshops, and completed the exercises in this document, you should able
to:
     </p><ul>
     <li class="compactitem">Use R to perform common statistical tests, particularly t, F, and <!--l. 49--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><msup><mrow 
><mi 
>&chi;</mi></mrow><mrow 
><mn>2</mn></mrow></msup 
></math>
     tests
     </li>
     <li class="compactitem">Develop ability to build, criticize and interpret linear models, especially linear regression
     and ANOVA.
     </li>
     <li class="compactitem">Interpret R output (particularly p values) correctly, and use them appropriately in practical
     write-ups and presentations</li></ul>
                                                                                 
                                                                                 
<!--l. 56--><p class="noindent" >
</p>
<h3 class="sectionHead"><span class="titlemark">0.2    </span> <a 
 id="x1-40000.2"></a>Conventions used in this document</h3>
<!--l. 58--><p class="noindent" >You will ﬁnd all R commandline/console arguments, code snippets and output in colored boxes like this:
</p><!--l. 60-->
<div class="lstlisting" id="listing-1"><span class="label"><a 
 id="x1-4001r1"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">ls</span><span 
class="pcrr8t-x-x-80">()</span>
</div>
<!--l. 63--><p class="noindent" >Here <!--l. 63--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" > <mo 
class="MathClass-rel">&#x003E;</mo></math>
is the R prompt, and will type (or copy-paste, but not recommended!) the commands/code
that you see from this document into the R command line. But please exclude the
<!--l. 65--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" > <mo 
class="MathClass-rel">&#x003E;</mo></math> as this is just the R command
prompt! I have included the <!--l. 66--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" > <mo 
class="MathClass-rel">&#x003E;</mo></math>
prompts in the code shown in this document so that you are forced to see what each line is doing. Indeed,
avoid copying-and-pasting chunks of R code you do not understand: blindly shovelling data into a black
box and assuming the output is correct and meaningful will eventually lead to frustrations, and if you are
unlucky, embarrassments or even catastrophes!
</p><!--l. 73--><p class="noindent" >The content of this document is computer platform (Mac, PC or Linux) independent because many of you
will probably also later be working with R on your personal laptops or desktops. Indeed,
platform-independence of your statistical analysis is one of the main reasons why you are using
R!
</p><!--l. 79--><p class="noindent" >Finally, note that:
    </p><ul>
    <li class="compactitem">In all subsequent chapters, lines marked with a star like this are things for you to do.</li></ul>
<!--l. 85--><p class="noindent" >
</p>
<h3 class="sectionHead"><span class="titlemark">0.3    </span> <a 
 id="x1-50000.3"></a>Readings</h3>
<!--l. 87--><p class="noindent" >Look up the Readings directory on the bitbucket repository (link given above).
</p>
     <ul class="itemize1">
     <li class="itemize">Bolker, B. M.: Ecological Models and Data in R (eBook and Hardcover available).
     </li>
     <li class="itemize">Beckerman,  A.  P.  &#x0026;  Petchey,  O.  L.  (2012)  Getting  started  with  R:  an  introduction  for
     biologists. Oxford, Oxford University Press. <br 
class="newline" />Good, short, general introduction
     </li>
     <li class="itemize">Crawley, R. (2013) The R book. 2nd edition. Chichester, Wiley. <br 
class="newline" />Excellent     but     enormous     reference     book,     code     and     data     available     from
     <a 
href="www.bio.ic.ac.uk/research/mjcraw/therbook/index.htm" class="url" ><span 
class="pcrr8t-x-x-109">www.bio.ic.ac.uk/research/mjcraw/therbook/index.htm</span></a>
     </li>
     <li class="itemize">Use the internet! Type “R tutorial”, and scores will pop up. Choose one that seems the most
     intuitive to you.</li></ul>
                                                                                 
                                                                                 
<!--l. 1--><p class="noindent" >
                                                                                 
                                                                                 
</p>
<h2 class="chapterHead"><span class="titlemark">Chapter 1</span><br /><a 
 id="x1-60001"></a>Experimental design and Data exploration</h2>
<!--l. 4--><p class="noindent" >Ideally, you would like to design experiments (manipulations and/or observations) that are appropriate for
the question you want to answer. However, you still need to explore you data to determine what kind of
statistical tests would be appropriate because: (a) Your experiments or observations may not go as
planned, and (b) You might have somebody else&#x2019;s data to analyse (very likely in your UG projects).
Building on the previous chapter, in this chapter we you learn how to use R to explore your data and
determine appropriate statistical tests. By the time you have worked through this chapter, you should be
able to
     </p><ul>
     <li class="compactitem">Provided suﬃcient information is available, be able to judge whether the sampling design
     used to generate a particular dataset was appropriate
     </li>
     <li class="compactitem">Determine  if  your  sample  sizes  are  adequate,  especially  for  a  speciﬁc  statistical  test
     qualitatively (Y1) or quantitatively using power analysis (Y2).
     </li>
     <li class="compactitem">Calculate basic statistical measures on your data to determine its properties</li></ul>
<!--l. 25--><p class="noindent" >We are going to start oﬀ in Y1 with the simplest of scenarios for statistical testing — that you want to
determine whether a sample, or a pair of samples meet some expectation (hypothesis) or
not.
</p><!--l. 29--><p class="noindent" >First, some conceptual preliminaries.
</p>
<h3 class="sectionHead"><span class="titlemark">1.1    </span> <a 
 id="x1-70001.1"></a>Some statistical parlance</h3>
<!--l. 32--><p class="noindent" >The following terms are important for you to get familiar with:
     </p><dl class="description"><dt class="description">
<span 
class="ptmb8t-x-x-109">(Statistical) Population</span> </dt><dd 
class="description">A statistical population is a <span 
class="ptmri8t-x-x-109">complete set </span>of items that share at least one
     <span 
class="ptmri8t-x-x-109">attribute </span>of interest. This attribute of interest is the target of your statistical analysis. For
     example,  if  we  are  interested  in  studying  the  weight  of  year-old  cod  in  the  Oceans,  the
     population consists of <span 
class="ptmri8t-x-x-109">all </span>year-old cod, but more speciﬁcally, the weight measurements of
     all the individuals of the cod population is what we want to analyse.
     </dd><dt class="description">
<span 
class="ptmb8t-x-x-109">(Statistical) Distribution</span> </dt><dd 
class="description">A  statistical  distribution  is  a  mathematical  description  (expressed  as
     a mathematical equation) of the properties of a population of interest. Theoreticians have
     come up with a bunch of distributions (e.g., Gaussian or Normal, Poisson, Binomial, etc.)
     that are appropriate for diﬀerent kinds of data. Figuring out which distribution best describes
     a population of interest is one of the ﬁrst steps in a statistical analysis. In reality, of course,
     even collecting and measuring all the individuals of a population may not be suﬃcient to
     characterize its statistical properties — imagine the situation where the cod population has
     declined to a few hundred individuals (not an impossibility in the future!).
     </dd><dt class="description">
<span 
class="ptmb8t-x-x-109">(Data or Population) Sample</span> </dt><dd 
class="description">A data <span 
class="ptmri8t-x-x-109">sample </span>is a set of measurements of the attribute of interest
     collected from a <span 
class="ptmri8t-x-x-109">statistical population </span>by a deﬁned procedure (<span 
class="ptmri8t-x-x-109">sampling methodology</span>). In
                                                                                 
                                                                                 
     the  cod  example  above,  this  would  be  the  weight  of  every  individual  of  a  <span 
class="ptmri8t-x-x-109">subset  </span>of  the
     year-old Cod population.
     </dd><dt class="description">
<span 
class="ptmb8t-x-x-109">(Statistical) Parameter</span> </dt><dd 
class="description">A statistical parameter is a measure of some attribute of the <span 
class="ptmri8t-x-x-109">theoretical</span>
     statistical distribution that is support to represent your population. An example would be the
     average weight of yearling cod. In practice, this is not measurable because the population is
     much too large or incompletely inaccessible/invisible — imagine measuring the weight of
     every year-old cod individual in the oceans!
     </dd><dt class="description">
<span 
class="ptmb8t-x-x-109">Statistic</span> </dt><dd 
class="description">A  statistic  (singular)  is  an  <span 
class="ptmri8t-x-x-109">estimate  </span>of  a  statistical  parameter  of  the  population  of
     interest, obtained by calculating the <span 
class="ptmri8t-x-x-109">measure </span>for a <span 
class="ptmri8t-x-x-109">sample </span>(e.g., the average or mean weight
     of  individuals  in  a  sample  of  one-year  old  cod).  This  is  also  your  <span 
class="ptmri8t-x-x-109">descriptive  statistic</span>.
     Therefore,  a  <span 
class="ptmri8t-x-x-109">Statistic  </span>is  to  a  <span 
class="ptmri8t-x-x-109">Statistical  Parameter  </span>what  a  <span 
class="ptmri8t-x-x-109">Sample  </span>is  to  the  <span 
class="ptmri8t-x-x-109">Statistical</span>
     <span 
class="ptmri8t-x-x-109">Population</span>. For example, the average of a sample of cod weights is a statistic that <span 
class="ptmri8t-x-x-109">estimates</span>
     the theoretical “real” average of the weights of the entire one-year Cod population, which is
     its statistical parameter.
     </dd><dt class="description">
<span 
class="ptmb8t-x-x-109">Hypothesis</span> </dt><dd 
class="description">A Hypothesis is an (hopefully) informed <span 
class="ptmri8t-x-x-109">postulate </span>about an attribute of your population
     of interest. For example, you may hypothesize that the one-year old cod population&#x2019;s mean
     weight has declined over the last two decades (it has!). You will want to confront your main
     hypothesis with a “null” Hypothesis, to minimize the risk of making a “type I” error. A type
     I error is the probability of accepting an alternative (or main) hypothesis (and rejecting the
     null hypothesis) that is not really valid (e.g., the yearling cods have actually not declined
     in weight). This is a big NO NO from a scientiﬁc and philosophical standpoint. The rate or
     probability of the type I error is denoted by the Greek letter <!--l. 94--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mi 
>&alpha;</mi></math>,
     and equals the <span 
class="ptmri8t-x-x-109">signiﬁcance level </span>of a statistical test.
     </dd></dl>
<!--l. 99--><p class="noindent" >
</p>
<h3 class="sectionHead"><span class="titlemark">1.2    </span> <a 
 id="x1-80001.2"></a>Descriptive Statistics</h3>
<!--l. 101--><p class="noindent" >The fundamental statistics that describe a sample (or a population) are, ﬁrstly, the mean, or average value of a sample,
typically denoted by a <!--l. 103--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mover 
accent="true"><mrow 
><mi 
>x</mi></mrow><mo 
class="MathClass-op">̄</mo></mover></math>:
</p>
<table class="equation"><tr><td> <a 
 id="x1-8001r1"></a>
                                                                                 
                                                                                 
<!--l. 105--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="block" class="equation">
                           <mover 
accent="true"><mrow 
><mi 
>x</mi></mrow><mo 
class="MathClass-op">̄</mo></mover> <mo 
class="MathClass-rel">=</mo> <mfrac><mrow 
><munderover accentunder="false" accent="false"><mrow  
><mo mathsize="big" 
>&sum;</mo></mrow><mrow 
><mi 
>i</mi><mo 
class="MathClass-rel">=</mo><mn>1</mn></mrow><mrow 
><mi 
>n</mi></mrow></munderover 
><msub><mrow 
><mi 
>x</mi></mrow><mrow 
><mi 
>i</mi></mrow></msub 
></mrow> 
     <mrow 
><mi 
>n</mi></mrow></mfrac>    <mo 
class="MathClass-rel">=</mo> <mfrac><mrow 
><msub><mrow 
><mi 
>x</mi></mrow><mrow 
><mn>1</mn></mrow></msub 
> <mo 
class="MathClass-bin">+</mo><msub><mrow 
><mi 
>x</mi></mrow><mrow 
><mn>2</mn></mrow></msub 
> <mo 
class="MathClass-bin">+</mo><mo 
class="MathClass-rel">⋯</mo><mo 
class="MathClass-bin">+</mo><msub><mrow 
><mi 
>x</mi></mrow><mrow 
><mi 
>n</mi></mrow></msub 
></mrow> 
          <mrow 
><mi 
>n</mi></mrow></mfrac>
</math></td><td class="eq-no">(1.1)</td></tr></table>
<!--l. 109--><p class="noindent" >That is, it is the sum of all the values in a sample divided by the number,
<!--l. 110--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mi 
>n</mi></math>, of
items in the sample. Thus it is a measure of the <span 
class="ptmri8t-x-x-109">central-tendency </span>of the sample and population.
</p><!--l. 113--><p class="noindent" >Second, the standard deviation (<!--l. 113--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mi 
>s</mi></math>)
is:
</p>
<table class="equation"><tr><td> <a 
 id="x1-8002r2"></a>
<!--l. 115--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="block" class="equation">
            <mi 
>s</mi> <mo 
class="MathClass-rel">=</mo> <msqrt><mrow><mfrac><mrow 
><munderover accentunder="false" accent="false"><mrow  
><mo mathsize="big" 
>&sum;</mo> </mrow><mrow 
><mi 
>i</mi><mo 
class="MathClass-rel">=</mo><mn>1</mn> </mrow> <mrow 
>  <mi 
>n</mi> </mrow> </munderover 
><msup><mrow 
><mrow ><mo 
class="MathClass-open">(</mo><mrow><mover 
accent="true"><mrow 
><mi 
>x</mi></mrow><mo>̄</mo> </mover> <mo 
class="MathClass-bin">&minus;</mo><msub><mrow 
><mi 
>x</mi></mrow><mrow 
><mi 
>i</mi> </mrow> </msub 
> </mrow><mo 
class="MathClass-close">)</mo></mrow></mrow><mrow 
><mn>2</mn> </mrow> </msup 
> </mrow>
         <mrow 
><mi 
>n</mi><mo 
class="MathClass-bin">&minus;</mo><mn>1</mn></mrow></mfrac></mrow></msqrt>        <mo 
class="MathClass-rel">=</mo> <msqrt><mrow><mfrac><mrow 
><msup><mrow 
><mrow ><mo 
class="MathClass-open">(</mo><mrow><mover 
accent="true"><mrow 
><mi 
>x</mi></mrow><mo 
class="MathClass-op">̄</mo> </mover> <mo 
class="MathClass-bin">&minus;</mo><msub><mrow 
><mi 
>x</mi></mrow><mrow 
><mn>1</mn> </mrow> </msub 
> </mrow><mo 
class="MathClass-close">)</mo></mrow></mrow><mrow 
><mn>2</mn> </mrow> </msup 
> <mo 
class="MathClass-bin">+</mo><msup><mrow 
><mrow ><mo 
class="MathClass-open">(</mo><mrow><mover 
accent="true"><mrow 
><mi 
>x</mi></mrow><mo 
class="MathClass-op">̄</mo> </mover> <mo 
class="MathClass-bin">&minus;</mo><msub><mrow 
><mi 
>x</mi></mrow><mrow 
><mn>2</mn> </mrow> </msub 
> </mrow><mo 
class="MathClass-close">)</mo></mrow></mrow><mrow 
><mn>2</mn> </mrow> </msup 
> <mo 
class="MathClass-bin">+</mo><mo 
class="MathClass-rel">⋯</mo> <mo 
class="MathClass-bin">+</mo><msup><mrow 
><mrow ><mo 
class="MathClass-open">(</mo><mrow><mover 
accent="true"><mrow 
><mi 
>x</mi></mrow><mo 
class="MathClass-op">̄</mo> </mover> <mo 
class="MathClass-bin">&minus;</mo><msub><mrow 
><mi 
>x</mi></mrow><mrow 
><mi 
>n</mi> </mrow> </msub 
> </mrow><mo 
class="MathClass-close">)</mo></mrow></mrow><mrow 
><mn>2</mn> </mrow> </msup 
> </mrow>
                             <mrow 
><mi 
>n</mi><mo 
class="MathClass-bin">&minus;</mo><mn>1</mn></mrow></mfrac></mrow></msqrt>
</math></td><td class="eq-no">(1.2)</td></tr></table>
<!--l. 121--><p class="noindent" >That is, the the square root of the sum of squares (“SS”) of the diﬀerences between each item in
the sample and the mean, divided by the <span 
class="ptmri8t-x-x-109">degrees of freedom, “df” </span>remaining in the data set
(<!--l. 123--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mi 
>n</mi><mo 
class="MathClass-bin">&minus;</mo><mn>1</mn></math>). df is the
sample size, <!--l. 124--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mi 
>n</mi></math>,
minus the number of statistical parameters estimated from the data set. This is to reduce the <span 
class="ptmri8t-x-x-109">bias </span>in your
<span 
class="ptmri8t-x-x-109">estimate </span>of the statistic, as you are calculating it from the sample, and not the whole theoretical
population.
</p><!--l. 129--><p class="noindent" >Thus, the formula for <!--l. 129--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mi 
>s</mi></math>
above has <!--l. 129--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mi 
>n</mi><mo 
class="MathClass-bin">&minus;</mo><mn>1</mn></math>
in its denominator because, to to work out the standard deviation, you must have already estimated the
mean (<!--l. 131--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mover 
accent="true"><mrow 
><mi 
>x</mi></mrow><mo 
class="MathClass-op">̄</mo></mover></math>)
from the same data set. This removes 1 degree of freedom. Also, note that the sample variance,
<!--l. 132--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><msup><mrow 
><mi 
>s</mi></mrow><mrow 
><mn>2</mn></mrow></msup 
></math> is the
square of standard deviation. Or, in other words, the standard deviation is the square-root of the
variance!
</p><!--l. 136--><p class="noindent" >The above two statistics (mean and sd) are particularly meaningful when the sample and population have
a symmetric distribution (e.g., normal or gaussian). When the distribution is not symmetric (that is, it is
<span 
class="ptmri8t-x-x-109">skewed</span>), another statistic, the <span 
class="ptmri8t-x-x-109">median </span>becomes important. This is the middle value in the
ordered set of data, that is exactly 50% of the data live below and 50% lie above the median. In
skewed distributions, the median is a better measure of the <span 
class="ptmri8t-x-x-109">central-tendency </span>of the sample and
population.
                                                                                 
                                                                                 
</p><!--l. 145--><p class="noindent" >Other descriptive statistics you should keep in mind are the range (diﬀerence between the
largest and smallest values), and the quartiles (values lying in the data divided into the intervals
<!--l. 147--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mrow ><mo 
class="MathClass-open">[</mo><mrow><mfrac><mrow><mn>1</mn></mrow><mrow><mn>4</mn></mrow></mfrac><mo 
class="MathClass-punc">,</mo><mfrac><mrow> <mn>1</mn></mrow> <mrow><mn>2</mn></mrow></mfrac><mo 
class="MathClass-punc">,</mo><mfrac><mrow> <mn>3</mn></mrow> <mrow><mn>4</mn></mrow></mfrac><mo 
class="MathClass-punc">,</mo><mn>1</mn></mrow><mo 
class="MathClass-close">]</mo></mrow></math> or at
1% intervals (percentiles). Box-plots, which you have seen, represent a number of these statistics in one
ﬁgure.
</p><!--l. 151--><p class="noindent" >
</p>
<h4 class="subsectionHead"><span class="titlemark">1.2.1    </span> <a 
 id="x1-90001.2.1"></a>Descriptive statistic functions in R</h4>
<div class="tabular"> <table id="TBL-2" class="tabular" 
cellspacing="0" cellpadding="0"  
><colgroup id="TBL-2-1g"><col 
id="TBL-2-1" /><col 
id="TBL-2-2" /></colgroup><tr  
 style="vertical-align:baseline;" id="TBL-2-1-"><td  style="text-align:left;" id="TBL-2-1-1"  
class="td11"> <!--l. 154--><p class="noindent" ><span 
class="pcrr8t-x-x-109">mean(x)</span>         </p></td><td  style="text-align:left;" id="TBL-2-1-2"  
class="td11"> <!--l. 154--><p class="noindent" >Compute mean (of a vector or matrix)                                        </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-2-2-"><td  style="text-align:left;" id="TBL-2-2-1"  
class="td11"> <!--l. 155--><p class="noindent" ><span 
class="pcrr8t-x-x-109">sd(x)</span>           </p></td><td  style="text-align:left;" id="TBL-2-2-2"  
class="td11"> <!--l. 155--><p class="noindent" >Standard deviation                                                                       </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-2-3-"><td  style="text-align:left;" id="TBL-2-3-1"  
class="td11"> <!--l. 156--><p class="noindent" ><span 
class="pcrr8t-x-x-109">var(x)</span>          </p></td><td  style="text-align:left;" id="TBL-2-3-2"  
class="td11"> <!--l. 156--><p class="noindent" >Variance                                                                                       </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-2-4-"><td  style="text-align:left;" id="TBL-2-4-1"  
class="td11"> <!--l. 157--><p class="noindent" ><span 
class="pcrr8t-x-x-109">median(x)</span>       </p></td><td  style="text-align:left;" id="TBL-2-4-2"  
class="td11"> <!--l. 157--><p class="noindent" >Median                                                                                         </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-2-5-"><td  style="text-align:left;" id="TBL-2-5-1"  
class="td11"> <!--l. 158--><p class="noindent" ><span 
class="pcrr8t-x-x-109">quantile(x,0.05)</span></p></td><td  style="text-align:left;" id="TBL-2-5-2"  
class="td11"> <!--l. 158--><p class="noindent" >Compute the 0.05 quantile                                                          </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-2-6-"><td  style="text-align:left;" id="TBL-2-6-1"  
class="td11"> <!--l. 159--><p class="noindent" ><span 
class="pcrr8t-x-x-109">range(x)</span>        </p></td><td  style="text-align:left;" id="TBL-2-6-2"  
class="td11"> <!--l. 159--><p class="noindent" >Range of the data                                                                         </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-2-7-"><td  style="text-align:left;" id="TBL-2-7-1"  
class="td11"> <!--l. 160--><p class="noindent" ><span 
class="pcrr8t-x-x-109">min(x)</span>          </p></td><td  style="text-align:left;" id="TBL-2-7-2"  
class="td11"> <!--l. 160--><p class="noindent" >Minimum                                                                                     </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-2-8-"><td  style="text-align:left;" id="TBL-2-8-1"  
class="td11"> <!--l. 161--><p class="noindent" ><span 
class="pcrr8t-x-x-109">max(x)</span>          </p></td><td  style="text-align:left;" id="TBL-2-8-2"  
class="td11"> <!--l. 161--><p class="noindent" >Maximum                                                                                    </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-2-9-"><td  style="text-align:left;" id="TBL-2-9-1"  
class="td11"> <!--l. 162--><p class="noindent" ><span 
class="pcrr8t-x-x-109">sum(x)</span>          </p></td><td  style="text-align:left;" id="TBL-2-9-2"  
class="td11"> <!--l. 162--><p class="noindent" >Sum all elements                                                                         </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-2-10-"><td  style="text-align:left;" id="TBL-2-10-1"  
class="td11">                 </td>
</tr></table></div><br 
class="newline" />
<!--l. 165--><p class="noindent" >
</p>
<h3 class="sectionHead"><span class="titlemark">1.3    </span> <a 
 id="x1-100001.3"></a>Data types and distributions</h3>
<!--l. 166--><p class="noindent" >You will typically encounter or sample the following main types of data:
     </p><dl class="description"><dt class="description">
<span 
class="ptmb8t-x-x-109">Continuous numeric</span> </dt><dd 
class="description">This  is  the  <span 
class="pcrr8t-x-x-109">numeric </span>or  <span 
class="pcrr8t-x-x-109">real </span>data  type  in  R,  and  as  far  as  you  are
     concerned, these data typically will be made up of (mathematically) real numbers such as
     human height or weight. These may be unbounded (any value between negative inﬁnity to
     positive inﬁnity), or bounded (e.g., between or zero and some upper positive number) like
     human weight.
     </dd><dt class="description">
<span 
class="ptmb8t-x-x-109">Discrete numeric</span> </dt><dd 
class="description">This is the <span 
class="pcrr8t-x-x-109">integer </span>data type in R, and consist of (mathematically) integer
     (whole) numbers such as counts of individuals in a population, e.g., The number of bacteria
     in a ml of pond water.
     </dd><dt class="description">
<span 
class="ptmb8t-x-x-109">Percentage (proportion)</span> </dt><dd 
class="description">Percentage  data  is  a  particular  kind  of  numeric  data  that  is  strictly
     bounded between 0 and 100. The fact that you can never get samples of percentages that
     exceed these bounds makes such data tricky to analyse.
     </dd><dt class="description">
<span 
class="ptmb8t-x-x-109">Categorical</span> </dt><dd 
class="description">These are typically stored as the <span 
class="pcrr8t-x-x-109">character </span>data type in R. Categorical data are
     discrete, typically expressed as a one of a ﬁxed number of <span 
class="ptmri8t-x-x-109">levels </span>of a <span 
class="ptmri8t-x-x-109">factor</span>. For example,
                                                                                 
                                                                                 
     the the factor “Type.of.feeding.interaction” from the predator-prey dataset you have seen
     previously had ﬁve levels: “insectivorous”, “piscivorous”, “planktivorous”, “predacious”,
     and “predacious/piscivorous”.
     </dd><dt class="description">
<span 
class="ptmb8t-x-x-109">Binary (presence/absence) data</span> </dt><dd 
class="description">A special type of categorical data are binary, where only two
     categories or states are possible: (1, 0) (or “present”, “absent”), e.g., a disease symptom.
     These may be stored as <span 
class="pcrr8t-x-x-109">integer </span>or <span 
class="pcrr8t-x-x-109">character </span>in R.</dd></dl>
<!--l. 203--><p class="noindent" >While designing experiments or exploring data obtained by somebody else, you need to keep in mind that
each type will typically be best represented a particular <span 
class="ptmri8t-x-x-109">statistical distribution</span>. For example, continuous
numeric data are <span 
class="ptmri8t-x-x-109">often </span>normally distributed. On the other hand, count data are likely to be distributed
according to the Poisson distribution.
</p><!--l. 210--><p class="noindent" >If you are lucky, you will mostly have to deal with data that are continuous or discrete numeric, which
are the most straightforward to analyse using Linear models (more on that in subsequent
chapters). However, some of the most interesting and important problems in biology involve
proportion (percentage), categorical and binary data (e.g., Presence or absence of a disease
symptom).
</p><!--l. 217--><p class="noindent" >For example, think about what type of data, and what type of distribution, a sample of the following is
likely to be:
     </p><ul>
     <li class="compactitem">Wavelength of light
     </li>
     <li class="compactitem">Temperature
     </li>
     <li class="compactitem">Egg clutch size
     </li>
     <li class="compactitem">Rate of a reaction
     </li>
     <li class="compactitem">Eye-colour
     </li>
     <li class="compactitem">Score in Scrabble
     </li>
     <li class="compactitem">UG Degree class
     </li>
     <li class="compactitem">Ground-cover of grass in a quadrat
     </li>
     <li class="compactitem">Winning side in chess</li></ul>
<!--l. 231--><p class="noindent" >
</p>
<h4 class="subsectionHead"><span class="titlemark">1.3.1    </span> <a 
 id="x1-110001.3.1"></a>Sampling from distributions in R</h4>
<!--l. 233--><p class="noindent" >You can generate samples form many distributions in R (and handy thing to know). In particular, the
following are important:
</p>
                                                                                 
                                                                                 
<div class="tabular"> <table id="TBL-3" class="tabular" 
cellspacing="0" cellpadding="0"  
><colgroup id="TBL-3-1g"><col 
id="TBL-3-1" /><col 
id="TBL-3-2" /></colgroup><tr  
 style="vertical-align:baseline;" id="TBL-3-1-"><td  style="text-align:left;" id="TBL-3-1-1"  
class="td11"> <!--l. 237--><p class="noindent" ><span 
class="pcrr8t-x-x-109">rnorm(10, m=0, sd=1)</span>    </p></td><td  style="text-align:left;" id="TBL-3-1-2"  
class="td11"> <!--l. 237--><p class="noindent" >Draw 10 normal random numbers with mean 0 and s.d. 1    </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-3-2-"><td  style="text-align:left;" id="TBL-3-2-1"  
class="td11"> <!--l. 238--><p class="noindent" ><span 
class="pcrr8t-x-x-109">dnorm(x, m=0, sd=1)</span>      </p></td><td  style="text-align:left;" id="TBL-3-2-2"  
class="td11"> <!--l. 238--><p class="noindent" >Density function                                                                     </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-3-3-"><td  style="text-align:left;" id="TBL-3-3-1"  
class="td11"> <!--l. 239--><p class="noindent" ><span 
class="pcrr8t-x-x-109">qnorm(x, m=0, sd=1)</span>      </p></td><td  style="text-align:left;" id="TBL-3-3-2"  
class="td11"> <!--l. 239--><p class="noindent" >Cumulative density function                                                  </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-3-4-"><td  style="text-align:left;" id="TBL-3-4-1"  
class="td11"> <!--l. 240--><p class="noindent" ><span 
class="pcrr8t-x-x-109">runif(20, min=0, max=2)</span>  </p></td><td  style="text-align:left;" id="TBL-3-4-2"  
class="td11"> <!--l. 240--><p class="noindent" >Twenty random numbers from uniform [0,2]                         </p></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-3-5-"><td  style="text-align:left;" id="TBL-3-5-1"  
class="td11"> <!--l. 242--><p class="noindent" ><span 
class="pcrr8t-x-x-109">rpois(20, lambda=10)</span>     </p></td><td  style="text-align:left;" id="TBL-3-5-2"  
class="td11"> <!--l. 242--><p class="noindent" >Twenty                random                numbers                from
  Poisson(<!--l. 243--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mi 
>&lambda;</mi></math>) </p> </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-3-6-"><td  style="text-align:left;" id="TBL-3-6-1"  
class="td11">                          </td>
</tr></table></div><br 
class="newline" />
<!--l. 246--><p class="noindent" >
</p>
<h3 class="sectionHead"><span class="titlemark">1.4    </span> <a 
 id="x1-120001.4"></a>Two basic rules of experimental design and sampling</h3>
<!--l. 248--><p class="noindent" >In general, while designing experiments, and sampling from a <span 
class="ptmri8t-x-x-109">population</span>, there are two key (and simple)
rules:
     </p><dl class="enumerate"><dt class="enumerate">
   1. </dt><dd 
class="enumerate"><span 
class="ptmb8t-x-x-109">The more you sample, the more your sample&#x2019;s distribution will look like the population</span>
     <span 
class="ptmb8t-x-x-109">distribution </span>(obviously!)
     </dd><dt class="enumerate">
   2. </dt><dd 
class="enumerate"><span 
class="ptmb8t-x-x-109">The  more  you  sample,  the  closer  will  your  sample  statistic  be  to  the  population&#x2019;s</span>
     <span 
class="ptmb8t-x-x-109">statistical parameter </span>(the central limit theorem)</dd></dl>
<!--l. 258--><p class="noindent" >Let&#x2019;s have a quick look at rule 1 using R (open R and <span 
class="pcrr8t-x-x-109">setwd </span>to <span 
class="pcrr8t-x-x-109">Code</span>): </p><!--l. 260-->
<div class="lstlisting" id="listing-2"><span class="label"><a 
 id="x1-12003r1"></a></span><span 
class="pcrr8t-x-x-80">#</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">Draw</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">5</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">normal</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">random</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">nos</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">w</span><span 
class="pcrr8t-x-x-80">/</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">mean</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">0</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">and</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">s</span><span 
class="pcrr8t-x-x-80">.</span><span 
class="pcrr8t-x-x-80">d</span><span 
class="pcrr8t-x-x-80">.</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">1:</span><span 
class="pcrr8t-x-x-80"> </span><br /><span class="label"><a 
 id="x1-12004r2"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">MySample5</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">&#x003C;-</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">rnorm</span><span 
class="pcrr8t-x-x-80">(5,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">m</span><span 
class="pcrr8t-x-x-80">=0,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">sd</span><span 
class="pcrr8t-x-x-80">=1)</span><span 
class="pcrr8t-x-x-80"> </span><br /><span class="label"><a 
 id="x1-12005r3"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">MySample10</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">&#x003C;-</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">rnorm</span><span 
class="pcrr8t-x-x-80">(10,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">m</span><span 
class="pcrr8t-x-x-80">=0,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">sd</span><span 
class="pcrr8t-x-x-80">=1)</span><span 
class="pcrr8t-x-x-80"> </span><br /><span class="label"><a 
 id="x1-12006r4"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">MySample20</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">&#x003C;-</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">rnorm</span><span 
class="pcrr8t-x-x-80">(20,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">m</span><span 
class="pcrr8t-x-x-80">=0,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">sd</span><span 
class="pcrr8t-x-x-80">=1)</span><span 
class="pcrr8t-x-x-80"> </span><br /><span class="label"><a 
 id="x1-12007r5"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">MySample40</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">&#x003C;-</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">rnorm</span><span 
class="pcrr8t-x-x-80">(40,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">m</span><span 
class="pcrr8t-x-x-80">=0,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">sd</span><span 
class="pcrr8t-x-x-80">=1)</span><span 
class="pcrr8t-x-x-80"> </span><br /><span class="label"><a 
 id="x1-12008r6"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">MySample80</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">&#x003C;-</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">rnorm</span><span 
class="pcrr8t-x-x-80">(80,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">m</span><span 
class="pcrr8t-x-x-80">=0,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">sd</span><span 
class="pcrr8t-x-x-80">=1)</span><span 
class="pcrr8t-x-x-80"> </span><br /><span class="label"><a 
 id="x1-12009r7"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">MySample160</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">&#x003C;-</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">rnorm</span><span 
class="pcrr8t-x-x-80">(160,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">m</span><span 
class="pcrr8t-x-x-80">=0,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">sd</span><span 
class="pcrr8t-x-x-80">=1)</span>
</div>
<!--l. 269--><p class="noindent" >Now let&#x2019;s visualize these “samples”: </p><!--l. 270-->
<div class="lstlisting" id="listing-3"><span class="label"><a 
 id="x1-12010r1"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">par</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">mfcol</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">c</span><span 
class="pcrr8t-x-x-80">(2,3)</span><span 
class="pcrr8t-x-x-80">)</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">#</span><span 
class="pcrr8t-x-x-80">initialize</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">multi</span><span 
class="pcrr8t-x-x-80">-</span><span 
class="pcrr8t-x-x-80">paneled</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">plot</span><span 
class="pcrr8t-x-x-80"> </span><br /><span class="label"><a 
 id="x1-12011r2"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">par</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">mfg</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">c</span><span 
class="pcrr8t-x-x-80">(1,1)</span><span 
class="pcrr8t-x-x-80">)</span><span 
class="pcrr8t-x-x-80">;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">hist</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">MySample5</span><span 
class="pcrr8t-x-x-80">,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">col</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">rgb</span><span 
class="pcrr8t-x-x-80">(1,1,0)</span><span 
class="pcrr8t-x-x-80">,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">main</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8c-x-x-80">&#x2019;</span><span 
class="pcrr8t-x-x-80">n</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">5</span><span 
class="pcrr8c-x-x-80">&#x2019;</span><span 
class="pcrr8t-x-x-80">)</span><span 
class="pcrr8t-x-x-80"> </span><br /><span class="label"><a 
 id="x1-12012r3"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">par</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">mfg</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">c</span><span 
class="pcrr8t-x-x-80">(1,2)</span><span 
class="pcrr8t-x-x-80">)</span><span 
class="pcrr8t-x-x-80">;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">hist</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">MySample10</span><span 
class="pcrr8t-x-x-80">,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">col</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">rgb</span><span 
class="pcrr8t-x-x-80">(1,1,0)</span><span 
class="pcrr8t-x-x-80">,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">main</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8c-x-x-80">&#x2019;</span><span 
class="pcrr8t-x-x-80">n</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">10</span><span 
class="pcrr8c-x-x-80">&#x2019;</span><span 
class="pcrr8t-x-x-80">)</span><span 
class="pcrr8t-x-x-80"> </span><br /><span class="label"><a 
 id="x1-12013r4"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">par</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">mfg</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">c</span><span 
class="pcrr8t-x-x-80">(1,3)</span><span 
class="pcrr8t-x-x-80">)</span><span 
class="pcrr8t-x-x-80">;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">hist</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">MySample20</span><span 
class="pcrr8t-x-x-80">,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">col</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">rgb</span><span 
class="pcrr8t-x-x-80">(1,1,0)</span><span 
class="pcrr8t-x-x-80">,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">main</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8c-x-x-80">&#x2019;</span><span 
class="pcrr8t-x-x-80">n</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">20</span><span 
class="pcrr8c-x-x-80">&#x2019;</span><span 
class="pcrr8t-x-x-80">)</span><span 
class="pcrr8t-x-x-80"> </span><br /><span class="label"><a 
 id="x1-12014r5"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">par</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">mfg</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">c</span><span 
class="pcrr8t-x-x-80">(2,1)</span><span 
class="pcrr8t-x-x-80">)</span><span 
class="pcrr8t-x-x-80">;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">hist</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">MySample40</span><span 
class="pcrr8t-x-x-80">,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">col</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">rgb</span><span 
class="pcrr8t-x-x-80">(1,1,0)</span><span 
class="pcrr8t-x-x-80">,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">main</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8c-x-x-80">&#x2019;</span><span 
class="pcrr8t-x-x-80">n</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">40</span><span 
class="pcrr8c-x-x-80">&#x2019;</span><span 
class="pcrr8t-x-x-80">)</span><span 
class="pcrr8t-x-x-80"> </span><br /><span class="label"><a 
 id="x1-12015r6"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">par</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">mfg</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">c</span><span 
class="pcrr8t-x-x-80">(2,2)</span><span 
class="pcrr8t-x-x-80">)</span><span 
class="pcrr8t-x-x-80">;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">hist</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">MySample80</span><span 
class="pcrr8t-x-x-80">,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">col</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">rgb</span><span 
class="pcrr8t-x-x-80">(1,1,0)</span><span 
class="pcrr8t-x-x-80">,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">main</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8c-x-x-80">&#x2019;</span><span 
class="pcrr8t-x-x-80">n</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">80</span><span 
class="pcrr8c-x-x-80">&#x2019;</span><span 
class="pcrr8t-x-x-80">)</span><span 
class="pcrr8t-x-x-80"> </span><br /><span class="label"><a 
 id="x1-12016r7"></a></span><span 
class="pcrr8t-x-x-80">&#x003E;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">par</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">mfg</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">c</span><span 
class="pcrr8t-x-x-80">(2,3)</span><span 
class="pcrr8t-x-x-80">)</span><span 
class="pcrr8t-x-x-80">;</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">hist</span><span 
class="pcrr8t-x-x-80">(</span><span 
class="pcrr8t-x-x-80">MySample160</span><span 
class="pcrr8t-x-x-80">,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">col</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">rgb</span><span 
class="pcrr8t-x-x-80">(1,1,0)</span><span 
class="pcrr8t-x-x-80">,</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">main</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8c-x-x-80">&#x2019;</span><span 
class="pcrr8t-x-x-80">n</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">=</span><span 
class="pcrr8t-x-x-80"> </span><span 
class="pcrr8t-x-x-80">160</span><span 
class="pcrr8c-x-x-80">&#x2019;</span><span 
class="pcrr8t-x-x-80">)</span>
</div>
<!--l. 279--><p class="noindent" >The second rule above states that if I was to repeat even
<!--l. 279--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mi 
>n</mi> <mo 
class="MathClass-rel">=</mo> <mn>5</mn></math>
suﬃcient number of times, I would get a good <span 
class="ptmri8t-x-x-109">estimate </span>of mean (= 0) and standard deviation (= 1) of the
normal distribution we sampled from.
</p><!--l. 285--><p class="noindent" >
</p>
<h3 class="sectionHead"><span class="titlemark">1.5    </span> <a 
 id="x1-130001.5"></a>A data exploration case study</h3>
                                                                                 
                                                                                 
<!--l. 287--><p class="noindent" >As a case study, we will use data from a paper looking at the relationship between genome size and body
size across species of dragonﬂies and damselﬂies (<a 
href="http://en.wikipedia.org/wiki/Odonata" >Odonata</a>):
</p><!--l. 292--><p class="noindent" >
     </p><div class="quote">
     <!--l. 293--><p class="noindent" >Ardila-Garcia, AM &#x0026; Gregory, TR (2009) ‘An exploration of genome size diversity
     in dragonﬂies and damselﬂies (Insecta: Odonata)&#x2019; Journal of Zoology, 278, 163 - 173</p></div>
<!--l. 298--><p class="noindent" >You will work with the script ﬁle <span 
class="pcrr8t-x-x-109">ExpDesign.R</span>, which performs exploratory analyses on the data in
<span 
class="pcrr8t-x-x-109">GenomeSize.csv</span>. Let&#x2019;s go through the code block by block.
</p><!--l. 302--><p class="noindent" >
    </p><ul>
    <li class="compactitem">Get the script <span 
class="pcrr8t-x-x-109">ExpDesign.R </span>from the Bitbucket repository and put it in your own <span 
class="pcrr8t-x-x-109">Code</span>
    directory.
    </li>
    <li class="compactitem">Also get <span 
class="pcrr8t-x-x-109">GenomeSize.csv</span>
    </li>
    <li class="compactitem">Open the script <span 
class="pcrr8t-x-x-109">ExpDesign.R </span>in RStudio (or some other text editor).
    </li>
    <li class="compactitem">Use the shift and arrow keys to select the code in block (2), including the comments. Now
    use the keyboard short cut (look back at the R Chapters if you don&#x2019;t know how!) to run the
    highlighted block of code.</li></ul>
<!--l. 314--><p class="noindent" >This ﬁrst line (block (1)) reads in the data, as you have learned previously.
    </p><ul>
    <li class="compactitem">Now the code in block (2) line by line of code.</li></ul>
<!--l. 319--><p class="noindent" >Have a good look at the data. There are three factors (categorical variables): Suborder,
splitting the species into dragonﬂies (Anisoptera) and damselﬂies (Zygoptera); Family,
splitting the species further into 9 taxonomic families; and Species, giving the latin binomial
for each species in the table. The remaining columns are measurements of genome size
(in picograms) and measurements of body size and morphology (in grams, mm and
mm<!--l. 325--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><msup><mrow 
></mrow><mrow 
><mn>2</mn></mrow></msup 
></math>).
There are two columns ending with an N that show the sample size from which the observations for each
species are taken and a column ending SE showing standard errors.
</p><!--l. 330--><p class="noindent" >One thing you should see in the output from <span 
class="pcrr8t-x-x-109">head </span>or <span 
class="pcrr8t-x-x-109">str </span>is that there are some observations marked as
<span 
class="pcrr8t-x-x-109">NA </span>– this is the way R shows <span 
class="ptmri8t-x-x-109">missing data</span>. It is important to check how much missing data there is an
dataset, so we&#x2019;ll use another function that includes this information. Many R functions refuse to use
variables containing missing data — this is just R being careful and you can add <span 
class="pcrr8t-x-x-109">na.rm=TRUE </span>into most
functions to avoid this problem.
</p><!--l. 338--><p class="noindent" >
    </p><ul>
    <li class="compactitem">Run the <span 
class="pcrr8t-x-x-109">summary </span>line from the script window (block 3).</li></ul>
<!--l. 341--><p class="noindent" >Look at the output. There is a column for each variable: for factors, it provides a short table of the number
of observations in each level and for continuous variables, it provides some simple summary
statistics about the distribution (range, quartiles, mean and median), and the number of missing
                                                                                 
                                                                                 
values
</p><!--l. 347--><p class="noindent" >
</p>
<h4 class="subsectionHead"><span class="titlemark">1.5.1    </span> <a 
 id="x1-140001.5.1"></a>Visualise distributions of the variables</h4>
<!--l. 348--><p class="noindent" >The <span 
class="pcrr8t-x-x-109">summary </span>function shows us the basic distribution (range, quartiles, mean and median)
of a continuous variable, but this is easier to interpret if we visualise it. We&#x2019;ll look at two
ways:
     </p><dl class="description"><dt class="description">
<span 
class="ptmb8t-x-x-109">Histogram</span> </dt><dd 
class="description">In the simplest form, this shows the number of observations of the variable falling into
     a set of bins spanning the range of the variable. The option <span 
class="pcrr8t-x-x-109">breaks </span>allows you to change
     the number of bins.
     </dd><dt class="description">
<span 
class="ptmb8t-x-x-109">Density plot</span> </dt><dd 
class="description">Rather than showing blocks of counts, the density plot shows a continuous smooth
     line. This is a <span 
class="ptmri8t-x-x-109">smoothed </span>estimate of the how frequently data is observed across the range of
     values and the <span 
class="ptmri8t-x-x-109">bandwidth </span>(<span 
class="pcrr8t-x-x-109">bw=0.1</span>) controls the degree of the smoothing.</dd></dl>
<!--l. 363--><p class="noindent" >
    </p><ul>
    <li class="compactitem">Go to block (4) of the script and run each line separately, looking at the output.
    </li>
    <li class="compactitem">In  the  editor,  change  the  values  of  <span 
class="pcrr8t-x-x-109">breaks </span>and  <span 
class="pcrr8t-x-x-109">bw </span>—  for  example  <span 
class="pcrr8t-x-x-109">breaks=5 </span>and
    <span 
class="pcrr8t-x-x-109">bw=0.05 </span>— and re-run these lines to see how this aﬀects the graph. Basically, with both
    types of graph you can look at the data too coarsely or too ﬁnely.
    </li>
    <li class="compactitem">The graphs you&#x2019;ve just created look at genome size. Add a copy of those two lines of code
    in the script and change them to look at the variable <span 
class="pcrr8t-x-x-109">TotalLength</span>. You will need to alter
    the <span 
class="pcrr8t-x-x-109">density </span>function to ignore missing values (<span 
class="pcrr8t-x-x-109">na.rm=TRUE</span>) and to play around with the
    bandwidth. You should get something like this:</li></ul>
<!--l. 382--><p class="noindent" >
</p>
<h4 class="subsectionHead"><span class="titlemark">1.5.2    </span> <a 
 id="x1-150001.5.2"></a>Take a quick look at eﬀects of certain factors</h4>
<!--l. 383--><p class="noindent" >R has a special way of describing a model that deﬁnes the response variable and the explanatory variables
(“factors”). This is called a &#x2019;formula&#x2019; and is used to deﬁne linear models (more on these in a later
chapters). The same structure is used in many plotting functions and will put the response variable on the
<!--l. 387--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mi 
>y</mi></math> axis and the explanatory
variable on the <!--l. 388--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" ><mi 
>x</mi></math> axis. The
structure is “response variable <!--l. 388--><math 
 xmlns="http://www.w3.org/1998/Math/MathML"  
display="inline" > <mo 
class="MathClass-rel">&sim;</mo></math>
explanatory variables”. We will look at multiple explanatory variables in a later practical but an example
with one explantory variable (factor) is:
                                                                                 


