{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!--NAVIGATION-->\n",
    "< [Biological Computing in Python &ndash; I](05-Python_I.ipynb) | [Main Contents](Index.ipynb) | [Biological Computing in Python &ndash; II](07-R_I.ipynb)>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Biological Computing in R &ndash; I <span class=\"tocSkip\"> <a name=\"chap:R_I\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Outline-of-the-the-R-chapters\" data-toc-modified-id=\"Outline-of-the-the-R-chapters-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Outline of the the R chapters</a></span></li><li><span><a href=\"#What-is-R?\" data-toc-modified-id=\"What-is-R?-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>What is R?</a></span></li><li><span><a href=\"#Why-R?\" data-toc-modified-id=\"Why-R?-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Why R?</a></span></li><li><span><a href=\"#Would-you-ever-need-anything-other-than-R?\" data-toc-modified-id=\"Would-you-ever-need-anything-other-than-R?-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Would you ever need anything other than R?</a></span></li><li><span><a href=\"#Installing-R\" data-toc-modified-id=\"Installing-R-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Installing R</a></span></li><li><span><a href=\"#Getting-started}\" data-toc-modified-id=\"Getting-started}-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Getting started}</a></span></li><li><span><a href=\"#Gooey-IDEs!\" data-toc-modified-id=\"Gooey-IDEs!-7\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>Gooey IDEs!</a></span></li><li><span><a href=\"#R-basics\" data-toc-modified-id=\"R-basics-8\"><span class=\"toc-item-num\">8&nbsp;&nbsp;</span>R basics</a></span><ul class=\"toc-item\"><li><span><a href=\"#Useful-R-commands\" data-toc-modified-id=\"Useful-R-commands-8.1\"><span class=\"toc-item-num\">8.1&nbsp;&nbsp;</span>Useful R commands</a></span></li></ul></li><li><span><a href=\"#Baby-steps\" data-toc-modified-id=\"Baby-steps-9\"><span class=\"toc-item-num\">9&nbsp;&nbsp;</span>Baby steps</a></span></li><li><span><a href=\"#Operators}\" data-toc-modified-id=\"Operators}-10\"><span class=\"toc-item-num\">10&nbsp;&nbsp;</span>Operators}</a></span></li><li><span><a href=\"#When-things-go-wrong}\" data-toc-modified-id=\"When-things-go-wrong}-11\"><span class=\"toc-item-num\">11&nbsp;&nbsp;</span>When things go wrong}</a></span></li><li><span><a href=\"#Types-of-parentheses}\" data-toc-modified-id=\"Types-of-parentheses}-12\"><span class=\"toc-item-num\">12&nbsp;&nbsp;</span>Types of parentheses}</a></span></li><li><span><a href=\"#Variable-Types}\" data-toc-modified-id=\"Variable-Types}-13\"><span class=\"toc-item-num\">13&nbsp;&nbsp;</span>Variable Types}</a></span></li><li><span><a href=\"#Type-Conversion-and-Special-Values}\" data-toc-modified-id=\"Type-Conversion-and-Special-Values}-14\"><span class=\"toc-item-num\">14&nbsp;&nbsp;</span>Type Conversion and Special Values}</a></span></li><li><span><a href=\"#Data-Structure-types}\" data-toc-modified-id=\"Data-Structure-types}-15\"><span class=\"toc-item-num\">15&nbsp;&nbsp;</span>Data Structure types}</a></span></li><li><span><a href=\"#Vectors}\" data-toc-modified-id=\"Vectors}-16\"><span class=\"toc-item-num\">16&nbsp;&nbsp;</span>Vectors}</a></span></li><li><span><a href=\"#Matrices-and-arrays`\" data-toc-modified-id=\"Matrices-and-arrays`-17\"><span class=\"toc-item-num\">17&nbsp;&nbsp;</span>Matrices and arrays`</a></span></li><li><span><a href=\"#Data-frames}\" data-toc-modified-id=\"Data-frames}-18\"><span class=\"toc-item-num\">18&nbsp;&nbsp;</span>Data frames}</a></span></li><li><span><a href=\"#Lists`\" data-toc-modified-id=\"Lists`-19\"><span class=\"toc-item-num\">19&nbsp;&nbsp;</span>Lists`</a></span></li><li><span><a href=\"#Creating-and-manipulating-data-structures}\" data-toc-modified-id=\"Creating-and-manipulating-data-structures}-20\"><span class=\"toc-item-num\">20&nbsp;&nbsp;</span>Creating and manipulating data structures}</a></span></li><li><span><a href=\"#Creating-Sequences}\" data-toc-modified-id=\"Creating-Sequences}-21\"><span class=\"toc-item-num\">21&nbsp;&nbsp;</span>Creating Sequences}</a></span></li><li><span><a href=\"#Acessing-parts-of-data-stuctures----Indices-and-Indexing}\" data-toc-modified-id=\"Acessing-parts-of-data-stuctures----Indices-and-Indexing}-22\"><span class=\"toc-item-num\">22&nbsp;&nbsp;</span>Acessing parts of data stuctures -- Indices and Indexing}</a></span></li><li><span><a href=\"#Recycling}\" data-toc-modified-id=\"Recycling}-23\"><span class=\"toc-item-num\">23&nbsp;&nbsp;</span>Recycling}</a></span></li><li><span><a href=\"#Basic-vector-matrix-operations}\" data-toc-modified-id=\"Basic-vector-matrix-operations}-24\"><span class=\"toc-item-num\">24&nbsp;&nbsp;</span>Basic vector-matrix operations}</a></span></li><li><span><a href=\"#Strings-and-Pasting}\" data-toc-modified-id=\"Strings-and-Pasting}-25\"><span class=\"toc-item-num\">25&nbsp;&nbsp;</span>Strings and Pasting}</a></span></li><li><span><a href=\"#Your-analysis-workflow}\" data-toc-modified-id=\"Your-analysis-workflow}-26\"><span class=\"toc-item-num\">26&nbsp;&nbsp;</span>Your analysis workflow}</a></span></li><li><span><a href=\"#The-R-Workspace-and-Working-Directory}\" data-toc-modified-id=\"The-R-Workspace-and-Working-Directory}-27\"><span class=\"toc-item-num\">27&nbsp;&nbsp;</span>The R Workspace and Working Directory}</a></span></li><li><span><a href=\"#Importing-and-Exporting-Data}\" data-toc-modified-id=\"Importing-and-Exporting-Data}-28\"><span class=\"toc-item-num\">28&nbsp;&nbsp;</span>Importing and Exporting Data}</a></span></li><li><span><a href=\"#Relative-paths!`\" data-toc-modified-id=\"Relative-paths!`-29\"><span class=\"toc-item-num\">29&nbsp;&nbsp;</span>Relative paths!`</a></span></li><li><span><a href=\"#Writing-out-to-and-saving-files}\" data-toc-modified-id=\"Writing-out-to-and-saving-files}-30\"><span class=\"toc-item-num\">30&nbsp;&nbsp;</span>Writing out to and saving files}</a></span></li><li><span><a href=\"#Writing-R-code}\" data-toc-modified-id=\"Writing-R-code}-31\"><span class=\"toc-item-num\">31&nbsp;&nbsp;</span>Writing R code}</a></span></li><li><span><a href=\"#Running-R-code}\" data-toc-modified-id=\"Running-R-code}-32\"><span class=\"toc-item-num\">32&nbsp;&nbsp;</span>Running R code}</a></span></li><li><span><a href=\"#Writing-R-Functions}\" data-toc-modified-id=\"Writing-R-Functions}-33\"><span class=\"toc-item-num\">33&nbsp;&nbsp;</span>Writing R Functions}</a></span></li><li><span><a href=\"#Practicals}\" data-toc-modified-id=\"Practicals}-34\"><span class=\"toc-item-num\">34&nbsp;&nbsp;</span>Practicals}</a></span></li><li><span><a href=\"#Control-statements}\" data-toc-modified-id=\"Control-statements}-35\"><span class=\"toc-item-num\">35&nbsp;&nbsp;</span>Control statements}</a></span></li><li><span><a href=\"#Useful-R-Functions}\" data-toc-modified-id=\"Useful-R-Functions}-36\"><span class=\"toc-item-num\">36&nbsp;&nbsp;</span>Useful R Functions}</a></span></li><li><span><a href=\"#Mathematical}\" data-toc-modified-id=\"Mathematical}-37\"><span class=\"toc-item-num\">37&nbsp;&nbsp;</span>Mathematical}</a></span></li><li><span><a href=\"#Strings}\" data-toc-modified-id=\"Strings}-38\"><span class=\"toc-item-num\">38&nbsp;&nbsp;</span>Strings}</a></span></li><li><span><a href=\"#Statistical}\" data-toc-modified-id=\"Statistical}-39\"><span class=\"toc-item-num\">39&nbsp;&nbsp;</span>Statistical}</a></span></li><li><span><a href=\"#Packages}\" data-toc-modified-id=\"Packages}-40\"><span class=\"toc-item-num\">40&nbsp;&nbsp;</span>Packages}</a></span></li><li><span><a href=\"#Practicals-wrap-up}\" data-toc-modified-id=\"Practicals-wrap-up}-41\"><span class=\"toc-item-num\">41&nbsp;&nbsp;</span>Practicals wrap-up}</a></span></li><li><span><a href=\"#Readings}\" data-toc-modified-id=\"Readings}-42\"><span class=\"toc-item-num\">42&nbsp;&nbsp;</span>Readings}</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Outline of the the R chapters \n",
    "\n",
    "You will use R a lot during the rest of your courses, your thesis  dissertation, and very likely, your career. The R part of this course aims to lay down the foundations for you to become comfortable with it. Specifically,\n",
    "\n",
    "* Giving you an introduction to R syntax and programming conventions, assuming you have never set your eyes on R\n",
    "* Teaching you principles of data processing and exploration (including visualization) using R\n",
    "* Teaching you principles of clean and efficient programming using R\n",
    "* Teaching you how to generate publication quality graphics in R \n",
    "* Teaching you how to develop reproducible data analysis \"work flows\" so you (or anybody else) can run and re-run your analyses, graphics outputs and all, in R \n",
    "\n",
    "\n",
    "## What is R?\n",
    "\n",
    "R is a freely available statistical software with strong  programming capabilities widely used by professional scientists around  the world. It was based on the commercial statistical software `S` by Robert Gentleman and Ross Ihaka. The first stable version appeared  in 2000. \n",
    "\n",
    "R was essentially designed for *programming* statistical  analysis and data-mining. It became the standard tool for data  analysis and visualization in biology in a matter of just 10 years or so. It is also increasingly being used for modelling in biology. This is because R has: \n",
    "\n",
    "* many tried and tested packages to perform practically all statistical analysis\n",
    "* numerous packages for data-handling and processing   \n",
    "* excellent graphing and visualization capabilities\n",
    "* good capabilities for mathematical calculations, including matrix algebra \n",
    "\n",
    "## Why R?\n",
    "\n",
    "There are many commercial statistical (minitab, SPSS, etc) software  packages in the world that are mouse-driven, warm, and friendly, and  have lots of statistical tests and plotting/graphing capabilities. Why not just use them? Here are some very good reasons:\n",
    "\n",
    "* R is scriptable, so you can build a perfectly repeatable record of your analysis. This in itself has several advantages:\n",
    "    * You can never replicate *exactly* the same analysis with all the same steps using a point-and-click approach/software. With R you can reproduce your full analysis for yourself (in the future!), your colleagues, your \n",
    "supervisor/employer, and any journal you might want to submit your work to.\n",
    "    * You may need to rerun your analysis every time you get new data. Once you have it all in a R script, you can just rerun your \tanalysis and go home!\n",
    "    * You may need to tweak your analysis many times (new data, supervisor changes mind, you change mind, paper reviewers want you do something differently). Having the analysis recorded as script then allows you to do so by revising the relevant parts of your analysis with relatively little pain.  \n",
    "* R provides basically every statistical test you'll ever need and is constantly being improved --  you can tailor your analyses rather than trying to use the more limited options each statistical software package can offer. \n",
    "* R can produce publication-quality graphics that can be re-produced with scripts -- you won't get RSI mouse-clicking your way though graphing and re-graphing your data every time you change your analysis!\n",
    "* R is freely available for all common computer operating systems -- if you want a copy on your laptop, help yourself  at the [CRAN website](https://cran.r-project.org).\n",
    "*Thus, being able to program in R means you can develop and automate your own data handling, statistical analysis, and graphing/plotting, a set of skills you are likely to need in many, if not most careers paths!}\n",
    "\n",
    "## Would you ever need anything other than R?\n",
    "\n",
    "Being able to program R means you can develop and automate your statistical analyses and the generation of figures into a reproducible work flow. For many of you, using R as your only programming language will do the job. However, if your work also includes extensive numerical simulations, manipulation of very large matrices, bioinformatics, relational database access and manipulation, or web development, you will be better-off *also* knowing another programming language that is more versatile and computationally efficient (like python or C).\n",
    "\n",
    "## Installing R\n",
    "\n",
    "If you are using a college computer, R will likely already be available. Otherwise you can install R on your own computer as follows:  \n",
    "\n",
    "On linux/ubuntu, run the following in terminal:\n",
    "\n",
    "```bash\n",
    "sudo apt install r-base r-base-dev\n",
    "```\n",
    "\n",
    "On Mac OS X, download and install from:\n",
    "\\url{https://cran.r-project.org/bin/macosx/}\n",
    "\n",
    "On Windows, download and install from:\n",
    "\\url{https://cran.r-project.org/bin/windows/base/}\n",
    "\n",
    "## Getting started}\n",
    "\n",
    "Let's briefly look at the bare-bones R interface and command line  interface (CLI), and then switch to a more Interactive Development  Environment (IDE) like vscode or RStudio. \n",
    "\n",
    "Launch R (From Applications menu on Window or Mac, from terminal in  Linux/Ubuntu) --- it should look something like this (on Linux/Ubuntu or Mac terminal):\n",
    "\n",
    "\n",
    "<figure>\n",
    "    <img src=\"./graphics/R_Linux.png\" alt=\"Unix R\" style=\"width:50%\">\n",
    "</figure>\n",
    "\n",
    "Or like this (Windows \"console\", similar in Mac):\n",
    " \n",
    "<figure>\n",
    "    <img src=\"./graphics/HorrorVacui.png\" alt=\"R console\" style=\"width:50%\">\n",
    "</figure>\n",
    "\n",
    "\n",
    "## Gooey IDEs!\n",
    "\n",
    "To develop your R code, you should ideally use an IDE (Interactive Development Environment) that offers useful features like syntax highlighting (google it!) and embedded data and plot viewing windows, such as RStudio, vscode vim}, etc. \n",
    "\n",
    "These IDEs come with graphic user interfaces (GUI's, or ``gooeys'') of differing levels of sophistication and shiny-ness.\n",
    "\n",
    "In fact, one of the reasons that R has become so popular is that there are some very nice, freely available GUI IDE's for it where you can use your mouse to do certain \"stuff\". \n",
    "\n",
    "In particular, [http://rstudio.org](RStudio) is very good. \n",
    "\n",
    "A big advantage of something like RStudio is that you will get ``syntax highlighting'' wherein R language elements such as variables, commands, and brackets are differently colored. This is very handy and will make your R programming far more convenient and error-free. If you are using your own desktop/laptop, you can download it freely from \n",
    "<https://www.rstudio.com> and install (versions are available for all computer platforms). If you are suing a college computer, it should already be available.\n",
    "\n",
    "The R chapters do not require you to use RStudio (everything will be command line based), but I would recommend that you to use it, or some other IDE like vscode or emacs. \n",
    "\n",
    "Here's what the RStudio window looks like: \n",
    "\n",
    "\n",
    "<figure>\n",
    "    <img src=\"./graphics/RStudio.png\" alt=\"Rstudio\" style=\"width:50%\">\n",
    "</figure>\n",
    "\n",
    "You are still in the R environment, but with some useful frills and  additional thrills. There are four main elements here (you can hide or resize any of them using the RStudio GUI options):\n",
    "\n",
    "* *Console*: This is the same R terminal you saw above. In this week we will work mainly in this console\n",
    "\n",
    "* *Source code*: This panel is the code editor with syntax  highlighting and other handy features that we will explore soon. This  is where you will write your scripts/programs, and save them. To save,  you will use `Ctrl + S` and to execute it you will use `Ctrl + Shift + S`\n",
    "\n",
    "* *Environment* This panel lists all the variables you created (more on this later). It has another tab that shows you the history of  the commands you typed\n",
    "\n",
    "* *Plots* This panel shows you all the plots you drew. Other tabs in this panel allow you to access the list of packages you have loaded, and the help page for commands (just type `\n",
    "help(command_name)` in the Console) and packages\n",
    "\n",
    "\n",
    "In RStudio, you can feed your mouse habit, but I strongly suggest using the terminal/console, and resist, to the extent possible, the  temptation to do everything using the shiny buttons you see (that is  the dark side pulling, my young *padawan*).\n",
    "\n",
    "## R basics\n",
    "\n",
    "Gets get started with some R basics. You will be working by  entering R commands interactively at the R user prompt (`>`). Up and down arrow keys scroll through your command history. \n",
    "\n",
    "### Useful R commands\n",
    "\n",
    "|||\n",
    "|:-|:-|\n",
    "|\t` ls()`| list all the variables in the work space |\n",
    "|\t` rm('a', 'b')`| remove variable(s) ` a`and ` b}|\n",
    "|\t` rm(list=ls())`| remove all variable(s)|\n",
    "|\t` getwd()`| get current working directory |\n",
    "|\t` setwd('Path')`| set working directory to ` Path`|\n",
    "|\t` q()`| quit R |\n",
    "|\t` ?Command`| show the documentation of ` Command`|\n",
    "|\t` ??Keyword`| search the all packages/functions with `Keyword', ``fuzzy search''|\n",
    "\n",
    "\n",
    "## Baby steps\n",
    "\n",
    "Like in any programming language, you will need to use \"variables\" to store information in a R session's workspace. Each Variable has a reserved location in your [RAM](https://en.wikipedia.org/wiki/Random-access_memory), and takes up \"real estate\" in it --- that is when you create a variable you reserve some space in your computer's memory.\n",
    "\n",
    "$\\star$ Now, let's try assigning a few variables in the R and doing things to them:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "a <- 4  # store 4 as variable a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4"
      ],
      "text/latex": [
       "4"
      ],
      "text/markdown": [
       "4"
      ],
      "text/plain": [
       "[1] 4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "a # display it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "16"
      ],
      "text/latex": [
       "16"
      ],
      "text/markdown": [
       "16"
      ],
      "text/plain": [
       "[1] 16"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "a*a # product"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_squared <- a*a "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4"
      ],
      "text/latex": [
       "4"
      ],
      "text/markdown": [
       "4"
      ],
      "text/plain": [
       "[1] 4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sqrt(a_squared) # square root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "v <- c(0, 1, 2, 3, 4) # build a vector with c (for \"concatenate\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that any text after a \"#\" is ignored by R, like in many other languages --- handy for \n",
    "commenting. \n",
    "\n",
    "*In general, please comment your code and scripts, for *everybody's* sake. You will be amazed by how difficult it is to read and understand what a certain R script does (or any other script, for that matter) without judicious comments --- even scripts you yourself wrote not so long ago!\n",
    "\n",
    "`c()`(concatenate) is one of the most commonly used functions --- it will appear again and again! (try `?c` at the R prompt). \n",
    "\n",
    "OK, continuing with our R warmup:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>0</li>\n",
       "\t<li>1</li>\n",
       "\t<li>2</li>\n",
       "\t<li>3</li>\n",
       "\t<li>4</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 3\n",
       "\\item 4\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0\n",
       "2. 1\n",
       "3. 2\n",
       "4. 3\n",
       "5. 4\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0 1 2 3 4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "v # Display the vector variable you created"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "is.vector(v) # check if it's a vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean(v) # mean"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Thus, a *vector* is like a single column or row in a *spreadsheet* (just like it is in [Python](05-Python_I.ipynb)). Multiple vectors can be combined to make a matrix (the full spreadsheet).  \n",
    "\n",
    "This is one of many ways R stores and processes data. More on R data types and objects below.  \n",
    "\n",
    "A single value (any kind) is a vector object of length 1 by default. That's why in the R console you see ` [1]` before any single-value output (e.g., type `8`, and you will see `[1] 8`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "2.5"
      ],
      "text/latex": [
       "2.5"
      ],
      "text/markdown": [
       "2.5"
      ],
      "text/plain": [
       "[1] 2.5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "var(v) # variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "median(v) # median"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(v) # sum all elements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prod(v + 1) # multiply"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "length(v) # length of vector"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variable names and Tabbing\n",
    "\n",
    "In R, you can name variables in the following way to keep track of related variables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wing.width.cm <- 1.2 #Using dot notation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wing.length.cm <- c(4.7, 5.2, 4.8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This can be handy; type:\n",
    "\n",
    "```r\n",
    "wing.\n",
    "```\n",
    "\n",
    "And then hit the ` tab`key to reveal all variables in that  category. This is nice --- variable names should be as obvious as possible. However, they should not be over-long either! Good style and readability is more important than just convenient variable names. \n",
    "\n",
    "### Operators\n",
    "The usual ``operators'' are available in R:\n",
    "\n",
    "|||\n",
    "|:-|:-|\n",
    "| ` +`| Addition \\\\\n",
    "| ` -`| Subtraction\\\\\n",
    "| ` *`| Multiplication\\\\\n",
    "| ` /`| Division\\\\\n",
    "| ` \\textasciicircum`| Power\\\\\n",
    "| ` \\%\\%`& Modulo\\\\\n",
    "| ` \\%/\\%`& Integer division\\\\\n",
    "| ` ==`& Equals\\\\\n",
    "| ` !=`& Differs\\\\\n",
    "| ` $>$`& Greater\\\\\n",
    "| ` $>$=`& Greater or equal\\\\\n",
    "| ` \\&`& Logical and\\\\\n",
    "| ` $\\vert$`& Logical or\\\\\n",
    "| ` !`& Logical not\\\\\n",
    "\n",
    "\n",
    "## When things go wrong}\n",
    "\n",
    "Syntax errors are those where you've just made a mistake while typing \n",
    "code. Here are some common problems in R: \n",
    "\n",
    "\n",
    "* missing close bracket leads to  continuation line.\n",
    "\n",
    "> x <- (1 + (2 * 3)\n",
    "+ \n",
    "\n",
    "Hit ` Ctrl-C`(UNIX terminal or base R command line) or ESC (in in RStudio) or keep typing!\n",
    "\n",
    "* Too many parentheses: ` 2 + (2*3))}\n",
    "\n",
    "* Wrong or mismatched brackets (see next subsection)\n",
    "\n",
    "* Do not mix double quotes and single quotes\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\tWhen things are taking too long and the R consile seems frozen, try \n",
    "\t` Ctrl + C`(UNIX terminal or base R command line) or ESC (in in RStudio) to force an exit from whatever is going on.\n",
    "\n",
    "\n",
    "## Types of parentheses}\n",
    "R has specific uses for different types of parentheses that you need to \n",
    "get used to:\n",
    "\n",
    "\n",
    " ` f(3,4)`& call the function (or command) f, with the \n",
    "\targuments 3 \\& 4. \\\\\n",
    " ` a + (b*c)`& to enforce order over which statements \n",
    "  or calculations are executed. Here ` (b*c)`is executed before \n",
    "  adding to ` a}; here is an alternative order: ` (a + b)*c}\\\\\n",
    " ` \\{ expr1; expr2; \\ldots exprn \\}`& group a set of expressions or \n",
    " commands into one compound expression. Value returned is value of last\n",
    " expression; used in building function, loops, and conditionals (more on \n",
    " these soon!).\\\\\n",
    " ` x[4] `& get the 4th element of the vector ` x}.\\\\\n",
    " ` li[[3]]`& get the 3rd element of some list ` li}, and return it.\n",
    " (compare with li[3], which returns a list with just the 3rd element\n",
    " inside). More on lists in next section\n",
    "\n",
    "\n",
    "## Variable Types}\n",
    "\n",
    "There are different kinds of data variable types in R, but you will \n",
    "basically need to know four for most of your work: integer, float (or \n",
    "``numeric'', including real numbers), string (or ``character'', e.g., \n",
    "text), and Boolean (``logical''; ` True`or ` False}).  Try this:\n",
    "\n",
    "\n",
    "> v <- TRUE\n",
    "> class(v)\n",
    "[1] \"logical\"\n",
    "\n",
    "> v <- 3.2\n",
    "> class(v)\n",
    "[1] \"numeric\"\n",
    "\n",
    "> v <- 2L\n",
    "> class(v)\n",
    "[1] \"integer\"\n",
    "\n",
    "> v <- \"A string\"\n",
    "> class(v)\n",
    "[1] \"character\"\n",
    "\n",
    "\n",
    "\n",
    "R will use ` E`notation in outputs of statistical tests to display \n",
    "very large or small numbers. If you are not used to different \n",
    "representations of long numbers, the ` E`notation might be \n",
    "confusing. Try this:\n",
    "\n",
    "> 1E4\n",
    "[1] 10000\n",
    "> 1e4\n",
    "[1] 10000\n",
    "> 5e-2\n",
    "[1] 0.05\n",
    "> 1E4^2\n",
    "[1] 1e+08\n",
    "> 1 / 3 / 1e8\n",
    "[1] 3.333333e-09\n",
    "\n",
    "Keep and eye out for ` E`notation! \n",
    "\n",
    "\n",
    "## Type Conversion and Special Values}\n",
    "In the following examples, the ` as.*`commands all convert a \n",
    "variable from one type to another:\t\n",
    "\n",
    "> as.integer(3.1)\n",
    "[1] 3\n",
    "> as.numeric(4)\n",
    "[1] 4\n",
    "> as.roman(155)\n",
    "[1] CLV\n",
    "> as.character(155) # same as converting to string\n",
    "[1] \"155\"\n",
    "> as.logical(5) #what's happening here?!\n",
    "[1] TRUE\n",
    "> as.logical(0)\n",
    "[1] FALSE\n",
    "> b <- NA\n",
    "> is.na(b)\n",
    "[1] TRUE\n",
    "> b <- 0/0\n",
    "> b\n",
    "[1] NaN\n",
    "> is.nan(b)\n",
    "[1] TRUE\n",
    "> b <- 5/0\n",
    "> b\n",
    "[1] Inf\n",
    "> is.nan(b)\n",
    "[1] FALSE\n",
    "> is.infinite(b)\n",
    "[1] TRUE\n",
    "> is.finite(b)\n",
    "[1] FALSE\n",
    "> is.finite(0/0)\n",
    "[1] FALSE\n",
    "\n",
    "\n",
    "\n",
    "Beware of the difference between ` NA`(` N}ot ` A}vailable) \n",
    "and ` NaN`(` N}ot ` a`` N}umber)!\n",
    "\n",
    "R will use ` NA`to represent/identify missing values in data or \n",
    "outputs, while  ` NaN`represent nonsense values (e.g., 0/0) that \n",
    "cannot be represented as a number or some other data type. See what R \n",
    "has to say about this: try ` ?is.nan}, ` ?is.na}, ` ?NA}, ` \n",
    "?NaN`in the R commandline (one at a time!).\n",
    "\n",
    "There are also ` Inf`(Infinity, e.g., 1/0), and ` NULL`\n",
    "(variable not set) value types. Look these up as well using ` ?}.\n",
    "\n",
    "\n",
    "## Data Structure types}\n",
    "R comes with different built-in structures (objects) for data storage and \n",
    "manipulation. Mastering these, and knowing which one to use when will \n",
    "help you write better, more efficient programs and also handle diverse \n",
    "datasets (numbers, counts, names, dates, etc). \n",
    "\n",
    "## Vectors}\n",
    "The Vector, which you first saw above, is a fundamental data object in \n",
    "R. Scalars (single data values) are treated as vector of length 1. *\n",
    "A vector is like a single column or row in a spreadsheet.`Now get back \n",
    "into R (if you somehow quit R using ` q()`or something else), and \n",
    "try this:\n",
    "\n",
    "> a <- 5\n",
    "> is.vector(a)\n",
    "[1] TRUE\n",
    "> v1 <- c(0.02, 0.5, 1)\n",
    "> v2 <- c(\"a\", \"bc\", \"def\", \"ghij\")\n",
    "> v3 <- c(TRUE, TRUE, FALSE)\n",
    "\n",
    "R vectors can only store data of a single type (e.g., all numeric \n",
    "or all character). If you try to combine different types, R will \n",
    "homogenize everything to the same data type. To see this, try the \n",
    "following:\n",
    "\n",
    "> v1 <- c(0.02, TRUE, 1)\n",
    "> v1 # TRUE gets converted to 1.00!\n",
    "[1] 0.02 1.00 1.00\n",
    "> v1 <- c(0.02, \"Mary\", 1)\n",
    "> v1 # Everything gets converted to text!\n",
    "[1] \"0.02\" \"Mary\" \"1\"\n",
    "\n",
    "\n",
    "## Matrices and arrays`\n",
    "\n",
    "A R ` matrix`is a 2 dimensional vector (has both rows and columns). \n",
    "and an ` R array`is can store data in more than two dimensions \n",
    "(e.g., a stack of 2-D matrices). \n",
    "\n",
    "R has many functions to build and manipulate matrices and arrays. \n",
    "Try:\n",
    "\n",
    "\n",
    "> mat1 <- matrix(1:25, 5, 5)\n",
    "> mat1\n",
    "\t [,1] [,2] [,3] [,4] [,5]\n",
    "[1,] 1 6   11   16   21\n",
    "[2,] 2 7   12   17   22\n",
    "[3,] 3 8   13   18   23\n",
    "[4,] 4 9   14   19   24\n",
    "[5,] 5   10   15   20   25\n",
    "> mat1 <- matrix(1:25, 5, 5, byrow=TRUE)\n",
    "> mat1\n",
    "\t [,1] [,2] [,3] [,4] [,5]\n",
    "[1,] 1 2 3 4 5\n",
    "[2,] 6 7 8 9   10\n",
    "[3,]   11   12   13   14   15\n",
    "[4,]   16   17   18   19   20\n",
    "[5,]   21   22   23   24   25\n",
    "> dim(mat1) #get the size of the matrix\n",
    "[1] 5 5\n",
    "\n",
    "\n",
    "Make an array consisting of two 5$\\times$5 matrices containing the \n",
    "integers 1--50:\n",
    "\n",
    "> arr1 <- array(1:50, c(5, 5, 2))\n",
    "> arr1\n",
    ", , 1\n",
    "\n",
    "\t [,1] [,2] [,3] [,4] [,5]\n",
    "[1,] 1 6   11   16   21\n",
    "[2,] 2 7   12   17   22\n",
    "[3,] 3 8   13   18   23\n",
    "[4,] 4 9   14   19   24\n",
    "[5,] 5   10   15   20   25\n",
    "\n",
    ", , 2\n",
    "\n",
    "\t [,1] [,2] [,3] [,4] [,5]\n",
    "[1,]   26   31   36   41   46\n",
    "[2,]   27   32   37   42   47\n",
    "[3,]   28   33   38   43   48\n",
    "[4,]   29   34   39   44   49\n",
    "[5,]   30   35   40   45   50\n",
    "\n",
    "\n",
    "Just like R vectors, R matrices and arrays have to be of a \n",
    "homogeneous type, and R will do the same sort of type homogenization \n",
    "you saw for R vectors above (try inserting a text value in ` mat1`and \n",
    "see what happens), and in ` python}'s ` numpy`array and matrix \n",
    "data structures. \n",
    "\n",
    "## Data frames}\n",
    "\n",
    "This is a very important data structure in R. Unlike matrices and \n",
    "vectors, R data frames can store data in which each column \n",
    "contains a different data type (e.g., numbers, strings, boolean) or \n",
    "even a combination of data types, just like a standard spreadsheet. \n",
    "Indeed, the dataframe data type was built to emulate some of the \n",
    "convenient properties of spreadsheets. Many statistical and plotting \n",
    "functions and packages in R naturally use data frames. Let's build and \n",
    "manipulate a dataframe:\n",
    "\n",
    "> Col1 <- 1:10\n",
    "> Col1\n",
    " [1]  1  2  3  4  5  6  7  8  9 10\n",
    "> Col2 <- LETTERS[1:10]\n",
    "> Col2\n",
    " [1] \"A\" \"B\" \"C\" \"D\" \"E\" \"F\" \"G\" \"H\" \"I\" \"J\"\n",
    "> Col3 <- runif(10) # 10 random numbers from a uniform distribution\n",
    "> Col3\n",
    " [1] 0.29109 0.91495 0.64962 0.95503 0.26589 0.02482 0.59718\n",
    " [8] 0.99134 0.98786 0.86168\n",
    "> MyDF <- data.frame(Col1, Col2, Col3)\n",
    "> MyDF\n",
    "\t Col1 Col2   Col3\n",
    "1  1 A 0.2910981\n",
    "2  2 B 0.9149558\n",
    "3  3 C 0.6496248\n",
    "4  4 D 0.9550331\n",
    "5  5 E 0.2658936\n",
    "6  6 F 0.0248217\n",
    "7  7 G 0.5971868\n",
    "8  8 H 0.9913407\n",
    "9  9 I 0.9878679\n",
    "10   10 J 0.8616854\n",
    "> names(MyDF) <- c(\"A.name\", \"another\", \"another.one\")\n",
    "> MyDF\n",
    "\t A.name another another.one\n",
    "1    1    A   0.2910981\n",
    "2    2    B   0.9149558\n",
    "3    3    C   0.6496248\n",
    "4    4    D   0.9550331\n",
    "5    5    E   0.2658936\n",
    "6    6    F   0.0248217\n",
    "7    7    G   0.5971868\n",
    "8    8    H   0.9913407\n",
    "9    9    I   0.9878679\n",
    "10  10    J   0.8616854\n",
    "\n",
    "\n",
    "Unlike matrices, you can access the contents of data frames by naming \n",
    "the columns using a \\$ sign: \n",
    "\n",
    "> MyDF$A.name\n",
    " [1]  1  2  3  4  5  6  7  8  9 10\n",
    "> MyDF[,1] #using numerical indexing instead\n",
    " [1]  1  2  3  4  5  6  7  8  9 10\n",
    "> MyDF[c(\"A.name\",\"another\")] # show two specific columns only\n",
    "\t A.name another\n",
    "1    1    A\n",
    "2    2    B\n",
    "3    3    C\n",
    "4    4    D\n",
    "5    5    E\n",
    "6    6    F\n",
    "7    7    G\n",
    "8    8    H\n",
    "9    9    I\n",
    "10  10    J\n",
    "\n",
    "\n",
    "You can check whether a particular object is a dataframe data structure with:\n",
    "\n",
    "> class(MyDF)\n",
    "[1] \"data.frame\"\n",
    "\n",
    "You can check the structure of a dataframe with ` str()}:\n",
    "\n",
    "> str(MyDF)\n",
    "'data.frame':\t10 obs. of  3 variables:\n",
    " $ A.name  : int  1 2 3 4 5 6 7 8 9 10\n",
    " $ another : Factor w/ 10 levels \"A\",\"B\",\"C\",\"D\",..\n",
    " $ another.one: num  0.291 0.915 0.65 0.955 0.266 ...\n",
    "\n",
    "You can print the column names and top few rows with ` head()},\n",
    "\n",
    "> head(MyDF)\n",
    "\n",
    "\n",
    "And the bottom few rows with ` tail()},\n",
    "\n",
    "> tail(MyDF)\n",
    "\n",
    "\n",
    "\n",
    "\tWhat is the ``factor'' data type in the data frame you created? You \n",
    "\twill see that R may mysteriously convert certain columns in dataframes \n",
    "\tto a factor type, which has \"levels\".  \n",
    "\t\n",
    "\tR has a special data type called `factor'. It basically considers \n",
    "\tcolumns containing strings-only in a data frame to be a grouping \n",
    "\tvariable. You can convert vectors to and from the ` factor`class. \n",
    "\tTry the following:\n",
    "\t\n",
    "\t\n",
    "\t> a <- as.factor(2)\n",
    "\t> a\n",
    "\t[1] 2\n",
    "\tLevels: 2\n",
    "\t> class(a)\n",
    "\t[1] \"factor\"\n",
    "\t\n",
    "\n",
    "\n",
    "## Lists`\n",
    "\n",
    "A list is used to collect a group of data objects of different sizes \n",
    "and types (e.g., one whole data frame and one vector can both be in a \n",
    "single list). It is simply an ordered collection of objects (that can \n",
    "be variables). The outputs of many statistical functions in R are lists \n",
    "(e.g. linear model fitting using ` lm()}), to return all relevant \n",
    "information in one output object. So you need to know how to unpack and \n",
    "manipulate lists. \n",
    "\n",
    "\n",
    "As a budding multilingual quantitative biologist, you should not be \n",
    "perturbed by the fact that a ` list`is a very different data structure in \n",
    "python vs R!\n",
    "\n",
    "\t\n",
    "\n",
    "> List1 <- list(species=c(\"Quercus robur\",\"Fraxinus excelsior\"), age=c(123, 84))\n",
    "> List1\n",
    "$species\n",
    "[1] \"Quercus robur\"   \"Fraxinus excelsior\"\n",
    "\n",
    "$age\n",
    "[1] 123 84\n",
    "\n",
    "\n",
    "You can access contents of a list item using number of the item instead of name:\n",
    "\n",
    "> List1[[1]]  \n",
    "[1] \"Quercus robur\"   \"Fraxinus excelsior\"\n",
    "> List1[[2]]\n",
    "[1] 123 84\n",
    "\n",
    "\n",
    "And you can access it using the name of the item in these two way:\n",
    "\n",
    "> List1[[\"age\"]]\n",
    "[1] 123 84\n",
    "> List1$age\n",
    "[1] 123 84\n",
    "\n",
    "\n",
    "You can build lists of lists too! Also, you have perhaps guessed by now \n",
    "that R dataframes are actually a kind of list.\n",
    "\n",
    "\n",
    "If dataframes are so nice, why use R matrices at all? The problem \n",
    "is that dataframes can be too slow when large numbers of mathematical \n",
    "calculations or operations need to be performed. In such cases, you \n",
    "will need to convert a dataframe to a matrix. But for statistical \n",
    "analyses and plotting, data frames will get the job done.    \n",
    "\n",
    "\n",
    "## Creating and manipulating data structures}\n",
    " \n",
    "## Creating Sequences}\n",
    "\n",
    "The ` :`operator creates vectors of sequential integers:\n",
    "\n",
    "\n",
    "> years <- 1990:2009\n",
    "> years\n",
    "[1] 1990 1991 1992 1993 1994 1995 1996 1997 1998 1999\n",
    "[11] 2000 2001 2002 2003 2004 2005 2006 2007 2008 2009\n",
    "\n",
    "> years <- 2009:1990 # or in reverse order \n",
    "> years\n",
    "[1] 2009 2008 2007 2006 2005 2004 2003 2002 2001 2000\n",
    "[11] 1999 1998 1997 1996 1995 1994 1993 1992 1991 1990\n",
    "\n",
    "\n",
    "For sequences of fractional numbers, you have to use ` seq()`:\n",
    "\n",
    "\n",
    "> seq(1, 10, 0.5)\n",
    " [1]  1.0  1.5  2.0  2.5  3.0  3.5  4.0  4.5  5.0  5.5  6.0  6.5  7.0  7.5  8.0\n",
    "[16]  8.5  9.0  9.5 10.0\n",
    "\n",
    "\n",
    "You can also ` seq(from=1,to=10, by=0.5)`OR ` seq(from=1, \n",
    "by=0.5, to=10)`with the same effect (try it) --- this explicit, \n",
    "\"argument matching\" approach is partly why R is so popular.\n",
    "\n",
    "## Acessing parts of data stuctures -- Indices and Indexing}\n",
    "Every element (entry) of a vector in R has an order: the first value, \n",
    "second, third, etc. To illustrate this, let's create a simple vector:\n",
    "\n",
    "> MyVar <- c( 'a' , 'b' , 'c' , 'd' , 'e' )\n",
    "\n",
    "Then, square brackets extract values based on their numerical order \n",
    "in the vector:\n",
    "\n",
    "> MyVar[1] # Show element in first position \n",
    "[1] \"a\"\n",
    "> MyVar[4]\n",
    "[1] \"d\" # Show element in fourth position \n",
    "\n",
    "The values in square brackets are called \"indices\" --- they \n",
    "give the index (position) of the required value. We can also select \n",
    "sets of values in different orders, or repeat values:\n",
    "\n",
    "> MyVar[c(3,2,1)] # reverse order\n",
    "[1] \"c\" \"b\" \"a\"\n",
    "MyVar[c(1,1,5,5)] # repeat indices\n",
    "[1] \"a\" \"a\" \"e\" \"e\"\n",
    "\n",
    "\n",
    "You can also manipulate data structures/objects by indexing:\n",
    "\n",
    "\n",
    "> v <- c(0, 1, 2, 3, 4) # Re-create the vector variable v\n",
    "> v[3] # access one element\n",
    "[1] 2\n",
    "> v[1:3] # access sequential elements\n",
    "[1] 0 1 2\n",
    "> v[-3] # remove elements\n",
    "[1] 0 1 3 4\n",
    "> v[c(1, 4)] # access non-sequential\n",
    "[1] 0 3\n",
    "\n",
    "\n",
    "For matrices, you need to use both row and column indices:\n",
    "\n",
    "> mat1 <- matrix(1:25, 5, 5, byrow=TRUE) #create a matrix\n",
    "> mat1\n",
    "\t [,1] [,2] [,3] [,4] [,5]\n",
    "[1,] 1 2 3 4 5\n",
    "[2,] 6 7 8 9   10\n",
    "[3,]   11   12   13   14   15\n",
    "[4,]   16   17   18   19   20\n",
    "[5,]   21   22   23   24   25\n",
    "> mat1[1,2]\n",
    "[1] 2\n",
    "> mat1[1,2:4]\n",
    "[1] 2 3 4\n",
    "> mat1[1:2,2:4]\n",
    "\t [,1] [,2] [,3]\n",
    "[1,] 2 3 4\n",
    "[2,] 7 8 9\n",
    "\n",
    "\n",
    "## Recycling}\n",
    "When vectors are of different lengths, R will recycle the shorter \n",
    "one to make a vector of the same length:\n",
    "\n",
    "a <- c(1,5) + 2\n",
    "x <- c(1,2); y <- c(5,3,9,2)\n",
    "x + y\n",
    "x + c(y,1)  ## somewhat strange!\n",
    "\n",
    "\n",
    "Recycling is convenient, but dangerous!\n",
    "\n",
    "## Basic vector-matrix operations}\n",
    "\n",
    "\n",
    "> v <- c(0, 1, 2, 3, 4)\n",
    "> v2 <- v*2 # multiply whole vector by 2\n",
    "> v2\n",
    "[1] 0 2 4 6 8\n",
    "> v * v2 # element-wise product\n",
    "[1]  0  2  8 18 32\n",
    "> t(v) # transpose the vector\n",
    "\t [,1] [,2] [,3] [,4] [,5]\n",
    "[1,] 0 1 2 3 4\n",
    "> v %*% t(v) # matrix/vector product\n",
    "\t [,1] [,2] [,3] [,4] [,5]\n",
    "[1,] 0 0 0 0 0\n",
    "[2,] 0 1 2 3 4\n",
    "[3,] 0 2 4 6 8\n",
    "[4,] 0 3 6 9   12\n",
    "[5,] 0 4 8   12   16\n",
    "> v3 <- 1:7 # assign using sequence\n",
    "> v3\n",
    "[1] 1 2 3 4 5 6 7\n",
    "> v4 <- c(v2, v3) # concatenate vectors\n",
    "> v4\n",
    " [1] 0 2 4 6 8 1 2 3 4 5 6 7\n",
    "\n",
    "\n",
    "## Strings and Pasting}\n",
    "It is important to know how to handle strings in R for two main reasons:\n",
    "\n",
    "* To deal with text data, such as names of experimental \n",
    "\ttreatments \n",
    "* To generate appropriate text labels and titles for figures \n",
    "\n",
    "\n",
    "Let's try creating and manipulating strings:\n",
    "\n",
    "> species.name <- \"Quercus robur\" #double quotes\n",
    "> species.name\n",
    "[1] \"Quercus robur\"\n",
    "> species.name <- 'Fraxinus excelsior' #single quotes \n",
    "> species.name\n",
    "[1] \"Fraxinus excelsior\"\n",
    "> paste(\"Quercus\", \"robur\")\n",
    "[1] \"Quercus robur\"\n",
    "> paste(\"Quercus\", \"robur\",sep = \"\") #Get rid of space\n",
    "\"Quercusrobur\"\n",
    "> paste(\"Quercus\", \"robur\",sep = \", \") #insert comma to separate\n",
    "\n",
    "As you can see above, both double and single quotes work, but I suggest \n",
    "that you use double quotes --- this will allow you to define strings \n",
    "that contain a single quotes, which is often necessary.\n",
    "\n",
    "And as is the case with so many R functions, pasting works on vectors:\n",
    "\n",
    "> paste('Year is:', 1990:2000)\n",
    " [1] \"Year is: 1990\" \"Year is: 1991\" \"Year is: 1992\" \"Year is: 1993\"\n",
    " [5] \"Year is: 1994\" \"Year is: 1995\" \"Year is: 1996\" \"Year is: 1997\"\n",
    " [9] \"Year is: 1998\" \"Year is: 1999\" \"Year is: 2000\"\n",
    "\n",
    "\n",
    "Note that this last example creates a vector of 11 strings as it is \n",
    "1990:2000 *inclusive}. \n",
    "\n",
    "\n",
    "*Data ``structures'' vs. ``objects'' in R}:\\\\\n",
    "\n",
    "You will often see the terms ``object'' and ``data structure'' thrown \n",
    "around this week and elsewhere. These two have a very distinct meaning \n",
    "in object-oriented programs (OOPs) like R. The main thing you \n",
    "need to keep in mind is that a data structrure is just a ``dumb'' \n",
    "container for data (e.g., a vector). An object, on the other hand can \n",
    "be a data structure, but also any other variables or functions in your R \n",
    "environment. R, being an OOP, needs to convert everything in the \n",
    "current environment to an object so that it knows what to do with each \n",
    "such entity --- each object type has its own set of rules for operations and \n",
    "manipulations that R uses when interpreting your commands. \n",
    "\n",
    "If all this sounds like gobbledygook, don't worry too much!   \n",
    "\t\n",
    "\n",
    "\n",
    "## Your analysis workflow}\n",
    "\n",
    "In using R for an analysis, you will likely use and create \n",
    "several files. As in the case of ` bash`and ` python`based \n",
    "projects, in R projects as well, you should keep your workflow \n",
    "well organized. For example, it is sensible to create a folder \n",
    "(directory) to keep all code files together. You can then set R \n",
    "to work from this directory, so that files are easy to find and run --- \n",
    "this will be your ``working directory'' (more on this below). Also, you \n",
    "don't want to mix code files with data and results files. So you should \n",
    "create separate directories for these as well. Thus, your typical ` \n",
    "R`analysis workflow will be:\n",
    "\n",
    "\n",
    "\n",
    "\t         minimum height=20pt, minimum width=4cm, text depth=0pt},\n",
    "     zn/.style={rectangle, rounded corners, fill=black!10, minimum width=5 cm, minimum height=5cm},\n",
    "     lb/.style={right=0pt, yshift=7pt, font=\\small, green!30!black},\n",
    "     every path/.style={-stealth, very thick, green!30!black}]\n",
    "\n",
    "\t\\node (R)    {\\includegraphics[width=2cm]{R_logo.png}};     \n",
    "\t\\node (inputs)  [zn, left= of R] {};\n",
    "\t\\node (outputs) [zn, right= of R] {};\n",
    "\t\\node (lab)  [lb, at=(inputs.north west), black] {Input files};\n",
    "\t\\node (lab)  [lb, at=(outputs.north west), black] {Output files (all to ` Results})};\n",
    "\n",
    "\t\\node (Rin)  [bx, above=1cm of inputs.center, anchor=center] {MyScript.R};\n",
    "\t\\node (data) [bx, below=1cm of inputs.center, anchor=center] {MyData.csv};\n",
    "\t\n",
    "\t\\node (lab)  [lb, at=(Rin.north west)] {R script file (from ` Code})};\n",
    "\t\\node (lab)  [lb, at=(data.north west)] {Text data file (from ` Data})`;\n",
    "\n",
    "\t\\node (Results) [bx, at=(outputs.center)]{MyTextResults.txt};\n",
    "\t\\node (image)   [bx, above=1.5cm of outputs.center, anchor=center]{MyGraphicResults.pdf};\n",
    "\t\\node (Rdata)   [bx, below=1.5cm of outputs.center, anchor=center]{MyRData.Rda};\n",
    "\n",
    "\t\\node (lab)  [lb, at=(Results.north west)] {Results output};\n",
    "\t\\node (lab)  [lb, at=(image.north west)] {Graphics file};\n",
    "\t\\node (lab)  [lb, at=(Rdata.north west)] {R data file};\n",
    "\n",
    "\t\\draw [bend left, shorten > = -8pt] (Rin.east) to (R);\n",
    "\t\\draw [bend right, shorten > = -8pt](data.east) to (R);\n",
    "\n",
    "\t\\draw [bend left, shorten < = -1pt]  (R) to (image.west);\n",
    "\t\\draw [bend right, shorten < = -1pt, stealth-stealth] (R) to (Rdata.west);\n",
    "\t\\draw (R) -- (Results.west);\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Some details on each kind of file:\n",
    "\n",
    "\n",
    "* [R script files] These are plain text files containing all the R \n",
    "code needed for an analysis. These should always be created with a \n",
    "simple text editor like Notepad (Windows), TextEdit (MacOS) or vscode \n",
    "(Linux) and saved with the extension ` *.R}. We will use RStudio in \n",
    "this class (more on this below). You should *never`use Word to \n",
    "save or edit these files as R can only read code from plain text files. \n",
    " \n",
    "* [Text data files] These are files of data in plain text format \n",
    "containing one or more columns of data (numbers, strings, or both). \n",
    "Although there are several format options, we will tyoucally be using \n",
    "csv files, where the entries are separated by commas. These are easy to \n",
    "create and export from Excel (if that's what you use...).\\footnote{If \n",
    "you are using a computer from elsewhere in the EU, Excel may use a \n",
    "comma ($\\pi=3,1416$) instead of a decimal point ($\\pi=3.1416$). In this \n",
    "case, *csv`files may use a semi-colon to separate columns and you \n",
    "can use the alternative function *read.csv2()`to read them into \n",
    "R.`\n",
    "\n",
    "* [Results output files] These are a plain text files contain your \n",
    "results, such the the summary of output of a regression or ANOVA  \n",
    "analysis. Typically, you will putput your results in a table format \n",
    "where the columns are separated by commas (csv) or tabs \n",
    "(tab-delimited) \n",
    "\n",
    "* [Graphics files] R can export graphics in a wide range of \n",
    "formats. This can be done automatically from R code and we will look at \n",
    "this later but you can also select a graphics window and click `File \n",
    "$\\triangleright$ Save as...'\n",
    "\n",
    "* [Rdata files] You can save any data loaded or created in R, \n",
    "including model outputs and other things, into a single` Rdata`\n",
    "file. These are not plain text and can only be read by R, but can hold \n",
    "all the data from an analysis in a single handy location. I never use \n",
    "these, but you can, if you want.  \n",
    "\n",
    "\n",
    "\n",
    "So let's build your R analysis project structure. \n",
    "\n",
    "Do the following:\n",
    "\n",
    "* Create a sensibly named directory (e.g., ` MyICRModule}, ` Week3}, etc in an \n",
    "\tappropriate location on your computer. If you are using a college Windows \n",
    "\tcomputer, you will need to create it in your ` H:`drive.  \n",
    "\t \n",
    "* Create subdirectories *within this directory`called ` \n",
    "\tCode}, ` Data}, and ` Results}\n",
    "\n",
    "\n",
    "\n",
    "\tAvoid including spaces in your file or directory names, as this will \n",
    "\toften create problems when you share your file or directory with \n",
    "\tsomebody else. Many software programs do not handle spaces in \n",
    "\tfile/directory names well. Use underscores instead of spaces. For \n",
    "\texample, instead of ` My IC R Module}, use ` \n",
    "\tMy\\_IC\\_R\\_Module}.  \n",
    "\t\n",
    "\n",
    "\n",
    "You can create directories using ` dir.create()`within R:\n",
    "\n",
    "\n",
    "> dir.create(\"MyICRModule\")\n",
    "> dir.create(\"MyICRModule/Code\")\n",
    "> dir.create(\"MyICRModule/Data\")\t\n",
    "> dir.create(\"MyICRModule/Results\")\t\n",
    "\n",
    "\n",
    "## The R Workspace and Working Directory}\n",
    "\n",
    "R has a ``workspace'' -- a current working environment that includes \n",
    "any user-defined data structures objects (vectors, matrices, data \n",
    "frames, lists) as well as other objects (e.g., functions). At the end \n",
    "of an R session, the user can save an image of the current workspace \n",
    "that is automatically reloaded the next time R is started. Your \n",
    "workspace is saved in your ``Working Directory'', which has to be set \n",
    "manually.\n",
    "\n",
    "So before we go any further, let's get sort out where your R ``Working \n",
    "Directory'' should be and how you should set it. R has a default \n",
    "location where it assumes your working directory is. \n",
    "\n",
    "in Windows, it is ` C:/Windows/system32`or similar.\n",
    "\n",
    "in Mac, it is ` /User/User Name`or similar.\n",
    "\n",
    "In UNIX/Linux, it is whichever directory you are in when you launch R.\n",
    "\n",
    "To see where your current working directory is, at the R command \n",
    "prompt, type:\n",
    "\n",
    "> getwd() \n",
    "\n",
    "This tells you what the current working directory (\"` wd}\") is. \n",
    "\n",
    "Now, set the working directory to be ` MyICRModule/Code}. For \n",
    "example, if you created ` MyICRModule`directly in your ` \n",
    "H:\\textbackslash }, the you would use:\n",
    "\n",
    "> setwd(\"H:/MyICRModule/Code\") \n",
    "> dir() #check what's in the current working directory\n",
    "\n",
    "On your own computer, you can also change R's default to a particular \n",
    "working directory where you would like to start (easily done in RStudio). \n",
    "\n",
    "In Linux, you can do this by editing the ` Rprofile.site`site with \n",
    "\\\\ ` sudo vscode /etc/R/Rprofile.site}. In that file, you would add \n",
    "your start-up parameters between the lines ` .First <- function() \n",
    "cat(\"\\textbackslash n   Welcome to R!\\textbackslash n\\textbackslash \n",
    "n\")`and ` .Last <- function() cat(\"\\textbackslash n   \n",
    "Goodbye!\\textbackslash n\\textbackslash n\")`--- between these lines, \n",
    "insert \\\\` setwd(\"/home/YourName/YourDirectoryPath\")}\n",
    "\n",
    "In Windows and Macs, you can find the ` Rprofile.site`file by \n",
    "searching for it. When I last checked for Windows, it used to be at ` \n",
    "C:\\textbackslash Program Files\\textbackslash R\\textbackslash \n",
    "etc\\textbackslash Rprofile.site }\n",
    "\n",
    "If you are using RStudio, you can change the default working directory \n",
    "by through the RStudio \"Options\" dialog.\n",
    "\n",
    "## Importing and Exporting Data}\n",
    "We are now ready to see how to import and export data in R, typically \n",
    "the first step of your analysis. The best option is to have your data \n",
    "in a ` c}omma ` s}eparated ` v}alue (` csv}) text file or \n",
    "in a tab separated text file. Then, \n",
    "you can use the function ` read.csv`(or ` read.table}) to import \n",
    "your data. Now, lets get some data into your ` Data`directory.\n",
    "\n",
    "\n",
    "* Go to the repository you downloaded from bitbucket and \n",
    "\tunzipped, and navigate to the ` Data`directory. \n",
    "* Copy the file ` trees.csv`into your own ` Data`\n",
    "\tdirectory.\n",
    "* Now, try the following:\n",
    "\n",
    "\n",
    "> MyData <- read.csv(\"../Data/trees.csv\")\n",
    "> ls() #Check that MyData has appeared \n",
    "> head(MyData) # Have a quick look at the data frame\n",
    "> str(MyData) # Have a quick look at the column types\n",
    "> MyData <- read.csv(\"../Data/trees.csv\", header = TRUE) # with headers\n",
    "> MyData <- read.table(\"../Data/trees.csv\", sep = ',', header = TRUE) #another way\n",
    "> head(MyData)\n",
    "> MyData <- read.csv(\"../Data/trees.csv\", skip = 5) # skip first 5 lines\n",
    "\n",
    "\n",
    "Note that the resulting ` MyData`in your workspace is a R \n",
    "dataframe. Also, note the UNIX-like paths using forward slashes (Windows uses \n",
    "back slashes). \n",
    "\n",
    "## Relative paths!`\n",
    "\n",
    "The $../$ in ` read.csv(\"../Data/trees.csv\")`above signifies a \n",
    "\"relative\" path. That is, you are asking R to load data that lies in a \n",
    "different directory (folder) relative your current location (in this \n",
    "case, you are in your ` Code`directory). In other, more dorky \n",
    "words, ` ../Data/trees.txt`points to a file named ` trees.txt`\n",
    "located in the \"parent\" of the current directory.\n",
    "\n",
    "*What is an absolute path?}--- one that specifies the whole path on \n",
    "your computer, say from ` C:/`\"upwards\".  \n",
    "\n",
    "Using relative paths in in your R scripts and code will make your code \n",
    "computer independent and your life better! The relative path way should \n",
    "always be the way you load data in your analyses scripts --- it will \n",
    "guarantee that your analysis works on every computer, not just your \n",
    "college computer. \n",
    "\n",
    "Also, *AVOID putting a ` setwd`command at the start of your R \n",
    "script}, as setting the working directory always requires an absolute \n",
    "directory path, which will differ across computers, platforms, and \n",
    "users. Let the end user sort out how to set the working directory. So \n",
    "to import data and export results, your script should *not`use \n",
    "absolute paths.  \n",
    "\n",
    "## Writing out to and saving files}\n",
    "\n",
    "You can also save your data frames using ` write.table`or ` \n",
    "write.csv}:\n",
    "\n",
    "\n",
    "> write.csv(MyData, \"../Results/MyData.csv\")\n",
    "> dir(\"../Results/\") # Check if it worked\n",
    "> write.table(MyData[1,], file = \"../Results/MyData.csv\",append=TRUE) # append\n",
    "> write.csv(MyData, \"../Results/MyData.csv\", row.names=TRUE) # write row names\n",
    "> write.table(MyData, \"../Results/MyData.csv\", col.names=FALSE) # ignore col names\n",
    "\n",
    "\n",
    "## Writing R code}\n",
    " \n",
    "Typing in commands interactively in the R console is good for starters, \n",
    "but you will want to switch to putting your sequence of commands into a \n",
    "script file, and then ask R to run those commands. \n",
    "\n",
    "\n",
    "* Open a new text file, call it ` basic\\_io.R}, and save it to \n",
    "\tyour ` Code`directory. \n",
    "* Write the above input-output commands in it: \n",
    "\n",
    "\\lstinputlisting[language=R]{Practicals/Code/basic_io.R}\n",
    "\n",
    "You will get a warning with ` write.table(MyData[1,], file = \n",
    "\"../Results/MyData.csv\",append=TRUE)`because R thinks it is silly that \n",
    "you are appending headers to a file that already has headers!\n",
    "\n",
    "\n",
    "* Place the cursor on the first line of code in the script file \n",
    "\tand run it by pressing the keyboard shortcut (PC: ctrl+R, Mac: \n",
    "\tcommand+enter, Linux: ctrl+enter if you are using ` vscode}).\n",
    "* Check after every line that you are getting the expected result.\n",
    "\n",
    "\n",
    "## Running R code}\n",
    "\n",
    "But even writing to a script file and running the code line-by-line or \n",
    "block-by-block is not your ultimate goal. What you would really like to \n",
    "do is to just run your full analysis and output all the results. There \n",
    "are two main approaches for running R script/code.\n",
    "\n",
    "\\subsubsection{Using ` source}}\n",
    "\n",
    "\n",
    "You can run all the contents of a ` *.R`script file from the ` \n",
    "R`command line by using ` source}. This causes R to accept code \n",
    "input from the named file and run it.\n",
    "\n",
    "* Try sourcing ` basic\\_io.R}: \n",
    "\n",
    "\n",
    "\n",
    "> source(\"basic_io.R\") # Assuming you are in Code directory!\n",
    "\n",
    "\n",
    "* If you get errors, fix them!\n",
    "\n",
    "\n",
    "\n",
    "Do not put a ` source()`command inside the script file you are \n",
    "sourcing, as it is then trying to run itself again and again and that's \n",
    "just cruel!\n",
    "\n",
    "\n",
    "Note that you will need to add the directory path to the file name \n",
    "(` basic\\_io.R`in the above example), if the script file is not in \n",
    "your working directory. For example, you will need \\\\` \n",
    "source(\"../Code/control.R\"`if your working directory is ` Data`and \n",
    "not ` Code}.\n",
    "\n",
    "\n",
    "The command ` source()`has a ` chdir`argument whose default \n",
    "value is FALSE. When set to TRUE, it will change the working directory \n",
    "to the directory of the file being sourced.  \n",
    "\n",
    "\n",
    "\\subsubsection{Using ` Rscript}}\n",
    "\n",
    "You can also run R script from the UNIX/Linux terminal by calling ` \n",
    "Rscript}. You can then easily automate the execution of your R scripts \n",
    "(e.g., by writing a bash script) and integrate R into a bigger \n",
    "computing pipeline/workflow by calling it through other tools or \n",
    "languages (such as ` Python}; see Chapter \\ref{chap:pythonII}). \n",
    "\n",
    "Let's try using ` Rscript`to run ` basic\\_io.R}: \n",
    "\n",
    "\n",
    "* Exit from the R console using ` ctrl+D}, or open a new \n",
    "\tbash terminal\n",
    "\t\n",
    "* ` cd`to the location of ` basic\\_io.R`(presumably, ` \n",
    "\tWeek3/code})\n",
    "\t\n",
    "* Then run the script using Rscript \n",
    "\n",
    "$ Rscript basic_io.R \n",
    "\n",
    "\t\n",
    "\n",
    "\n",
    "Also, please have a look at ` man Rscript`in a bash terminal.\n",
    "\n",
    "\n",
    "Thus, you have to be inside an R session to use ` source}, while \n",
    "you can only call ` Rscript`from the UNIX/Linux terminal.\n",
    "\n",
    "\n",
    "\\subsubsection{Running R in batch mode}\n",
    "\n",
    "In addition to ` Rscript}, there is another way to run you R script \n",
    "without opening the R console. In Mac or linux, you can do so by \n",
    "typing:\n",
    "\n",
    "\\noindent ` R CMD BATCH MyCode.R MyResults.Rout}\n",
    "\n",
    "This will create an ` MyResults.Rout`file containing all the output. \n",
    "On Microsoft Windows, it's more complicated --- change the path to ` \n",
    "R.exe`and output file as needed: \n",
    "\n",
    "\\noindent ` \"C:\\textbackslash{}Program \n",
    "Files\\textbackslash{}R\\textbackslash{}R-3.1.1\\textbackslash{}bin\\textbackslash{}R.exe\" \n",
    "CMD BATCH --vanilla --slave\\\\\n",
    " \"C:\\textbackslash{}PathToMyResults\\textbackslash{}Results\\textbackslash{}MyCode.R\"}\n",
    "  \n",
    "## Writing R Functions}\n",
    "Like any other programming language, R lets you write your own \n",
    "functions. A function is a block of re-useable code that takes an \n",
    "input, does something with it (or to it!), and returns the result. All \n",
    "the ``commands'' that you have been using, such as ` ls()}, ` \n",
    "mean()}, ` c()}, etc are basically functions. You will want to write \n",
    "your own function for every scenario where a particular, task or set \n",
    "of analysis steps need to be performed again and again.   \n",
    "\n",
    "The syntax for R functions is quite simple, with each function accepting \n",
    "``arguments'' and ``returning'' a value. \n",
    "\n",
    "Type the following into a script file called  ` boilerplate.R`and \n",
    "save it in your ` Code`directory.\n",
    " \n",
    "\\lstinputlisting[language=R]{Practicals/Code/boilerplate.R}\n",
    "\n",
    "Note the curly brackets -- these are necessary for R to know where the \n",
    "specification of the function starts and ends. Also, note the \n",
    "indentation. Not necessary (unlike Python), but recommended to make\n",
    "code more readable.\n",
    "\n",
    "Now source the script\n",
    "\n",
    "> source(\"boilerplate.R\")\t\n",
    "\n",
    "This will run the script, and also save your function ` MyFunction`\n",
    "as an object into your workspace (try ` ls()}, and you will see ` \n",
    "MyFunction`appear in the list of objects). \n",
    "\n",
    "Also try: \n",
    "\n",
    "> class(MyFunction)\n",
    "[1] \"function\"\t\n",
    "\n",
    "So, yes, ` MyFunction`is a ` function`object! \n",
    "\n",
    "Now let's write an script containing a more useful function:\n",
    "\n",
    "* In your text editor type the following in a file called ` \n",
    "\tTreeHeight.R}, and save it in your ` Code`directory:\n",
    "\n",
    "\n",
    "\\lstinputlisting[language=R]{Practicals/Code/TreeHeight.R}\n",
    "\n",
    "\n",
    "* Run ` TreeHeight.R`block by block and check what each line \n",
    "\tis doing.\n",
    "* Now run it using ` source`and/or ` Rscript}.\n",
    "* If you get errors, carefully read the error messages and fix \n",
    "\tthem!\n",
    "\n",
    "\n",
    "## Practicals}\n",
    "\n",
    "\n",
    "\t\n",
    "* Modify the script ` TreeHeight.R`so that it does the following:\n",
    "\t\n",
    "\n",
    "\t * Loads ` trees.csv`and calculates tree heights for all \n",
    "\t trees in the data. Note that the distances have been measured in \n",
    "\t meters. (Hint: use relative paths))\n",
    "\t \n",
    "* Creates a csv output file called ` TreeHts.csv`in ` \n",
    "Results`that contains the calculated tree heights along with the \n",
    "original data in the following format (only first two rows and \n",
    "headers shown):\t\n",
    "\t\n",
    "\t\"Species\",\"Distance.m\",\"Angle.degrees\",\"Tree.Height.m\"\n",
    "\t\"Populus tremula\",31.6658337740228,41.2826361937914,25.462680727681\n",
    "\t\"Quercus robur\",45.984992608428,44.5359166583512,46.094124200205\n",
    "\t\n",
    "\n",
    "* This script should work using either ` source`or ` \n",
    "\tRscript}\n",
    "\n",
    "\t\n",
    "\n",
    "* Make the ` TreeHeight.R`script more general so that it \n",
    "\tcould be used for other datasets, not just ` trees.csv}:\n",
    "\t\n",
    "\t\n",
    "\n",
    "* Write another R script called ` get\\_TreeHeight.R`\n",
    "\tthat takes a csv file name from the command line (e.g., ` \n",
    "\tget\\_TreeHeight.R Trees.csv}) and outputs the result to a file just \n",
    "\tlike ` TreeHeight.R`above, but this time includes the input file \n",
    "\tname in the output file name as ` InputFileName\\_treeheights.csv}. \n",
    "\tNote that you will have to strip the ` .csv`or whatever the \n",
    "\textension is from the filename, and also ` ../`etc., if you are \n",
    "\tusing relative paths \n",
    "\t\n",
    "\t(Hint: Command-line parameters are accessible within the R running \n",
    "\tenvironment via \\\\` commandArgs()`--- so help(commandArgs) might \n",
    "\tbe your starting point.)\n",
    "\n",
    "* Write a Unix shell script called ` run\\_get\\_TreeHeight.sh`\n",
    "\tthat tests \\\\ ` get\\_TreeHeight.R`--- include ` \n",
    "\ttrees.csv`as your example file. Note that \t` source`will not \n",
    "\twork in this case as it does not allow scripts with arguments to be \n",
    "\trun; you will have to use ` Rscript`instead. \n",
    "\n",
    "\t\n",
    "\n",
    "* {\\bf Extra credit}: If you have already started on ` python`\n",
    "\t(Chapter \\ref{chap:pythonI}), write a python version of ` \n",
    "\tget\\_TreeHeight.R`(call it ` get\\_TreeHeight.py}). Include a test \n",
    "\tof this script into ` run\\_get\\_TreeHeight.sh}.\t \n",
    "\n",
    "\n",
    "## Control statements}\n",
    "In R, you can write ` if}, ` then}, ` else`statements, and \n",
    "` for`and ` while`loops like any programming language.\n",
    "However, loops are slow in R, so use them sparingly. Such statements \n",
    "are useful to include in functions and scripts because you may only \n",
    "want to do certain calculations or otehr tasks, under certain \n",
    "conditions (e.g., ` if`the dataset is from a particular year, do \n",
    "something different). \n",
    "\n",
    "\n",
    "* Type the following in a script file called ` control.R`\n",
    "\t(save it in your ` Code`directory) \t\n",
    "* Run ` control.R`function block by block (not line by line!) \n",
    "\tand check what each line is doing.\n",
    "* Now run it using ` source`and ` Rscript}.\n",
    "* If you get errors, fix them (OK, I am going to stop saying this \n",
    "\thenceforth!).\n",
    "\n",
    "\n",
    "\\lstinputlisting[language=R]{Practicals/Code/control.R}\n",
    "\n",
    "## Useful R Functions}\n",
    "There are a number of very useful functions available by default (in \n",
    "the \"base packages\"). Here are some particularly useful ones:\n",
    "\n",
    "## Mathematical}\n",
    "\n",
    "\n",
    "` log(x)`& Natural logarithm\\\\\n",
    "` log10(x)`& Logarithm in base 10\\\\\n",
    "` exp(x)`& $e^x$\\\\\n",
    "` abs(x)`& Absolute value\\\\\n",
    "` floor(x)`& Largest integer $<x$\\\\\n",
    "` ceiling(x)`& Smallest integer $>x$\\\\\n",
    "` pi`& $\\pi$\\\\\n",
    "` sqrt(x)`& $\\sqrt{x}$\\\\\n",
    "` sin(x)`& Sinus function\\\\\n",
    "\n",
    "\n",
    "## Strings}\n",
    "\n",
    "\n",
    "` strsplit(x,';')`& Split the string at ';' \\\\\n",
    "` nchar(x)`& Number of characters\\\\\n",
    "` toupper(x)`& Set to upper case\\\\\n",
    "` tolower(x)`& Set to lower case\\\\\n",
    "` paste(x1,x2,sep=';')`& Join the strings using ';'\\\\\n",
    "\n",
    "\n",
    "## Statistical}\n",
    "\n",
    "\n",
    "\t` mean(x)`& Compute mean (of a vector or matrix)\\\\\n",
    "\t` sd(x)`& Standard deviation\\\\\n",
    "\t` var(x)`& Variance\\\\\n",
    "\t` median(x)`& Median\\\\\n",
    "\t` quantile(x,0.05)`& Compute the 0.05 quantile\\\\\n",
    "\t` range(x)`& Range of the data\\\\\n",
    "\t` min(x)`& Minimum\\\\\n",
    "\t` max(x)`& Maximum\\\\\n",
    "\t` sum(x)`& Sum all elements\\\\\n",
    "\n",
    "\n",
    "## Packages}\n",
    "The main strength of R is that users can easily build packages and \n",
    "share them through \\url{https://cran.r-project.org/}. There are packages to do \n",
    "most statistical and mathematical analysis you might conceive, so check \n",
    "them out before reinventing the wheel! Visit \\url{https://cran.r-project.org/`\n",
    "and go to packages to see a list and a brief description. \n",
    "\n",
    "In UNIX, you can use the terminal to install R packages:\n",
    "\n",
    "$ sudo apt-get install r-cran-ggplot2 r-cran-plyr r-cran-reshape2\n",
    "\n",
    "\n",
    "In Windows and Macs, you can install a package within R by using the \n",
    "` install.packages()`command. \n",
    "\n",
    "Let's install the following packages:\n",
    "\n",
    "> install.packages(c(\"ggplot2\",\"plyr\",\"reshape2\"))\n",
    "\n",
    "\n",
    "You can also use the RStudio GUI to install packages using your mouse \n",
    "and menu. \n",
    "\n",
    "\n",
    "In UNIX, you will have to launch a ` sudo R`session to get \n",
    "installation from within R using  ` install.packages()`to work \n",
    "properly. Otherwise, you be forced to install the package in a \n",
    "non-standard location on your drive that does not require ` sudo`\n",
    "privileges.\n",
    "\n",
    "\n",
    "\n",
    "## Practicals wrap-up}\n",
    "\n",
    "  \n",
    "\n",
    "* Review and make sure you can run all the commands, code \n",
    "\tfragments, and named scripts we have built till now and get the \n",
    "\texpected outputs.\n",
    "\n",
    "* Annotate/comment your code lines as much and as often as necessary \n",
    "\tusing \\#.\n",
    "\t\n",
    "* Keep all files organized in ` code}, ` data`and ` \n",
    "\tresults`directories.\n",
    "\t \n",
    "   \n",
    "\n",
    "\n",
    "\t\\it ` git add}, ` commit`and ` push`all your code and data \n",
    "\tfrom this chapter to your git repository by the *following \n",
    "\tWednesday 5PM}.\n",
    "\n",
    "\n",
    "## Readings}\n",
    "Check the readings under the R directory in the bitbucket master \n",
    "repository! Also, google \"R tutorial\", and plenty will pop up. Choose \n",
    "ones that seem the most intuitive to you.\n",
    "\n",
    "* The Use R! series (the yellow books) by Springer are really \n",
    "\tgood. In particular, consider: ``A Beginner's Guide to R'', ``R by \n",
    "\tExample'', ``Numerical Ecology With R'', ``ggplot2'' (coming up in \n",
    "\tChapter \\ref{chap:R_data}), ``A Primer of \n",
    "\tEcology with R'', ``Nonlinear Regression with R'', ``Analysis of \n",
    "\tPhylogenetics and Evolution with R''.\n",
    "* For more focus on dynamical models: Soetaert \\& Herman. 2009 \n",
    "\t``A practical guide to ecological modelling: using R as a \n",
    "\tsimulation platform''.\n",
    "* There are excellent websites besides cran. In particular, check out \n",
    "\t\\url{https://www.statmethods.net/`and  \\url{https://en.wikibooks.org/wiki/R_Programming}. \t\n",
    "* For those who are coming with Matlab experience:\n",
    "   \\url{http://www.math.umaine.edu/~hiebeler/comp/matlabR.html}\n",
    "  * Bolker, B. M.: Ecological Models and Data in R (eBook and \n",
    " Hardcover available).\n",
    "  * Beckerman, A. P. \\& Petchey, O. L. (2012) Getting started \n",
    " with R: an introduction for biologists. Oxford, Oxford University \n",
    " Press. \\\\ Very basic, good if you are really stuck at the outset. \n",
    "  * Crawley, R. (2013) The R book. 2nd edition. Chichester, Wiley. \\\\\n",
    " Excellent but enormous reference book, code and data available from\n",
    " \\url{www.bio.ic.ac.uk/research/mjcraw/therbook/index.htm}\n",
    "\t\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.4"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": false,
   "skip_h1_title": false,
   "title_cell": "Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": false,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
