{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!--NAVIGATION-->\n",
    "< [Multiple Explanatory Variables](16-MulExpl.ipynb) | [Main Contents](Index.ipynb) | [Model Simplification](18-ModelSimp.ipynb)>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Linear Models: Multiple variables and interactions  <span class=\"tocSkip\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Introduction\" data-toc-modified-id=\"Introduction-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Introduction</a></span><ul class=\"toc-item\"><li><span><a href=\"#Formulae-with-interactions-in-R\" data-toc-modified-id=\"Formulae-with-interactions-in-R-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Formulae with interactions in R</a></span></li><li><span><a href=\"#Model-1:-Mammalian-genome-size\" data-toc-modified-id=\"Model-1:-Mammalian-genome-size-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Model 1: Mammalian genome size</a></span></li><li><span><a href=\"#ANCOVA:-Body-Weight-in-Odonata\" data-toc-modified-id=\"ANCOVA:-Body-Weight-in-Odonata-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>ANCOVA: Body Weight in Odonata</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction \n",
    "\n",
    "Aims of this chapter[$^{[1]}$](#fn1):\n",
    "\n",
    "Creating more complex models, including ANCOVA\n",
    "\n",
    "Looking at interactions between variables\n",
    "\n",
    "Plotting predictions from models\n",
    "\n",
    "We will look at two models in this chapter:\n",
    "\n",
    "Model 1: Is mammalian genome size predicted by interactions between\n",
    "trophic level and whether species are ground dwelling?\n",
    "\n",
    "ANCOVA: Is body size in Odonata predicted by interactions between genome\n",
    "size and taxonomic suborder?\n",
    "\n",
    "So far, we have only looked at the independent effects of variables. For\n",
    "example, in the trophic level and ground dwelling model from [MulExpl](16-MulExpl.ipynb), we only looked for specific differences for being a\n",
    "omnivore *or* being ground dwelling, not for being\n",
    "specifically a *ground dwelling omnivore*. These\n",
    "independent effects of a variable are known as *main\n",
    "effects* and the effects of combinations of variables acting\n",
    "together are known as *interactions* — they describe how\n",
    "the variables *interact*.\n",
    "\n",
    "## Formulae with interactions in R\n",
    "\n",
    "We’ve already seen a number of different model formulae in R. They all\n",
    "use this syntax:\n",
    "\n",
    "`response variable ~ explanatory variable(s)`\n",
    "\n",
    "but we are now going to add two extra pieces of syntax:\n",
    "\n",
    "The `a:b` means the interaction between `a` and\n",
    "`b` — do combinations of these variables lead to different\n",
    "outcomes?\n",
    "\n",
    "This a shorthand for the model above. The means fit `a` and\n",
    "`b` as main effects and their interaction `a:b`.\n",
    "\n",
    "## Model 1: Mammalian genome size\n",
    "\n",
    "$\\star$ Make sure you have changed the working directory to `Code` in\n",
    "your stats coursework directory.\n",
    "\n",
    "Create a new blank script called ‘Interactions.R’ and add some\n",
    "introductory comments.\n",
    "\n",
    "Use `load(’mammals.Rdata’)` to load the data.\n",
    "\n",
    "If `mammals.Rdata` is missing, just import the data again\n",
    "using `read.csv(“../Data/MammalData.csv”)`. You will then\n",
    "have to add the log C Value column to the imported data frame again.\n",
    "\n",
    "Let’s refit the model from [MulExpl](16-MulExpl.ipynb), but including the\n",
    "interaction between trophic level and ground dwelling. We’ll immediately\n",
    "check the model is appropriate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> model <- lm(logCvalue ~ TrophicLevel * GroundDwelling, data= mammals)\n",
    "> par(mfrow=c(2,2), mar=c(3,3,1,1), mgp=c(2, 0.8,0))\n",
    "> plot(model)   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This gives:\n",
    "\n",
    "<a id=\"fig:mamMod\"></a>\n",
    "<figure>\n",
    "    <img src=\"./graphics/mamMod.svg\" alt=\"mamMod\" style=\"width:70%\">\n",
    "    <small> \n",
    "        <center>\n",
    "            <figcaption> \n",
    "           Figure 1\n",
    "            </figcaption>\n",
    "        </center>\n",
    "    </small>\n",
    "</figure>\n",
    "\n",
    "Now, we’ll examine the `anova` and `summary`\n",
    "outputs for the model:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> anova(model)\n",
    "\n",
    " Analysis of Variance Table\n",
    " \n",
    " Response: logCvalue\n",
    "                              Df Sum Sq Mean Sq F value  Pr(>F)    \n",
    " TrophicLevel                  2   0.81   0.407    8.06  0.0004 ***\n",
    " GroundDwelling                1   2.75   2.747   54.40 2.3e-12 ***\n",
    " TrophicLevel:GroundDwelling   2   0.43   0.216    4.27  0.0150 *  \n",
    " Residuals                   253  12.77   0.050                    \n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compared to the model from [MulExpl](16-MulExpl.ipynb), there is an extra\n",
    "line at the bottom. The top two are the same and show that trophic level\n",
    "and ground dwelling both have independent main effects. The extra line\n",
    "shows that there is also an interaction between the two. It doesn’t\n",
    "explain a huge amount of variation, about half as much as trophic level,\n",
    "but it is significant.\n",
    "\n",
    "Again, we can calculate the $r^2$ for the model: $$\\frac{0.81 + 2.75 + \n",
    "0.43}{0.81+2.75+0.43+12.77} = 0.238$$ The model from [MulExpl](16-MulExpl.ipynb) without the interaction had an $r^2 = 0.212$ — our new\n",
    "model explains 2.6% more of the variation in the data.\n",
    "\n",
    "The summary table is as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ">summary(model)\n",
    " \n",
    " Call:\n",
    " lm(formula = logCvalue ~ TrophicLevel * GroundDwelling, data = mammals)\n",
    " \n",
    " Residuals:\n",
    "    Min     1Q Median     3Q    Max \n",
    " -0.523 -0.171 -0.010  0.119  0.831 \n",
    " \n",
    " Coefficients:\n",
    "                                         Estimate Std. Error t value Pr(>|t|)    \n",
    " (Intercept)                               0.9589     0.0441   21.76  < 2e-16 ***\n",
    " TrophicLevelHerbivore                     0.0535     0.0554    0.97  0.33460    \n",
    " TrophicLevelOmnivore                      0.2328     0.0523    4.45  1.3e-05 ***\n",
    " GroundDwellingYes                         0.2549     0.0651    3.92  0.00012 ***\n",
    " TrophicLevelHerbivore:GroundDwellingYes   0.0303     0.0786    0.39  0.69979    \n",
    " TrophicLevelOmnivore:GroundDwellingYes   -0.1476     0.0793   -1.86  0.06384 .  \n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 \n",
    " \n",
    " Residual standard error: 0.225 on 253 degrees of freedom\n",
    "   (120 observations deleted due to missingness)\n",
    " Multiple R-squared: 0.238, Adjusted R-squared: 0.223 \n",
    " F-statistic: 15.8 on 5 and 253 DF,  p-value: 1.5e-13 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The lines in this are:\n",
    "\n",
    "1. The reference level (intercept) for non ground dwelling carnivores. (The reference level is decided just by the alphabetic order of the levels)\n",
    "2. Two differences for being in different trophic levels.\n",
    "3. One difference for being ground dwelling\n",
    "4. Two new differences that give specific differences for ground dwelling herbivores and omnivores.\n",
    "\n",
    "The first four lines, as in the model from [ANOVA](15-anova.ipynb), which\n",
    "would allow us to find the predicted values for each group *if the\n",
    "size of the differences did not vary between levels because of the\n",
    "interactions*. That is, this part of the model only includes a\n",
    "single difference ground and non-ground species, which has to be the\n",
    "same for each trophic group because it ignores interactions between\n",
    "trophic level and ground / non-ground identity of each species. The last\n",
    "two lines then give the estimated coefficients associated with the\n",
    "interaction terms, and allow cause the size of differences to vary\n",
    "between levels because of the further effects of interactions.\n",
    "\n",
    "The table below show how these combine to give the predictions for each\n",
    "group combination, with those two new lines show in red:\n",
    "\n",
    "$$\\begin{array}{|r|r|r|}\n",
    "\\hline\n",
    " & \\textrm{Not ground} &  \\textrm{Ground} \\\\\n",
    "\\hline\n",
    "\\textrm{Carnivore} & 0.96 = 0.96 &  0.96+0.25=1.21 \\\\\n",
    "\\textrm{Herbivore} & 0.96 + 0.05 = 1.01 & 0.96+0.05+0.25{\\color{red}+0.03}=1.29\\\\\n",
    "\\textrm{Omnivore} & 0.96 + 0.23 = 1.19 & 0.96+0.23+0.25{\\color{red}-0.15}=1.29\\\\\n",
    "\\hline\n",
    "\\end{array}$$\n",
    "\n",
    "So why are there two new coefficients? For interactions between two\n",
    "factors, there are always $(n-1)\\times(m-1)$ new coefficients, where $n$\n",
    "and $m$ are the number of levels in the two factors (Ground dwelling or\n",
    "not: 2 levels and trophic level: 3 levels, in our current example). So\n",
    "in this model, $(3-1) \\times (2-1) =2$. It is easier to understand why\n",
    "graphically: the prediction for the white boxes below can be found by\n",
    "adding the main effects together but for the grey boxes we need to find\n",
    "specific differences and so there are $(n-1)\\times(m-1)$ interaction\n",
    "coefficients to add.\n",
    "\n",
    "<a id=\"fig:interactionsdiag\"></a>\n",
    "<figure>\n",
    "    <img src=\"./graphics/interactionsdiag.png\" alt=\"interactionsdiag\" style=\"width:50%\">\n",
    "    <small> \n",
    "        <center>\n",
    "            <figcaption> \n",
    "           Figure 2\n",
    "            </figcaption>\n",
    "        </center>\n",
    "    </small>\n",
    "</figure>\n",
    "If we put this together, what is the model telling us?\n",
    "\n",
    "Herbivores have the same genome sizes as carnivores, but omnivores have\n",
    "larger genomes.\n",
    "\n",
    "Ground dwelling mammals have larger genomes.\n",
    "\n",
    "These two findings suggest that ground dwelling omnivores should have\n",
    "extra big genomes. However, the interaction shows they are smaller than\n",
    "expected and are, in fact, similar to ground dwelling herbivores.\n",
    "\n",
    "Note that although the interaction term in the `anova` output\n",
    "is significant, neither of the two coefficients in the\n",
    "`summary` has a $p<0.05$. There are two weak differences (one\n",
    "very weak, one nearly significant) that together explain significant\n",
    "variance in the data.\n",
    "\n",
    "$\\star$ Copy the code above into your script and run the model.\n",
    "\n",
    "Make sure you understand the output!\n",
    "\n",
    "Just to make sure the sums above are correct, we’ll use the same code as\n",
    "in [MulExpl](16-MulExpl.ipynb) to get R to calculate predictions for us:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# a data frame of combinations of variables\n",
    "> gd <- rep(levels(mammals$GroundDwelling), times = 3)\n",
    "> print(gd)\n",
    " [1] \"No\"  \"Yes\" \"No\"  \"Yes\" \"No\"  \"Yes\"\n",
    "\n",
    "> tl <- rep(levels(mammals$TrophicLevel), each = 2)\n",
    "> print(tl)\n",
    "\n",
    " [1] \"Carnivore\" \"Carnivore\" \"Herbivore\" \"Herbivore\" \"Omnivore\"  \"Omnivore\" \n",
    "\n",
    "# New data frame\n",
    "> predVals <- data.frame(GroundDwelling = gd, TrophicLevel = tl)\n",
    "\n",
    "# predict using the new data frame\n",
    "> predVals$predict <- predict(model, newdata = predVals)\n",
    "> print(predVals)\n",
    " \n",
    "    GroundDwelling TrophicLevel predict\n",
    " 1             No    Carnivore  0.9589\n",
    " 2            Yes    Carnivore  1.2138\n",
    " 3             No    Herbivore  1.0125\n",
    " 4            Yes    Herbivore  1.2977\n",
    " 5             No     Omnivore  1.1918\n",
    " 6            Yes     Omnivore  1.2990"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$\\star$ Run these predictions in your script.\n",
    "\n",
    "If we plot these data points onto the barplot from [MulExpl](16-MulExpl.ipynb), they now lie exactly on the mean values, because we’ve\n",
    "allowed for interactions. The triangle on this plot shows the\n",
    "predictions for ground dwelling omnivores from the main effects\n",
    "($0.96 + 0.23  + 0.25 = 1.44$), the interaction of $-0.15$ pushes the\n",
    "prediction back down.\n",
    "\n",
    "<a id=\"fig:predPlot\"></a>\n",
    "<figure>\n",
    "    <img src=\"./graphics/predPlot.svg\" alt=\"predPlot\" style=\"width:80%\">\n",
    "</figure>\n",
    "\n",
    "\n",
    "## ANCOVA: Body Weight in Odonata\n",
    "\n",
    "We’ll go all the way back to the regression analyses from [regress](14-regress.ipynb). Remember that we fitted two separate regression\n",
    "lines to the data for damselflies and dragonflies. We’ll now use an\n",
    "interaction to fit these in a single model. This kind of linear model —\n",
    "with a mixture of continuous variables and factors — is often called an\n",
    "*analysis of covariance*, or ANCOVA. That is, ANCOVA is a\n",
    "type of linear model that blends ANOVA and regression. ANCOVA evaluates\n",
    "whether population means of a dependent variable are equal across levels\n",
    "of a categorical independent variable, while statistically controlling\n",
    "for the effects of other continuous variables that are not of primary\n",
    "interest, known as covariates.\n",
    "\n",
    "*That is, this is still a linear model, but with one categorical\n",
    "and one or more continuous predictors*.\n",
    "\n",
    "$\\star$ Load the data:\n",
    "\n",
    "`odonata <- read.csv(’../Data/GenomeSize.csv’)`\n",
    "\n",
    "Create two new variables in the `odonata` data set called\n",
    "`logGS` and `logBW` containing log genome size and\n",
    "log body weight.\n",
    "\n",
    "The models we fitted before looked like this:\n",
    "\n",
    "<a id=\"fig:dragonData\"></a>\n",
    "<figure>\n",
    "    <img src=\"./graphics/dragonData.svg\" alt=\"dragonData\" style=\"width:70%\">\n",
    "    <small> \n",
    "        <center>\n",
    "            <figcaption> \n",
    "           Figure 3\n",
    "            </figcaption>\n",
    "        </center>\n",
    "    </small>\n",
    "</figure>\n",
    "\n",
    "We can now fit the model of body weight as a function of both genome\n",
    "size and suborder:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> odonModel <- lm(logBW ~ logGS * Suborder, data = odonata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Again, we’ll look at the <span>anova</span> table first:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> anova(odonModel)\n",
    " \n",
    " Analysis of Variance Table\n",
    " \n",
    " Response: logBW\n",
    "                Df Sum Sq Mean Sq F value  Pr(>F)    \n",
    " logGS           1    1.1     1.1    2.71     0.1    \n",
    " Suborder        1  112.0   112.0  265.13 < 2e-16 ***\n",
    " logGS:Suborder  1    9.1     9.1   21.65 1.1e-05 ***\n",
    " Residuals      94   39.7     0.4                    \n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Interpreting this gives the following:\n",
    "\n",
    "There is no significant main effect of log genome size. The\n",
    "*main* effect is the important thing here — genome size is\n",
    "hugely important but does very different things for the two different\n",
    "suborders. If we ignored `Suborder`, there isn’t an overall\n",
    "relationship: the average of those two lines is pretty much flat.\n",
    "\n",
    "There is a very strong main effect of Suborder: the mean body weight in\n",
    "the two groups are very different.\n",
    "\n",
    "There is a strong interaction between suborder and genome size. This is\n",
    "an interaction between a factor and a continuous variable and shows that\n",
    "the *slopes* are different for the different factor levels.\n",
    "\n",
    "The summary table looks like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> summary(odonModel)\n",
    " \n",
    " Call:\n",
    " lm(formula = logBW ~ logGS * Suborder, data = odonata)\n",
    " \n",
    " Residuals:\n",
    "     Min      1Q  Median      3Q     Max \n",
    " -1.3243 -0.3225  0.0073  0.3962  1.4976 \n",
    " \n",
    " Coefficients:\n",
    "                         Estimate Std. Error t value Pr(>|t|)    \n",
    " (Intercept)              -2.3995     0.0848  -28.31  < 2e-16 ***\n",
    " logGS                     1.0052     0.2237    4.49  2.0e-05 ***\n",
    " SuborderZygoptera        -2.2489     0.1354  -16.61  < 2e-16 ***\n",
    " logGS:SuborderZygoptera  -2.1492     0.4619   -4.65  1.1e-05 ***\n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 \n",
    " \n",
    " Residual standard error: 0.65 on 94 degrees of freedom\n",
    "   (2 observations deleted due to missingness)\n",
    " Multiple R-squared: 0.755, Adjusted R-squared: 0.747 \n",
    " F-statistic: 96.5 on 3 and 94 DF,  p-value: <2e-16 \n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first thing to note is that the $r^2$ value is really high. The\n",
    "model explains three quarters (0.752) of the variation in the data.\n",
    "Next, there are four coefficients:\n",
    "\n",
    "The intercept is for the first level of `Suborder`, which is\n",
    "Anisoptera (dragonflies).\n",
    "\n",
    "The next line, for `log genome size`, is the slope for Anisoptera.\n",
    "\n",
    "We then have a coefficient for the second level of `\n",
    "Suborder`, which is Zygoptera (damselflies). As with the first\n",
    "model, this difference in factor levels is a difference in mean values\n",
    "and shows the difference in the intercept for Zygoptera.\n",
    "\n",
    "The last line is the interaction between `Suborder` and\n",
    "`logGS`. This shows how the slope for Zygoptera differs from\n",
    "the slope for Anisoptera.\n",
    "\n",
    "How do these hang together to give the two lines shown in the model? We\n",
    "can calculate these by hand: $$\\begin{aligned}\n",
    "    \\textrm{Body Weight} &= -2.40 + 1.01 \\times \\textrm{logGS} & \\textrm{[Anisoptera]}\\\\\n",
    "    \\textrm{Body Weight} &= (-2.40 -2.25) + (1.01 - 2.15) \\times \\textrm{logGS} & \\textrm{[Zygoptera]}\\\\\n",
    "                         &= -4.65 - 1.14 \\times \\textrm{logGS} \\\\\\end{aligned}$$\n",
    "\n",
    "$\\star$ Add the code into your script and check that you understand the outputs.\n",
    "\n",
    "We’ll use the `predict` function to get the predicted values\n",
    "from the model and add lines to the plot above.\n",
    "\n",
    "First, we’ll create a set of numbers spanning the range of genome size:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the range of the data\n",
    "> rng <- range(odonata$logGS)\n",
    "#get a sequence from the min to the max with 100 equally spaced values\n",
    "> LogGSForFitting <- seq(rng[1], rng[2], length = 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Have a look at these numbers:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(LogGSForFitting)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now use the model to predict the values of body weight at each of\n",
    "those points for each of the two suborders. We’ve added `se.fit=TRUE` to the function to get the standard error around the\n",
    "regression lines. Note that we are now using"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get a data frame of new data for the order\n",
    "> ZygoVals <- data.frame(logGS = LogGSForFitting, Suborder = \"Zygoptera\")\n",
    "\n",
    "#get the predictions and standard error\n",
    "> ZygoPred <- predict(odonModel, newdata = ZygoVals, se.fit = TRUE)\n",
    "\n",
    "#repeat for anisoptera\n",
    "AnisoVals <- data.frame(logGS = LogGSForFitting, Suborder = \"Anisoptera\")\n",
    "\n",
    "AnisoPred <- predict(odonModel, newdata = AnisoVals, se.fit = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Both `AnisoPred` and `ZygoPred` contain predicted values (called `fit`) and standard error values (called\n",
    "`se.fit`) for each of the values in our generated values in\n",
    "`LogGSForFitting` for each of the two suborders.\n",
    "\n",
    "We can add the predictions onto a plot like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the scatterplot of the data\n",
    "> plot(logBW ~ logGS, data = odonata, col = Suborder)\n",
    "# add the predicted lines\n",
    "> lines(AnisoPred$fit ~ LogGSForFitting, col = \"black\")\n",
    "> lines(AnisoPred$fit + AnisoPred$se.fit ~ LogGSForFitting, col = \"black\", lty = 2)\n",
    "> lines(AnisoPred$fit - AnisoPred$se.fit ~ LogGSForFitting, col = \"black\", lty = 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$\\star$ Copy the prediction code into your script and run the plot above.\n",
    "\n",
    "Copy and modify the last three lines to add the lines for the Zygoptera. Your final plot should look like this.\n",
    "\n",
    "<a id=\"fig:odonPlot\"></a>\n",
    "<figure>\n",
    "    <img src=\"./graphics/odonPlot.svg\" alt=\"odonPlot\" style=\"width:70%\">\n",
    "    <small> \n",
    "        <center>\n",
    "            <figcaption> \n",
    "           Figure 4\n",
    "            </figcaption>\n",
    "        </center>\n",
    "    </small>\n",
    "</figure>\n",
    "\n",
    "---\n",
    "\n",
    "<a id=\"fn1\"></a>\n",
    "[1]: Here you work with the script file `MulExplInter.R`"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.4"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": false,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": false,
   "skip_h1_title": false,
   "title_cell": "Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
