{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Experimental design and Data exploration <a class=\"tocSkip\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div id=\"toc\"> </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction and preliminaries\n",
    "\n",
    "Ideally, you would like to design experiments (manipulations and/orobservations) that are appropriate for the question you want to answer. However, you still need to explore you data to determine what kind of \n",
    "statistical tests would be appropriate because: (a) Your experiments or observations may not go as planned, and (b) You might have somebody else's data to analyse (very likely in your UG projects). Building on the previous chapter, in this chapter we you learn how to use R to explore your data and determine appropriate statistical tests. By the \n",
    "time you have worked through this chapter, you should be able to \n",
    "\n",
    "* Provided sufficient information is available, be able to judge \n",
    " whether the sampling design used to generate a particular dataset was \n",
    " appropriate\n",
    "* Determine if your sample sizes are adequate, especially for a \n",
    " specific statistical test qualitatively (Y1) or quantitatively using \n",
    " power analysis (Y2). \n",
    "* Calculate basic statistical measures on your data to determine \n",
    " its properties\n",
    "\n",
    "We are going to start off in Year 1 with the simplest of scenarios for  statistical testing --- that you want to determine whether a sample, or a pair of samples meet some expectation (hypothesis) or not. \n",
    "\n",
    "First, some conceptual preliminaries.\n",
    "\n",
    "### Some statistical parlance\n",
    "The following terms are important for you to get familiar with:\n",
    "\n",
    "   \n",
    "* **(Statistical) Population**: A statistical population is a *complete set* of items that share at least one **  attribute* of interest. This attribute of interest is the target of your statistical analysis. For example, if we are interested in studying the weight of year-old cod in the Oceans, the population  consists of {\\it all} year-old cod, but more specifically, the weight measurements of all the individuals of the cod population is what we want to analyse. \n",
    "    \n",
    "* **(Statistical) Distribution** A statistical distribution is a mathematical description (expressed as a mathematical equation) of the properties of a population of interest. Theoreticians have come up with a bunch of distributions (e.g., Gaussian or Normal, Poisson, Binomial, etc.) that are appropriate for different kinds of data. Figuring out which distribution best describes a population of interest is one of the first steps in a statistical \n",
    "  analysis. In reality, of course, even collecting and measuring all \n",
    "  the individuals of a population may not be sufficient to \n",
    "  characterize its statistical properties --- imagine the situation \n",
    "  where the cod population has declined to a few hundred individuals \n",
    "  (not an impossibility in the future!).\n",
    "    \n",
    "  *[(Data or Population) Sample] A data {\\it sample} is a set of \n",
    "  measurements of the attribute of interest collected from a {\\it \n",
    "  statistical population} by a defined procedure ({\\it sampling \n",
    "  methodology}). In the cod example above, this would be the weight \n",
    "  of every individual of a {\\it subset} of the year-old Cod \n",
    "  population.\n",
    "\n",
    "  *[(Statistical) Parameter]  A statistical parameter is a \n",
    "  measure of some attribute of the {\\it theoretical} statistical \n",
    "  distribution that is support to represent your population. An \n",
    "  example would be the average weight of yearling cod. In \n",
    "  practice, this is not measurable because the population is much too \n",
    "  large or incompletely inaccessible/invisible --- imagine measuring \n",
    "  the weight of every year-old cod individual in the oceans!   \n",
    "\n",
    "  *[Statistic] A statistic (singular) is an {\\it estimate} of a \n",
    "  statistical parameter of the population of interest, obtained by \n",
    "  calculating the {\\it measure} for a {\\it sample} (e.g.,  the \n",
    "  average or mean weight of individuals in a sample of one-year old \n",
    "  cod). This is also your {\\it descriptive statistic}. Therefore, a \n",
    "  {\\it Statistic} is to a {\\it Statistical Parameter} what a {\\it \n",
    "  Sample} is to the {\\it Statistical Population}. For example, the \n",
    "  average of a sample of cod weights is a statistic that {\\it \n",
    "  estimates} the theoretical ``real'' average of the weights of the \n",
    "  entire one-year Cod population, which is its statistical parameter.\n",
    "    \n",
    "  *[Hypothesis] A Hypothesis is an (hopefully) informed {\\it \n",
    "  postulate} about an attribute of your population of interest. For \n",
    "  example, you may hypothesize that the one-year old cod population's \n",
    "  mean weight has declined over the last two decades (it has!). You \n",
    "  will want to confront your main hypothesis with a ``null'' \n",
    "  Hypothesis, to minimize the risk of making a ``type I'' error. A \n",
    "  type I error is the probability of accepting an alternative (or \n",
    "  main) hypothesis (and rejecting the null hypothesis) that is not \n",
    "  really valid (e.g., the yearling cods have actually not declined in \n",
    "  weight). This is a big NO NO from a scientific and philosophical \n",
    "  standpoint. The rate or probability of the type I error is denoted \n",
    "  by the Greek letter $\\alpha$, and equals the {\\it significance \n",
    "  level} of a statistical test.\n",
    "\n",
    "\\end{description}\n",
    "\n",
    "## Descriptive Statistics}\n",
    "\n",
    "The fundamental statistics that describe a sample (or a population) \n",
    "are, firstly, the mean, or average value of a sample, typically denoted \n",
    "by a $\\bar{x}$: \n",
    "\n",
    "\\begin{equation}\n",
    " \\bar{x} =  \\frac{\\sum\\limits_{i=1}^n x_i}{n} = \\frac{x_{1} + x_{2} + \\dots +x_{n}}{n}\n",
    "\\end{equation}\n",
    "\n",
    "That is, it is the sum of all the values in a sample divided by the \n",
    "number, $n$, of items in the sample. Thus it is a measure of the {\\it \n",
    "central-tendency} of the sample and population.\n",
    "\n",
    "Second, the standard deviation ($s$) is:\n",
    "\n",
    "\\begin{equation}\n",
    " s =  \\sqrt{\\frac{\\sum\\limits_{i=1}^n (\\bar{x} - x_{i})^2}{n-1}} = \n",
    " \\sqrt{\\frac{(\\bar{x} - x_{1})^{2} + (\\bar{x} - x_{2})^2 + \\dots + (\\bar{x} \n",
    " - x_{n})^{2}}{n-1}}\n",
    "\\end{equation}\n",
    "\n",
    "That is, the the square root of the sum of squares (\"SS\") of the \n",
    "differences between each item in the sample and the mean, divided by \n",
    "the {\\it degrees of freedom, \"df\"} remaining in the data set ($ \n",
    "n-1$). df is the sample size, $n$, minus the number of statistical \n",
    "parameters estimated from the data set. This is to reduce the {\\it \n",
    "bias} in your {\\it estimate} of the statistic, as you are calculating \n",
    "it from the sample, and not the whole theoretical population.\n",
    "\n",
    "Thus, the formula for $s$ above has $n-1$ in its denominator because, \n",
    "to to work out the standard deviation, you must have already estimated \n",
    "the mean ($\\bar{x}$) from the same data set. This removes 1 degree of \n",
    "freedom. Also, note that the sample variance, $s^2$ is the square of \n",
    "standard deviation. Or, in other words, the standard deviation is the \n",
    "square-root of the variance! \n",
    "\n",
    "The above two statistics (mean and sd) are particularly meaningful when \n",
    "the sample and population have a symmetric distribution (e.g., normal \n",
    "or gaussian). When the distribution is not symmetric (that is, it is \n",
    "{\\it skewed}), another statistic, the {\\it median} becomes important. \n",
    "This is the middle value in the ordered set of data, that is exactly \n",
    "50\\% of the data live below and 50\\% lie above the median. In skewed \n",
    "distributions, the median is a better measure of the \n",
    "{\\it central-tendency} of the sample and population. \n",
    "\n",
    "Other descriptive statistics you should keep in mind are the range \n",
    "(difference between the largest and smallest values), and the quartiles \n",
    "(values lying in the data divided into the intervals $[{1\\over4}, {1\\over \n",
    "2},{3\\over 4}, 1]$ or at 1\\% intervals (percentiles). Box-plots, which \n",
    "you have seen, represent a number of these statistics in one figure.   \n",
    "\n",
    "\\subsection{Descriptive statistic functions in R}\n",
    "\n",
    "\\begin{tabular}{p{3.5cm} p{9.5cm}}\n",
    " {\\tt mean(x)} & Compute mean (of a vector or matrix)\\\\\n",
    " {\\tt sd(x)} & Standard deviation\\\\\n",
    " {\\tt var(x)} & Variance\\\\\n",
    " {\\tt median(x)} & Median\\\\\n",
    " {\\tt quantile(x,0.05)} & Compute the 0.05 quantile\\\\\n",
    " {\\tt range(x)} & Range of the data\\\\\n",
    " {\\tt min(x)} & Minimum\\\\\n",
    " {\\tt max(x)} & Maximum\\\\\n",
    " {\\tt sum(x)} & Sum all elements\\\\\n",
    "\\end{tabular}\\\\\n",
    "\n",
    "## Data types and distributions}\n",
    "You will typically encounter or sample the following main types of \n",
    "data:\n",
    "\n",
    "\\begin{description}\n",
    "\n",
    "  *[Continuous numeric] This is the {\\tt  numeric} or {\\tt real} \n",
    "  data type in R, and as far as you are concerned, these data \n",
    "  typically will be made up of (mathematically) real numbers such as \n",
    "  human height or weight. These may be unbounded (any value between \n",
    "  negative infinity to positive infinity), or bounded (e.g., between  \n",
    "  or zero and some upper positive number) like human weight.\n",
    "    \n",
    "    *[Discrete numeric] This is the {\\tt  integer} data type in R, \n",
    "    and consist of (mathematically) integer (whole) numbers such as \n",
    "    counts of individuals in a population, e.g., The number of bacteria \n",
    "    in a ml of pond water. \n",
    "    \n",
    "  *[Percentage (proportion)] Percentage data is a particular kind \n",
    "  of numeric data that is strictly bounded between 0 and 100. The \n",
    "  fact that you can never get samples of percentages that exceed \n",
    "  these bounds makes such data tricky to analyse.\n",
    "    \n",
    "    * [Categorical] These are typically stored as the {\\tt  \n",
    "   character} data type in R. Categorical data are discrete, \n",
    "   typically expressed as a one of a fixed number of {\\it levels} of a \n",
    "   {\\it factor}. For example, the the factor \"Type.of.feeding.interaction\" \n",
    "   from the predator-prey dataset you have seen previously had five \n",
    "   levels: \"insectivorous\", \"piscivorous\", \"planktivorous\", \n",
    "   \"predacious\", and \"predacious/piscivorous\".\n",
    "      \n",
    "     * [Binary (presence/absence) data] A special type of \n",
    "     categorical data are binary, where only two categories or states \n",
    "     are possible: (1, 0) (or \"present\", \"absent\"), e.g., a disease \n",
    "     symptom. These may be stored as {\\tt  integer} or {\\tt  \n",
    "     character} in R.\n",
    "\\end{description} \n",
    "\n",
    "While designing experiments or exploring data obtained by somebody \n",
    "else, you need to keep in mind that each type will typically be best \n",
    "represented a particular {\\it statistical distribution}. For example, \n",
    "continuous numeric data are {\\it often} normally distributed. On the \n",
    "other hand, count data are likely to be distributed according to the \n",
    "Poisson distribution.\n",
    "\n",
    "If you are lucky, you will mostly have to deal with data that are \n",
    "continuous or discrete numeric, which are the most straightforward to \n",
    "analyse using Linear models (more on that in subsequent chapters). \n",
    "However, some of the most interesting and important problems in biology \n",
    "involve proportion (percentage), categorical and binary data (e.g., \n",
    "Presence or absence of a disease symptom). \n",
    "\n",
    "For example, think about what type of data, and what type of \n",
    "distribution, a sample of the following is likely to be:\n",
    "\\begin{compactitem}\n",
    "    * Wavelength of light\n",
    "    * Temperature\n",
    "    * Egg clutch size\n",
    "    * Rate of a reaction\n",
    "    * Eye-colour\n",
    "    * Score in Scrabble\n",
    "    * UG Degree class\n",
    "    * Ground-cover of grass in a quadrat\n",
    "    * Winning side in chess\n",
    "\\end{compactitem}\n",
    "\n",
    "\\subsection{Sampling from distributions in R}\n",
    "\n",
    "You can generate samples form many distributions in R (and handy thing \n",
    "to know). In particular, the following are important: \n",
    " \n",
    "\\begin{tabular}{p{5.4cm} p{9cm}}\n",
    "    {\\tt rnorm(10, m=0, sd=1)} & Draw 10 normal random numbers with mean 0 and s.d. 1\\\\\n",
    "    {\\tt dnorm(x, m=0, sd=1)} & Density function\\\\\n",
    "    {\\tt qnorm(x, m=0, sd=1)} & Cumulative density function\\\\\n",
    "    {\\tt runif(20, min=0, max=2)} & Twenty random numbers from uniform\n",
    "        [0,2]\\\\\n",
    "    {\\tt rpois(20, lambda=10)} & Twenty random numbers from\n",
    "        Poisson($\\lambda$)\\\\\n",
    "\\end{tabular}\\\\\n",
    "\n",
    "## Two basic rules of experimental design and sampling}\n",
    "\n",
    "In general, while designing experiments, and sampling from a {\\it \n",
    "population}, there are two key (and simple) rules:\n",
    "\\begin{enumerate}\n",
    "    * {\\bf The more you sample, the more your sample's distribution \n",
    "    will look like the population distribution} (obviously!)\n",
    "    * {\\bf The more you sample, the closer will your sample \n",
    "    statistic be to the population's statistical parameter} (the \n",
    "    central limit theorem)\n",
    "\\end{enumerate}\n",
    "\n",
    "Let's have a quick look at rule 1 using R (open R and {\\tt setwd} to \n",
    "{\\tt Code}):\n",
    "\n",
    "```r\n",
    "# Draw 5 normal random nos w/ mean 0 and s.d. 1:\n",
    "> MySample5 <- rnorm(5, m=0, sd=1) \n",
    "> MySample10 <- rnorm(10, m=0, sd=1) \n",
    "> MySample20 <- rnorm(20, m=0, sd=1) \n",
    "> MySample40 <- rnorm(40, m=0, sd=1)\n",
    "> MySample80 <- rnorm(80, m=0, sd=1)\n",
    "> MySample160 <- rnorm(160, m=0, sd=1)\n",
    "```\n",
    "\n",
    "Now let's visualize these \"samples\":\n",
    "\n",
    "```r\n",
    "> par(mfcol = c(2,3)) #initialize multi-paneled plot\n",
    "> par(mfg = c(1,1)); hist(MySample5, col = rgb(1,1,0), main = 'n = 5') \n",
    "> par(mfg = c(1,2)); hist(MySample10, col = rgb(1,1,0), main = 'n = 10') \n",
    "> par(mfg = c(1,3)); hist(MySample20, col = rgb(1,1,0), main = 'n = 20') \n",
    "> par(mfg = c(2,1)); hist(MySample40, col = rgb(1,1,0), main = 'n = 40') \n",
    "> par(mfg = c(2,2)); hist(MySample80, col = rgb(1,1,0), main = 'n = 80') \n",
    "> par(mfg = c(2,3)); hist(MySample160, col = rgb(1,1,0), main = 'n = 160') \n",
    "```\n",
    "\n",
    "The second rule above states that if I was to repeat even $n \n",
    "= 5$ sufficient number of times, I would get a good {\\it estimate} of \n",
    "mean (= 0) and standard deviation (= 1) of the normal distribution we \n",
    "sampled from. \n",
    "\n",
    "\n",
    "## A data exploration case study}\n",
    "\n",
    "As a case study, we will use data from a paper looking at the \n",
    "relationship between genome size and body size across species of \n",
    "dragonflies and damselflies \n",
    "(\\href{http://en.wikipedia.org/wiki/Odonata}{Odonata}):\n",
    "\n",
    "\\begin{quote}\n",
    "Ardila-Garcia, AM \\& Gregory, TR (2009) `An exploration of genome size \n",
    "diversity in dragonflies and damselflies (Insecta: Odonata)' Journal of \n",
    "Zoology, 278, 163 - 173\n",
    "\\end{quote}\n",
    "\n",
    "You will work with the script file {\\tt  ExpDesign.R}, which performs \n",
    "exploratory analyses on the data in {\\tt GenomeSize.csv}. Let's go \n",
    "through the code block by block. \n",
    "\n",
    "\\begin{compactitem}[$\\quad\\star$]\n",
    "    * Get the script {\\tt ExpDesign.R} from the Bitbucket \n",
    "    repository and put it in your own {\\tt Code} directory.\n",
    "  * Also get {\\tt GenomeSize.csv}\n",
    "  * Open the script {\\tt ExpDesign.R} in RStudio (or some other \n",
    "  text editor). \n",
    "  * Use the shift and arrow keys to select the code in block (2), \n",
    "  including the comments. Now use the keyboard short cut (look back \n",
    "  at the R Chapters if you don't know how!) to run the highlighted block \n",
    "  of code.\n",
    "\\end{compactitem}\n",
    "\n",
    "This first line (block (1)) reads in the data, as you have learned previously.\n",
    "\\begin{compactitem}[$\\quad\\star$]\n",
    "    * Now the code in block (2) line by line  \n",
    "    of code.\n",
    "\\end{compactitem}\n",
    "Have a good look at the data. There are three factors \n",
    "(categorical variables): Suborder, splitting the species into \n",
    "dragonflies (Anisoptera) and damselflies (Zygoptera); Family, splitting \n",
    "the species further into 9 taxonomic families; and Species, giving the \n",
    "latin binomial for each species in the table. The remaining columns are \n",
    "measurements of genome size (in picograms) and measurements of body \n",
    "size and morphology (in grams, mm and mm$^2$). There are two columns \n",
    "ending with an N that show the sample size from which the observations \n",
    "for each species are taken and a column ending SE showing standard \n",
    "errors.\n",
    "\n",
    "One thing you should see in the output from {\\tt head} or {\\tt str} is \n",
    "that there are some observations marked as {\\tt NA} -- this is the way R \n",
    "shows {\\it missing data}. It is important to check how much missing \n",
    "data there is an dataset, so we'll use another function that includes \n",
    "this information. Many R functions refuse to use variables containing \n",
    "missing data --- this is just R being careful and you can add {\\tt \n",
    "na.rm=TRUE} into most functions to avoid this problem.\n",
    " \n",
    "\\begin{compactitem}[$\\quad\\star$]\n",
    "* Run the {\\tt summary} line from the script window (block 3). \n",
    "\\end{compactitem}\n",
    "Look at the output. There is a column for each variable: for \n",
    "factors, it provides a short table of the number of observations in \n",
    "each level and for continuous variables, it provides some simple \n",
    "summary statistics about the distribution (range,  quartiles, mean and \n",
    "median), and the number of missing values \n",
    "\n",
    "\\subsection{Visualise distributions of the variables}\n",
    "The {\\tt summary} function shows us the basic distribution (range, \n",
    "quartiles, mean and median) of a continuous variable, but this is \n",
    "easier to interpret if we visualise it. We'll look at two ways: \n",
    "\\begin{description}\n",
    "    * [Histogram] In the simplest form, this shows the number of \n",
    "    observations of the variable falling into a set of bins spanning \n",
    "    the range of the variable. The option {\\tt breaks} allows you to \n",
    "    change the number of bins.\n",
    "    * [Density plot] Rather than showing blocks of counts, the \n",
    "    density plot shows a continuous smooth line. This is a {\\it \n",
    "    smoothed} estimate of the how frequently data is observed across \n",
    "    the range of values and the {\\it bandwidth} ({\\tt bw=0.1}) controls \n",
    "    the degree of the smoothing.\n",
    "\\end{description}\n",
    " \n",
    "\\begin{compactitem}[$\\quad\\star$]\n",
    "* Go to block (4) of the script and run each line separately, \n",
    "looking at the output. \n",
    "* In the editor, change the values of {\\tt breaks} and {\\tt bw} --- \n",
    "for example {\\tt breaks=5} and {\\tt bw=0.05} --- and re-run these lines \n",
    "to see how this affects the graph. Basically, with both types of graph \n",
    "you can look at the data too coarsely or too finely.\n",
    "* The graphs you've just created look at genome size. Add a copy of \n",
    "those two lines of code in the script and change them to look at the \n",
    "variable {\\tt TotalLength}. You will need to alter the {\\tt density} \n",
    "function to ignore missing values ({\\tt na.rm=TRUE}) and to play around \n",
    "with the bandwidth. You should get something like this: \n",
    "\\end{compactitem}\n",
    "% \n",
    "\\begin{center}\n",
    "\\includegraphics[width=0.45\\textwidth]{histTL1}\n",
    "\\includegraphics[width=0.45\\textwidth]{histTL2} \n",
    "\\end{center}\n",
    " \n",
    "\\subsection{Take a quick look at effects of certain factors}\n",
    "R has a special way of describing a model that defines the response \n",
    "variable and the explanatory variables (\"factors\"). This is called a \n",
    "'formula' and is used to define linear models (more on these in a later \n",
    "chapters). The same structure is used in many plotting functions and \n",
    "will put the response variable on the $y$ axis and the explanatory \n",
    "variable on the $x$ axis. The structure is \"response variable $\\sim$ \n",
    "explanatory variables\". We will look at multiple explanatory variables \n",
    "in a later practical but an example with one explantory variable \n",
    "(factor) is:\n",
    "\n",
    " {\\tt Genome Size $\\sim$ Suborder}\n",
    "\n",
    "This formula tells R to model genome size `as a function of' \n",
    "($\\sim$) the suborders of Odonata. In a plot function, the \n",
    "result will be to plot genome size as a function of the suborders.\n",
    " \n",
    "\\subsection{Compare distribution of the variable across levels of a \n",
    "factor}\n",
    "\n",
    "Although looking at the distribution of variables is a good first step, \n",
    "we often want to compare distributions. In this case, we might want to \n",
    "know how genome size varies between dragonflies and damselflies. The \n",
    "first way we will look at is using boxplots --- these show the median \n",
    "and the 25\\% and 75\\% quantiles as a box, with whiskers extending to \n",
    "the minimum and maximum. More extreme outliers are plotted \n",
    "independently as points. The {\\tt plot} function in R automatically \n",
    "generates a boxplot when the explanatory variable is a factor.\n",
    "\n",
    "\\begin{compactitem}[$\\quad\\star$]\n",
    "* Go to block 5 of the script and run the first line, looking at \n",
    "genome size between the two suborders.\n",
    "* Duplicate and alter this line to look at the same plot for total \n",
    "length. You should get a plot like this:\n",
    "\\end{compactitem}\n",
    " \n",
    "\\begin{center}\n",
    "    \\includegraphics[width=0.55\\textwidth]{bxpTL} \n",
    "\\end{center}\n",
    " \n",
    "Although histograms are great for one variable, plotting two histograms \n",
    "on top of one another rarely works well because the overlapping bars \n",
    "are hard to interpret (recall the predator-prey body size example). \n",
    "Density plots don't have this problem but it takes a bit more code to \n",
    "create the plot. \n",
    "\n",
    "\\begin{compactitem}[$\\quad\\star$]\n",
    "* block 6 of the script uses the {\\tt subset} function to create \n",
    "two new data frames separating the data for dragonflies and \n",
    "damselflies. Run the first two lines of this block. Remember that the \n",
    "arrow symbol ({\\tt <-}) is used to save the output of a function into a \n",
    "new object in R --- if you use {\\tt ls()} in the console, you will see \n",
    "the two new data frames.\n",
    "* In the console, use {\\tt str} and {\\tt summary} to explore these \n",
    "two new dataframes.\n",
    "\\end{compactitem}\n",
    "Now that we've got the data separated we can go about plotting the two \n",
    "curves. \n",
    "\\begin{compactitem}[$\\quad\\star$]\n",
    "* Run the next two lines of code in block 6. The first draws the \n",
    "plot for damselflies and the second adds a line for the dragonflies. \n",
    "* Duplicate these last two lines of code and edit them to generate \n",
    "a similar plot for total body length. You will need to edit the code to \n",
    "change the range of the $x$ and $y$ axes ({\\tt xlim} and {\\tt ylim}) to \n",
    "get both curves to fit neatly on to the graph. It should look like \n",
    "this:\n",
    "\\end{compactitem}\n",
    "\n",
    "\\begin{center}\n",
    "    \\includegraphics[width=0.45\\textwidth]{densTL} \n",
    "\\end{center}\n",
    "\n",
    "\\subsection{Explore further by scatter-plotting two variables}\n",
    " \n",
    "Once we've looked at the distribution of variables, the next thing is \n",
    "to look at the relationships between continuous variables using \n",
    "scatterplots. The {\\tt plot} function in R automatically generates a \n",
    "scatterplot when the explanatory variable is continuous, so we can use \n",
    "the same syntax and structure as for the boxplot above.\n",
    " \n",
    "\\begin{compactitem}[$\\quad\\star$]\n",
    "* Go to block (7) of the script and run the first plot command to \n",
    "plot body weight as a function of genome size.\n",
    "* Create a new line of code to plot forewing area as a function of \n",
    "genome size. It should look like this:\n",
    "\\end{compactitem}\n",
    "\n",
    "\\begin{center}\n",
    "\\includegraphics[width=0.5\\textwidth]{scatWing} \n",
    "\\end{center}\n",
    "The scatterplot seems to show a weak relationship between \n",
    "genome size and morphology. But maybe dragonflies and damselflies show \n",
    "different relationships, and we can't distinguish between them! To \n",
    "explore this possibility, we need to plot the two orders using \n",
    "different colours or plot characters. In the next code block, we want \n",
    "customize the plots to show different types of points for each \n",
    "suborder. It is done by using {\\it indexing} (covered in the first R Chapter in SilBioComp.pdf).\n",
    "\n",
    "\\begin{compactitem}[$\\quad\\star$]\n",
    "* Run the first two lines of code in block 8. There are two levels \n",
    "of suborder and these two lines set up a colour and a plot symbol that \n",
    "will be used for each one.\n",
    "* Run the next line which shows the structure of the factor {\\tt \n",
    "Suborder}. You can see that there are two levels, with Anisoptera first \n",
    "and then Zygoptera. You can also see that these are stored as numeric \n",
    "values: 1 refers to the first level and 2 will refer to the second. We \n",
    "can use these as {\\it indices} to pair the colours and plot symbols to \n",
    "each suborder. These are set in the {\\tt plot} function using the \n",
    "options {\\tt col=} and {\\tt pch=}, which stands for \"plot character\".\n",
    "* Run the next plot command  to see the resulting plot --- each \n",
    "point gets the appropriate colour and symbol for its group.\n",
    "\\end{compactitem}\n",
    " \n",
    "There are a lot of built in colours and plot symbols in R, so the next \n",
    "thing to experiment with is changing these to your own versions.\n",
    "\n",
    "\\begin{compactitem}[$\\quad\\star$]\n",
    "* In the console, type in the function {\\tt colors()}. You'll see a \n",
    "long list of options to choose from, so pick two to replace red and \n",
    "blue in the script window.\n",
    "* The options for the plot symbols are shown below. Pick two to \n",
    "replace the current symbol choices.\n",
    "* Rerun the {\\tt plot} function and see what you get!\n",
    "\\end{compactitem}\n",
    "\n",
    "\\begin{center}\n",
    "    \\includegraphics[width=0.55\\textwidth]{pch} \n",
    "\\end{center}\n",
    " \n",
    "\\subsection{Saving the exploratory graphics}\n",
    "The file `GenomeSize.pdf' in the practical folder was created using the \n",
    "next block of code using the approach you learned previously. \n",
    "The function {\\tt pdf} opens a new empty pdf file  which can then be \n",
    "used to plot graphs. You can set the width and the height of the page \n",
    "size in the pdf but note that this is set in {\\it inches}. When you \n",
    "have finished creating a plot, the function {\\tt dev.off} closes the \n",
    "pdf file and makes it readable.\n",
    "\n",
    "\\begin{compactitem}[$\\quad\\star$]\n",
    "* Open 'GenomeSize.pdf' in a PDF reader. It uses the original \n",
    "colours and plot symbols. Close the file and then delete it from the \n",
    "folder.\n",
    "* Now go back to the script in R and select and run all the code in \n",
    "block (9)\n",
    "* Go back to the {\\tt Results} folder. The pdf file should have \n",
    "been recreated --- open it and it should now use your choice of colours \n",
    "and symbols.\n",
    "\\end{compactitem}\n",
    "\n",
    "\\subsection{Saving data}\n",
    "\n",
    "One last thing you can do is to save the data and variables in R format \n",
    "--- the original data, two subsets of the data and the two sets of \n",
    "colours and symbols. We can recreate the subsets easily, so we'll just \n",
    "save the data and your colour sets.\n",
    "\n",
    "\\begin{compactitem}[$\\quad\\star$]\n",
    "* Go to the script window and run the final line in block (10)\n",
    "* Still in the script window, choose `File $\\triangleright$ Save' \n",
    "to save your changes to the script file.\n",
    "* Quit from R {\\tt q()}.\n",
    "\\end{compactitem}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Readings and Resources <a id='Readings'></a>\n",
    "\n",
    "* Motulsky, Harvey, and Arthur Christopoulos. Fitting models to biological data using linear and nonlinear regression: a practical guide to curve fitting. OUP USA, 2004.\n",
    "* Johnson, J. B. & Omland, K. S. 2004 Model selection in ecology and evolution. Trends Ecol. Evol. 19, 101–108.\n",
    "* The [NCEAS non-linear modelling working group](https://groups.nceas.ucsb.edu/non-linear-modeling/projects/OrangeTree)\n",
    "* [Mixed-Effects Models in S and S-PLUS](https://link.springer.com/book/10.1007/b98882)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.3"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autocomplete": false,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "171px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
