{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Introduction\n",
    "============\n",
    "\n",
    "> <span>Donald Knuth, 1995: </span> <span>*Science is what we understand\n",
    "> well enough to explain to a computer. Art is everything else we\n",
    "> do.*</span>\n",
    "\n",
    "What is this document about?\n",
    "----------------------------\n",
    "\n",
    "This document contains the content of the Imperial College London\n",
    "Department of Life Sciences BSc Year 1 (Y1) and 2 (Y2) statistics\n",
    "modules, including workshops and practicals. Different subsets of the\n",
    "content of this document will be covered in Y1 and Y2. You will be given\n",
    "instructions about which sections will be covered in your respective (Y1\n",
    "or Y2) modules. You will learn statistics with R in a hands-on on way in\n",
    "this module. A few “pure” lectures on statistical concepts will also be\n",
    "delivered.\n",
    "\n",
    "This document is accompanied by data and code on which you can practice\n",
    "your skills in your own time and during the Practical sessions. These\n",
    "materials are available (and will be updated regularly) at bitbucket\n",
    "(it’s like GitHub): <https://bitbucket.org/mhasoba/iccompbiostat>.\n",
    "\n",
    "<span>**Note that you can download the code, data and everything from\n",
    "the bitbucket repository at one go, by going to\n",
    "<https://bitbucket.org/mhasoba/iccompbiostat> and then clicking on the\n",
    "“Download repository” link. You can then unzip the downloaded .zip and\n",
    "grab the files you need.**</span>\n",
    "\n",
    "These notes and the scripts that you create form a valuable reference\n",
    "for using and interpreting statistics. You will be expected to be able\n",
    "to use your experience and these notes to complete analyses for\n",
    "statistical components of other courses, in the second and third year in\n",
    "particular.\n",
    "\n",
    "<span> The topics covered here assume that you have already worked\n",
    "through at least the basic R sections of the <span>SilBioComp.pdf</span>\n",
    "document (also available in the same bitbucket repository)</span>\n",
    "\n",
    "Finally, it is important that you work through the problems in each\n",
    "chapter, particularly as some of the questions ask you to find out about\n",
    "commands and functions not introduced in the chapter’s text itself, but\n",
    "which will be relied on in later chapters. By the time you have attended\n",
    "the lectures and workshops, and completed the exercises in this\n",
    "document, you should able to:\n",
    "\n",
    "Use <span>R</span> to perform common statistical tests, particularly t,\n",
    "F, and $\\chi^{2}$ tests\n",
    "\n",
    "Develop ability to build, criticize and interpret linear models,\n",
    "especially linear regression and ANOVA.\n",
    "\n",
    "Interpret R output (particularly p values) correctly, and use them\n",
    "appropriately in practical write-ups and presentations\n",
    "\n",
    "Conventions used in this document\n",
    "---------------------------------\n",
    "\n",
    "You will find all R commandline/console arguments, code snippets and\n",
    "output in colored boxes like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> ls()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here $>$ is the R prompt, and will type (or copy-paste, but not\n",
    "recommended!) the commands/code that you see from this document into the\n",
    "R command line. But please exclude the $>$ as this is just the R command\n",
    "prompt! I have included the $>$ prompts in the code shown in this\n",
    "document so that you are forced to see what each line is doing. Indeed,\n",
    "avoid copying-and-pasting chunks of R code you do not understand:\n",
    "blindly shovelling data into a black box and assuming the output is\n",
    "correct and meaningful will eventually lead to frustrations, and if you\n",
    "are unlucky, embarrassments or even catastrophes!\n",
    "\n",
    "The content of this document is computer platform (Mac, PC or Linux)\n",
    "independent because many of you will probably also later be working with\n",
    "R on your personal laptops or desktops. Indeed, platform-independence of\n",
    "your statistical analysis is one of the main reasons why you are using\n",
    "R!\n",
    "\n",
    "Finally, note that:\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "In all subsequent chapters, lines marked with a star like this are\n",
    "things for you to do.\n",
    "\n",
    "Readings\n",
    "--------\n",
    "\n",
    "Look up the Readings directory on the bitbucket repository (link given\n",
    "above).\n",
    "\n",
    "-   Bolker, B. M.: Ecological Models and Data in R (eBook and\n",
    "    Hardcover available).\n",
    "\n",
    "-   Beckerman, A. P. & Petchey, O. L. (2012) Getting started with R: an\n",
    "    introduction for biologists. Oxford, Oxford University Press.\\\n",
    "    Good, short, general introduction\n",
    "\n",
    "-   Crawley, R. (2013) The R book. 2nd edition. Chichester, Wiley.\\\n",
    "    Excellent but enormous reference book, code and data available from\n",
    "    [www.bio.ic.ac.uk/research/mjcraw/therbook/index.htm](www.bio.ic.ac.uk/research/mjcraw/therbook/index.htm)\n",
    "\n",
    "-   Use the internet! Type “R tutorial”, and scores will pop up. Choose\n",
    "    one that seems the most intuitive to you.\n",
    "\n",
    "Experimental design and Data exploration {#ch:ExpDesign}\n",
    "========================================\n",
    "\n",
    "Ideally, you would like to design experiments (manipulations and/or\n",
    "observations) that are appropriate for the question you want to answer.\n",
    "However, you still need to explore you data to determine what kind of\n",
    "statistical tests would be appropriate because: (a) Your experiments or\n",
    "observations may not go as planned, and (b) You might have somebody\n",
    "else’s data to analyse (very likely in your UG projects). Building on\n",
    "the previous chapter, in this chapter we you learn how to use R to\n",
    "explore your data and determine appropriate statistical tests. By the\n",
    "time you have worked through this chapter, you should be able to\n",
    "\n",
    "Provided sufficient information is available, be able to judge whether\n",
    "the sampling design used to generate a particular dataset was\n",
    "appropriate\n",
    "\n",
    "Determine if your sample sizes are adequate, especially for a specific\n",
    "statistical test qualitatively (Y1) or quantitatively using power\n",
    "analysis (Y2).\n",
    "\n",
    "Calculate basic statistical measures on your data to determine its\n",
    "properties\n",
    "\n",
    "We are going to start off in Y1 with the simplest of scenarios for\n",
    "statistical testing — that you want to determine whether a sample, or a\n",
    "pair of samples meet some expectation (hypothesis) or not.\n",
    "\n",
    "First, some conceptual preliminaries.\n",
    "\n",
    "Some statistical parlance\n",
    "-------------------------\n",
    "\n",
    "The following terms are important for you to get familiar with:\n",
    "\n",
    "(Statistical) Population\n",
    "\n",
    ":   A statistical population is a <span> *complete set*</span> of items\n",
    "    that share at least one <span>*attribute*</span> of interest. This\n",
    "    attribute of interest is the target of your statistical analysis.\n",
    "    For example, if we are interested in studying the weight of year-old\n",
    "    cod in the Oceans, the population consists of <span>*all*</span>\n",
    "    year-old cod, but more specifically, the weight measurements of all\n",
    "    the individuals of the cod population is what we want to analyse.\n",
    "\n",
    "(Statistical) Distribution\n",
    "\n",
    ":   A statistical distribution is a mathematical description (expressed\n",
    "    as a mathematical equation) of the properties of a population\n",
    "    of interest. Theoreticians have come up with a bunch of\n",
    "    distributions (e.g., Gaussian or Normal, Poisson, Binomial, etc.)\n",
    "    that are appropriate for different kinds of data. Figuring out which\n",
    "    distribution best describes a population of interest is one of the\n",
    "    first steps in a statistical analysis. In reality, of course, even\n",
    "    collecting and measuring all the individuals of a population may not\n",
    "    be sufficient to characterize its statistical properties — imagine\n",
    "    the situation where the cod population has declined to a few hundred\n",
    "    individuals (not an impossibility in the future!).\n",
    "\n",
    "(Data or Population) Sample\n",
    "\n",
    ":   A data <span>*sample*</span> is a set of measurements of the\n",
    "    attribute of interest collected from a <span> *statistical\n",
    "    population*</span> by a defined procedure (<span>*sampling\n",
    "    methodology*</span>). In the cod example above, this would be the\n",
    "    weight of every individual of a <span>*subset*</span> of the\n",
    "    year-old Cod population.\n",
    "\n",
    "(Statistical) Parameter\n",
    "\n",
    ":   A statistical parameter is a measure of some attribute of the\n",
    "    <span>*theoretical*</span> statistical distribution that is support\n",
    "    to represent your population. An example would be the average weight\n",
    "    of yearling cod. In practice, this is not measurable because the\n",
    "    population is much too large or incompletely inaccessible/invisible\n",
    "    — imagine measuring the weight of every year-old cod individual in\n",
    "    the oceans!\n",
    "\n",
    "Statistic\n",
    "\n",
    ":   A statistic (singular) is an <span>*estimate*</span> of a\n",
    "    statistical parameter of the population of interest, obtained by\n",
    "    calculating the <span>*measure*</span> for a <span>*sample*</span>\n",
    "    (e.g., the average or mean weight of individuals in a sample of\n",
    "    one-year old cod). This is also your <span>*descriptive\n",
    "    statistic*</span>. Therefore, a <span>*Statistic*</span> is to a\n",
    "    <span>*Statistical Parameter*</span> what a <span> *Sample*</span>\n",
    "    is to the <span>*Statistical Population*</span>. For example, the\n",
    "    average of a sample of cod weights is a statistic that <span>\n",
    "    *estimates*</span> the theoretical “real” average of the weights of\n",
    "    the entire one-year Cod population, which is its\n",
    "    statistical parameter.\n",
    "\n",
    "Hypothesis\n",
    "\n",
    ":   A Hypothesis is an (hopefully) informed <span> *postulate*</span>\n",
    "    about an attribute of your population of interest. For example, you\n",
    "    may hypothesize that the one-year old cod population’s mean weight\n",
    "    has declined over the last two decades (it has!). You will want to\n",
    "    confront your main hypothesis with a “null” Hypothesis, to minimize\n",
    "    the risk of making a “type I” error. A type I error is the\n",
    "    probability of accepting an alternative (or main) hypothesis (and\n",
    "    rejecting the null hypothesis) that is not really valid (e.g., the\n",
    "    yearling cods have actually not declined in weight). This is a big\n",
    "    NO NO from a scientific and philosophical standpoint. The rate or\n",
    "    probability of the type I error is denoted by the Greek letter\n",
    "    $\\alpha$, and equals the <span>*significance level*</span> of a\n",
    "    statistical test.\n",
    "\n",
    "Descriptive Statistics\n",
    "----------------------\n",
    "\n",
    "The fundamental statistics that describe a sample (or a population) are,\n",
    "firstly, the mean, or average value of a sample, typically denoted by a\n",
    "$\\bar{x}$:\n",
    "\n",
    "$$\\bar{x} =  \\frac{\\sum\\limits_{i=1}^n x_i}{n} = \\frac{x_{1} + x_{2} + \\dots +x_{n}}{n}$$\n",
    "\n",
    "That is, it is the sum of all the values in a sample divided by the\n",
    "number, $n$, of items in the sample. Thus it is a measure of the <span>\n",
    "*central-tendency*</span> of the sample and population.\n",
    "\n",
    "Second, the standard deviation ($s$) is:\n",
    "\n",
    "$$s =  \\sqrt{\\frac{\\sum\\limits_{i=1}^n (\\bar{x} - x_{i})^2}{n-1}} = \n",
    "\t\\sqrt{\\frac{(\\bar{x} - x_{1})^{2} + (\\bar{x} - x_{2})^2 + \\dots + (\\bar{x} \n",
    "\t- x_{n})^{2}}{n-1}}$$\n",
    "\n",
    "That is, the the square root of the sum of squares (“SS”) of the\n",
    "differences between each item in the sample and the mean, divided by the\n",
    "<span>*degrees of freedom, “df”*</span> remaining in the data set ($ \n",
    "n-1$). df is the sample size, $n$, minus the number of statistical\n",
    "parameters estimated from the data set. This is to reduce the <span>\n",
    "*bias*</span> in your <span>*estimate*</span> of the statistic, as you\n",
    "are calculating it from the sample, and not the whole theoretical\n",
    "population.\n",
    "\n",
    "Thus, the formula for $s$ above has $n-1$ in its denominator because, to\n",
    "to work out the standard deviation, you must have already estimated the\n",
    "mean ($\\bar{x}$) from the same data set. This removes 1 degree of\n",
    "freedom. Also, note that the sample variance, $s^2$ is the square of\n",
    "standard deviation. Or, in other words, the standard deviation is the\n",
    "square-root of the variance!\n",
    "\n",
    "The above two statistics (mean and sd) are particularly meaningful when\n",
    "the sample and population have a symmetric distribution (e.g., normal or\n",
    "gaussian). When the distribution is not symmetric (that is, it is\n",
    "<span>*skewed*</span>), another statistic, the <span>*median*</span>\n",
    "becomes important. This is the middle value in the ordered set of data,\n",
    "that is exactly 50% of the data live below and 50% lie above the median.\n",
    "In skewed distributions, the median is a better measure of the\n",
    "<span>*central-tendency*</span> of the sample and population.\n",
    "\n",
    "Other descriptive statistics you should keep in mind are the range\n",
    "(difference between the largest and smallest values), and the quartiles\n",
    "(values lying in the data divided into the intervals\n",
    "$[{1\\over4}, {1\\over \n",
    "2},{3\\over 4}, 1]$ or at 1% intervals (percentiles). Box-plots, which\n",
    "you have seen, represent a number of these statistics in one figure.\n",
    "\n",
    "### Descriptive statistic functions in R\n",
    "\n",
    "  ------------------------------- --------------------------------------\n",
    "  <span>mean(x)</span>            Compute mean (of a vector or matrix)\n",
    "  <span>sd(x)</span>              Standard deviation\n",
    "  <span>var(x)</span>             Variance\n",
    "  <span>median(x)</span>          Median\n",
    "  <span>quantile(x,0.05)</span>   Compute the 0.05 quantile\n",
    "  <span>range(x)</span>           Range of the data\n",
    "  <span>min(x)</span>             Minimum\n",
    "  <span>max(x)</span>             Maximum\n",
    "  <span>sum(x)</span>             Sum all elements\n",
    "  ------------------------------- --------------------------------------\n",
    "\n",
    "\\\n",
    "\n",
    "Data types and distributions\n",
    "----------------------------\n",
    "\n",
    "You will typically encounter or sample the following main types of data:\n",
    "\n",
    "Continuous numeric\n",
    "\n",
    ":   This is the <span>numeric</span> or <span>real</span> data type in\n",
    "    R, and as far as you are concerned, these data typically will be\n",
    "    made up of (mathematically) real numbers such as human height\n",
    "    or weight. These may be unbounded (any value between negative\n",
    "    infinity to positive infinity), or bounded (e.g., between or zero\n",
    "    and some upper positive number) like human weight.\n",
    "\n",
    "Discrete numeric\n",
    "\n",
    ":   This is the <span>integer</span> data type in R, and consist\n",
    "    of (mathematically) integer (whole) numbers such as counts of\n",
    "    individuals in a population, e.g., The number of bacteria in a ml of\n",
    "    pond water.\n",
    "\n",
    "Percentage (proportion)\n",
    "\n",
    ":   Percentage data is a particular kind of numeric data that is\n",
    "    strictly bounded between 0 and 100. The fact that you can never get\n",
    "    samples of percentages that exceed these bounds makes such data\n",
    "    tricky to analyse.\n",
    "\n",
    "Categorical\n",
    "\n",
    ":   These are typically stored as the <span> character</span> data type\n",
    "    in R. Categorical data are discrete, typically expressed as a one of\n",
    "    a fixed number of <span>*levels*</span> of a <span>*factor*</span>.\n",
    "    For example, the the factor “Type.of.feeding.interaction” from the\n",
    "    predator-prey dataset you have seen previously had five levels:\n",
    "    “insectivorous”, “piscivorous”, “planktivorous”, “predacious”, and\n",
    "    “predacious/piscivorous”.\n",
    "\n",
    "Binary (presence/absence) data\n",
    "\n",
    ":   A special type of categorical data are binary, where only two\n",
    "    categories or states are possible: (1, 0) (or “present”, “absent”),\n",
    "    e.g., a disease symptom. These may be stored as <span>integer</span>\n",
    "    or <span> character</span> in R.\n",
    "\n",
    "While designing experiments or exploring data obtained by somebody else,\n",
    "you need to keep in mind that each type will typically be best\n",
    "represented a particular <span>*statistical distribution*</span>. For\n",
    "example, continuous numeric data are <span>*often*</span> normally\n",
    "distributed. On the other hand, count data are likely to be distributed\n",
    "according to the Poisson distribution.\n",
    "\n",
    "If you are lucky, you will mostly have to deal with data that are\n",
    "continuous or discrete numeric, which are the most straightforward to\n",
    "analyse using Linear models (more on that in subsequent chapters).\n",
    "However, some of the most interesting and important problems in biology\n",
    "involve proportion (percentage), categorical and binary data (e.g.,\n",
    "Presence or absence of a disease symptom).\n",
    "\n",
    "For example, think about what type of data, and what type of\n",
    "distribution, a sample of the following is likely to be:\n",
    "\n",
    "Wavelength of light\n",
    "\n",
    "Temperature\n",
    "\n",
    "Egg clutch size\n",
    "\n",
    "Rate of a reaction\n",
    "\n",
    "Eye-colour\n",
    "\n",
    "Score in Scrabble\n",
    "\n",
    "UG Degree class\n",
    "\n",
    "Ground-cover of grass in a quadrat\n",
    "\n",
    "Winning side in chess\n",
    "\n",
    "### Sampling from distributions in R\n",
    "\n",
    "You can generate samples form many distributions in R (and handy thing\n",
    "to know). In particular, the following are important:\n",
    "\n",
    "  -------------------------------------- ------------------------------------------------------\n",
    "  <span>rnorm(10, m=0, sd=1)</span>      Draw 10 normal random numbers with mean 0 and s.d. 1\n",
    "  <span>dnorm(x, m=0, sd=1)</span>       Density function\n",
    "  <span>qnorm(x, m=0, sd=1)</span>       Cumulative density function\n",
    "  <span>runif(20, min=0, max=2)</span>   Twenty random numbers from uniform \\[0,2\\]\n",
    "  <span>rpois(20, lambda=10)</span>      Twenty random numbers from Poisson($\\lambda$)\n",
    "  -------------------------------------- ------------------------------------------------------\n",
    "\n",
    "\\\n",
    "\n",
    "Two basic rules of experimental design and sampling\n",
    "---------------------------------------------------\n",
    "\n",
    "In general, while designing experiments, and sampling from a <span>\n",
    "*population*</span>, there are two key (and simple) rules:\n",
    "\n",
    "1.  <span>**The more you sample, the more your sample’s distribution\n",
    "    will look like the population distribution**</span> (obviously!)\n",
    "\n",
    "2.  <span>**The more you sample, the closer will your sample statistic\n",
    "    be to the population’s statistical parameter**</span> (the central\n",
    "    limit theorem)\n",
    "\n",
    "Let’s have a quick look at rule 1 using R (open R and <span>setwd</span>\n",
    "to <span>Code</span>):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Draw 5 normal random nos w/ mean 0 and s.d. 1:\n",
    "> MySample5 <- rnorm(5, m=0, sd=1) \n",
    "> MySample10 <- rnorm(10, m=0, sd=1) \n",
    "> MySample20 <- rnorm(20, m=0, sd=1) \n",
    "> MySample40 <- rnorm(40, m=0, sd=1)\n",
    "> MySample80 <- rnorm(80, m=0, sd=1)\n",
    "> MySample160 <- rnorm(160, m=0, sd=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let’s visualize these “samples”:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> par(mfcol = c(2,3)) #initialize multi-paneled plot\n",
    "> par(mfg = c(1,1)); hist(MySample5, col = rgb(1,1,0), main = 'n = 5') \n",
    "> par(mfg = c(1,2)); hist(MySample10, col = rgb(1,1,0), main = 'n = 10') \n",
    "> par(mfg = c(1,3)); hist(MySample20, col = rgb(1,1,0), main = 'n = 20') \n",
    "> par(mfg = c(2,1)); hist(MySample40, col = rgb(1,1,0), main = 'n = 40') \n",
    "> par(mfg = c(2,2)); hist(MySample80, col = rgb(1,1,0), main = 'n = 80') \n",
    "> par(mfg = c(2,3)); hist(MySample160, col = rgb(1,1,0), main = 'n = 160') "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The second rule above states that if I was to repeat even $n \n",
    "= 5$ sufficient number of times, I would get a good\n",
    "<span>*estimate*</span> of mean (= 0) and standard deviation (= 1) of\n",
    "the normal distribution we sampled from.\n",
    "\n",
    "A data exploration case study\n",
    "-----------------------------\n",
    "\n",
    "As a case study, we will use data from a paper looking at the\n",
    "relationship between genome size and body size across species of\n",
    "dragonflies and damselflies\n",
    "([Odonata](http://en.wikipedia.org/wiki/Odonata)):\n",
    "\n",
    "> Ardila-Garcia, AM & Gregory, TR (2009) ‘An exploration of genome size\n",
    "> diversity in dragonflies and damselflies (Insecta: Odonata)’ Journal\n",
    "> of Zoology, 278, 163 - 173\n",
    "\n",
    "You will work with the script file <span>ExpDesign.R</span>, which\n",
    "performs exploratory analyses on the data in\n",
    "<span>GenomeSize.csv</span>. Let’s go through the code block by block.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Get the script <span>ExpDesign.R</span> from the Bitbucket repository\n",
    "and put it in your own <span>Code</span> directory.\n",
    "\n",
    "Also get <span>GenomeSize.csv</span>\n",
    "\n",
    "Open the script <span>ExpDesign.R</span> in RStudio (or some other text\n",
    "editor).\n",
    "\n",
    "Use the shift and arrow keys to select the code in block (2), including\n",
    "the comments. Now use the keyboard short cut (look back at the R\n",
    "Chapters if you don’t know how!) to run the highlighted block of code.\n",
    "\n",
    "This first line (block (1)) reads in the data, as you have learned\n",
    "previously.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Now the code in block (2) line by line of code.\n",
    "\n",
    "Have a good look at the data. There are three factors (categorical\n",
    "variables): Suborder, splitting the species into dragonflies\n",
    "(Anisoptera) and damselflies (Zygoptera); Family, splitting the species\n",
    "further into 9 taxonomic families; and Species, giving the latin\n",
    "binomial for each species in the table. The remaining columns are\n",
    "measurements of genome size (in picograms) and measurements of body size\n",
    "and morphology (in grams, mm and mm$^2$). There are two columns ending\n",
    "with an N that show the sample size from which the observations for each\n",
    "species are taken and a column ending SE showing standard errors.\n",
    "\n",
    "One thing you should see in the output from <span>head</span> or\n",
    "<span>str</span> is that there are some observations marked as\n",
    "<span>NA</span> – this is the way R shows <span>*missing data*</span>.\n",
    "It is important to check how much missing data there is an dataset, so\n",
    "we’ll use another function that includes this information. Many R\n",
    "functions refuse to use variables containing missing data — this is just\n",
    "R being careful and you can add <span> na.rm=TRUE</span> into most\n",
    "functions to avoid this problem.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Run the <span>summary</span> line from the script window (block 3).\n",
    "\n",
    "Look at the output. There is a column for each variable: for factors, it\n",
    "provides a short table of the number of observations in each level and\n",
    "for continuous variables, it provides some simple summary statistics\n",
    "about the distribution (range, quartiles, mean and median), and the\n",
    "number of missing values\n",
    "\n",
    "### Visualise distributions of the variables\n",
    "\n",
    "The <span>summary</span> function shows us the basic distribution\n",
    "(range, quartiles, mean and median) of a continuous variable, but this\n",
    "is easier to interpret if we visualise it. We’ll look at two ways:\n",
    "\n",
    "Histogram\n",
    "\n",
    ":   In the simplest form, this shows the number of observations of the\n",
    "    variable falling into a set of bins spanning the range of\n",
    "    the variable. The option <span>breaks</span> allows you to change\n",
    "    the number of bins.\n",
    "\n",
    "Density plot\n",
    "\n",
    ":   Rather than showing blocks of counts, the density plot shows a\n",
    "    continuous smooth line. This is a <span> *smoothed*</span> estimate\n",
    "    of the how frequently data is observed across the range of values\n",
    "    and the <span>*bandwidth*</span> (<span>bw=0.1</span>) controls the\n",
    "    degree of the smoothing.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Go to block (4) of the script and run each line separately, looking at\n",
    "the output.\n",
    "\n",
    "In the editor, change the values of <span>breaks</span> and\n",
    "<span>bw</span> — for example <span>breaks=5</span> and\n",
    "<span>bw=0.05</span> — and re-run these lines to see how this affects\n",
    "the graph. Basically, with both types of graph you can look at the data\n",
    "too coarsely or too finely.\n",
    "\n",
    "The graphs you’ve just created look at genome size. Add a copy of those\n",
    "two lines of code in the script and change them to look at the variable\n",
    "<span>TotalLength</span>. You will need to alter the\n",
    "<span>density</span> function to ignore missing values\n",
    "(<span>na.rm=TRUE</span>) and to play around with the bandwidth. You\n",
    "should get something like this:\n",
    "\n",
    "![image](histTL1){width=\"45.00000%\"}\n",
    "![image](histTL2){width=\"45.00000%\"}\n",
    "\n",
    "### Take a quick look at effects of certain factors\n",
    "\n",
    "R has a special way of describing a model that defines the response\n",
    "variable and the explanatory variables (“factors”). This is called a\n",
    "’formula’ and is used to define linear models (more on these in a later\n",
    "chapters). The same structure is used in many plotting functions and\n",
    "will put the response variable on the $y$ axis and the explanatory\n",
    "variable on the $x$ axis. The structure is “response variable $\\sim$\n",
    "explanatory variables”. We will look at multiple explanatory variables\n",
    "in a later practical but an example with one explantory variable\n",
    "(factor) is:\n",
    "\n",
    "<span>Genome Size $\\sim$ Suborder</span>\n",
    "\n",
    "This formula tells R to model genome size ‘as a function of’ ($\\sim$)\n",
    "the suborders of Odonata. In a plot function, the result will be to plot\n",
    "genome size as a function of the suborders.\n",
    "\n",
    "### Compare distribution of the variable across levels of a factor\n",
    "\n",
    "Although looking at the distribution of variables is a good first step,\n",
    "we often want to compare distributions. In this case, we might want to\n",
    "know how genome size varies between dragonflies and damselflies. The\n",
    "first way we will look at is using boxplots — these show the median and\n",
    "the 25% and 75% quantiles as a box, with whiskers extending to the\n",
    "minimum and maximum. More extreme outliers are plotted independently as\n",
    "points. The <span>plot</span> function in R automatically generates a\n",
    "boxplot when the explanatory variable is a factor.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Go to block 5 of the script and run the first line, looking at genome\n",
    "size between the two suborders.\n",
    "\n",
    "Duplicate and alter this line to look at the same plot for total length.\n",
    "You should get a plot like this:\n",
    "\n",
    "![image](bxpTL){width=\"55.00000%\"}\n",
    "\n",
    "Although histograms are great for one variable, plotting two histograms\n",
    "on top of one another rarely works well because the overlapping bars are\n",
    "hard to interpret (recall the predator-prey body size example). Density\n",
    "plots don’t have this problem but it takes a bit more code to create the\n",
    "plot.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "block 6 of the script uses the <span>subset</span> function to create\n",
    "two new data frames separating the data for dragonflies and damselflies.\n",
    "Run the first two lines of this block. Remember that the arrow symbol\n",
    "(<span>&lt;-</span>) is used to save the output of a function into a new\n",
    "object in R — if you use <span>ls()</span> in the console, you will see\n",
    "the two new data frames.\n",
    "\n",
    "In the console, use <span>str</span> and <span>summary</span> to explore\n",
    "these two new dataframes.\n",
    "\n",
    "Now that we’ve got the data separated we can go about plotting the two\n",
    "curves.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Run the next two lines of code in block 6. The first draws the plot for\n",
    "damselflies and the second adds a line for the dragonflies.\n",
    "\n",
    "Duplicate these last two lines of code and edit them to generate a\n",
    "similar plot for total body length. You will need to edit the code to\n",
    "change the range of the $x$ and $y$ axes (<span>xlim</span> and\n",
    "<span>ylim</span>) to get both curves to fit neatly on to the graph. It\n",
    "should look like this:\n",
    "\n",
    "![image](densTL){width=\"45.00000%\"}\n",
    "\n",
    "### Explore further by scatter-plotting two variables\n",
    "\n",
    "Once we’ve looked at the distribution of variables, the next thing is to\n",
    "look at the relationships between continuous variables using\n",
    "scatterplots. The <span>plot</span> function in R automatically\n",
    "generates a scatterplot when the explanatory variable is continuous, so\n",
    "we can use the same syntax and structure as for the boxplot above.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Go to block (7) of the script and run the first plot command to plot\n",
    "body weight as a function of genome size.\n",
    "\n",
    "Create a new line of code to plot forewing area as a function of genome\n",
    "size. It should look like this:\n",
    "\n",
    "![image](scatWing){width=\"50.00000%\"}\n",
    "\n",
    "The scatterplot seems to show a weak relationship between genome size\n",
    "and morphology. But maybe dragonflies and damselflies show different\n",
    "relationships, and we can’t distinguish between them! To explore this\n",
    "possibility, we need to plot the two orders using different colours or\n",
    "plot characters. In the next code block, we want customize the plots to\n",
    "show different types of points for each suborder. It is done by using\n",
    "<span>*indexing*</span> (covered in the first R Chapter in\n",
    "SilBioComp.pdf).\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Run the first two lines of code in block 8. There are two levels of\n",
    "suborder and these two lines set up a colour and a plot symbol that will\n",
    "be used for each one.\n",
    "\n",
    "Run the next line which shows the structure of the factor <span>\n",
    "Suborder</span>. You can see that there are two levels, with Anisoptera\n",
    "first and then Zygoptera. You can also see that these are stored as\n",
    "numeric values: 1 refers to the first level and 2 will refer to the\n",
    "second. We can use these as <span>*indices*</span> to pair the colours\n",
    "and plot symbols to each suborder. These are set in the\n",
    "<span>plot</span> function using the options <span>col=</span> and\n",
    "<span>pch=</span>, which stands for “plot character”.\n",
    "\n",
    "Run the next plot command to see the resulting plot — each point gets\n",
    "the appropriate colour and symbol for its group.\n",
    "\n",
    "There are a lot of built in colours and plot symbols in R, so the next\n",
    "thing to experiment with is changing these to your own versions.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "In the console, type in the function <span>colors()</span>. You’ll see a\n",
    "long list of options to choose from, so pick two to replace red and blue\n",
    "in the script window.\n",
    "\n",
    "The options for the plot symbols are shown below. Pick two to replace\n",
    "the current symbol choices.\n",
    "\n",
    "Rerun the <span>plot</span> function and see what you get!\n",
    "\n",
    "![image](pch){width=\"55.00000%\"}\n",
    "\n",
    "### Saving the exploratory graphics\n",
    "\n",
    "The file ‘GenomeSize.pdf’ in the practical folder was created using the\n",
    "next block of code using the approach you learned previously. The\n",
    "function <span>pdf</span> opens a new empty pdf file which can then be\n",
    "used to plot graphs. You can set the width and the height of the page\n",
    "size in the pdf but note that this is set in <span>*inches*</span>. When\n",
    "you have finished creating a plot, the function <span>dev.off</span>\n",
    "closes the pdf file and makes it readable.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Open ’GenomeSize.pdf’ in a PDF reader. It uses the original colours and\n",
    "plot symbols. Close the file and then delete it from the folder.\n",
    "\n",
    "Now go back to the script in R and select and run all the code in block\n",
    "(9)\n",
    "\n",
    "Go back to the <span>Results</span> folder. The pdf file should have\n",
    "been recreated — open it and it should now use your choice of colours\n",
    "and symbols.\n",
    "\n",
    "### Saving data\n",
    "\n",
    "One last thing you can do is to save the data and variables in R format\n",
    "— the original data, two subsets of the data and the two sets of colours\n",
    "and symbols. We can recreate the subsets easily, so we’ll just save the\n",
    "data and your colour sets.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Go to the script window and run the final line in block (10)\n",
    "\n",
    "Still in the script window, choose ‘File $\\triangleright$ Save’ to save\n",
    "your changes to the script file.\n",
    "\n",
    "Quit from R <span>q()</span>.\n",
    "\n",
    "$t$ tests and $F$ tests {#ch:t_F_tests}\n",
    "=======================\n",
    "\n",
    "Aims of this chapter[^1]:\n",
    "\n",
    "Using $t$ tests to look at differences between means\n",
    "\n",
    "Using $F$ tests to compare the variance of two samples\n",
    "\n",
    "Using non-parametric tests for differences\n",
    "\n",
    "In the last chapter, we looked at the genome size and morphology of\n",
    "species of dragonflies and damselflies (Odonates: Anisoptera and\n",
    "Zygoptera). Box and whisker plots and density plots both show that the\n",
    "two groups have rather different genome size and morphology. We can use\n",
    "$t$ tests to test whether the means of the variables of the two groups\n",
    "are different and we can use the $F$ test to check whether they have the\n",
    "same variance.\n",
    "\n",
    "In this chapter, we will continue to practise building scripts and\n",
    "creating your own R code, so we will start from an empty script file.\n",
    "Use this script to store the code you run in this practical session and\n",
    "add notes to keep a record of what each bit is doing:\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Open R and change to the <span>Code</span> directory.\n",
    "\n",
    "If you have misplaced the data then it is on Blackboard / Bitbucket, so\n",
    "download it again.\n",
    "\n",
    "Create a new blank script called <span>ttests.R</span> and save it to\n",
    "the working directory.\n",
    "\n",
    "Put a comment at the top (using \\#) to describe the script and.\n",
    "\n",
    "For the rest of this session, type your code into this script, adding\n",
    "comments and then run them in R using Ctrl+R. If you make mistakes,\n",
    "correct the script and run the code again. This way you will end up with\n",
    "a complete neat version of the commands for your analysis.\n",
    "\n",
    "Add code to your script to load the genome size data into R, assigning\n",
    "the object name <span>genome</span> and use <span>str(genome)</span> to\n",
    "check it has loaded correctly.\n",
    "\n",
    "Using $t$ tests\n",
    "---------------\n",
    "\n",
    "The $t$ test is used to compare the mean of a sample to another value:\n",
    "either some reference point (Is the mean different from 5?) or another\n",
    "mean (Does the mean of A differ from the mean of B?). If you have a\n",
    "factor with two levels then a $t$ test is a good way to compare the\n",
    "means of those samples. If you have more than two levels, then you have\n",
    "a problem: as the number of levels ($n$) increases, the number of\n",
    "possible comparisons between pairs of levels ($p$) increases very\n",
    "rapidly[^2]. Making these many comparisons using something like a t-test\n",
    "is a problem because we neglect the covariance among measures and\n",
    "inflate the chance of falsely rejecting at least one hypothesis (Type I\n",
    "error – recall hypothesis testing in chapter \\[ch:ExpDesign\\]).\n",
    "\n",
    "\\[label distance=1.2cm\\] (2) at (0,0)\n",
    "\\[label=below:<span>$n=2, p=1$</span>\\]; (3) at (3,0)\n",
    "\\[label=below:<span>$n=3, p=3$</span>\\]; (4) at (6,0)\n",
    "\\[label=below:<span>$n=4, p=6$</span>\\]; (5) at (9,0)\n",
    "\\[label=below:<span>$n=5, p=10$</span>\\]; (6) at\n",
    "(12,0)\\[label=below:<span>$n=6, p=15$</span>\\]; (2) ++(0:1) node (a)\n",
    "<span>$a$</span>; (2) ++(180:1) node (b) <span>$b$</span>; (a) – (b);\n",
    "(3) ++(0:1) node (a) <span>$a$</span>; (3) ++(120:1) node (b)\n",
    "<span>$b$</span>; (3) ++(240:1) node (c) <span>$c$</span>; (a) – (b) –\n",
    "(c) – (a); (4) ++(0:1) node (a) <span>$a$</span>; (4) ++(90:1) node (b)\n",
    "<span>$b$</span>; (4) ++(180:1) node (c) <span>$c$</span>; (4) ++(270:1)\n",
    "node (d) <span>$d$</span>; (a) – (b) – (c) – (d) – (a) ; (a) – (c) (d) –\n",
    "(b) ; (5) ++(0:1) node (a) <span>$a$</span>; (5) ++(72:1) node (b)\n",
    "<span>$b$</span>; (5) ++(144:1) node (c) <span>$c$</span>; (5) ++(216:1)\n",
    "node (d) <span>$d$</span>; (5) ++(288:1) node (e) <span>$e$</span>; (a)\n",
    "– (b) – (c) – (d) – (e) – (a) ; (a) – (c) (d) – (b) (a) – (d) ; (e) –\n",
    "(b) (e) – (c) ; (6) ++(0:1) node (a) <span>$a$</span>; (6) ++(60:1) node\n",
    "(b) <span>$b$</span>; (6) ++(120:1) node (c) <span>$c$</span>; (6)\n",
    "++(180:1) node (d) <span>$d$</span>; (6) ++(240:1) node (e)\n",
    "<span>$e$</span>; (6) ++(300:1) node (f) <span>$f$</span>; (a) – (b) –\n",
    "(c) – (d) – (e) – (f) – (a) ; (a) – (c) (d) – (b) (a) – (d) ; (e) – (b)\n",
    "(e) – (c) (a) – (f) ; (d) – (f) (b) – (f) (a) – (f) ; (c) – (f) (a) –\n",
    "(e) ;\n",
    "\n",
    "The basic idea behind the $t$ test is to divide the difference between\n",
    "two values by a measure of how much uncertainty there is about the size\n",
    "of that difference. The measure of uncertainty used is the *standard\n",
    "error*.\n",
    "\n",
    "$$t=\\frac{\\textrm{difference between values}}{\\textrm{standard error}}$$\n",
    "\n",
    "When there is no difference between the values $t$ will be zero, but\n",
    "with big differences and/or small errors, will be larger. The $t$\n",
    "distribution below shows how commonly different values of $t$ are found\n",
    "under the null hypothesis.\n",
    "\n",
    "![image](tvals.pdf){width=\"70.00000%\"}\n",
    "\n",
    "Some points about the plot above:\n",
    "\n",
    "The null hypothesis is that there is no difference between the values\n",
    "but because we only estimate the values from samples, differences will\n",
    "creep in by chance.\n",
    "\n",
    "Mostly these differences will be small — hence the peak in the middle —\n",
    "but sometimes the differences will be large and the errors will be\n",
    "small.\n",
    "\n",
    "95% of the area under the curves are between these two sets of vertical\n",
    "lines. Values of $t$ more extreme than this will only occur 1 in 20\n",
    "times or with a probability ($p$) of 0.05.\n",
    "\n",
    "The means of small samples are more easily influenced by extreme values\n",
    "and so produce extreme $t$ values more frequently. This is why the red\n",
    "curve above for smaller samples is more flattened out and why the 95%\n",
    "limits are more spread out.\n",
    "\n",
    "One sample $t$ tests\n",
    "--------------------\n",
    "\n",
    "In the simplest example, a $t$ test can be used to test whether the mean\n",
    "of a sample is different from a specific value. For example:\n",
    "\n",
    "Is the ozone in a set of air samples above the legal limit?\n",
    "\n",
    "Is the change in a set of patient’ weights different from zero?\n",
    "\n",
    "Is the mean genome size for Odonata smaller than 1.25 pg, which is the\n",
    "average for insects [\\[see\n",
    "here\\]](http://www.genomesize.com/statistics.php?stats=insects)?\n",
    "\n",
    "Oh look! We can test that last one…\n",
    "\n",
    "To calculate $t$, we need that observed difference and then the standard\n",
    "error of the difference between the mean of our sample and the known\n",
    "value. This is calculated using the *variance* and the *sample size*\n",
    "($n$) of the sample ($s$).\n",
    "\n",
    "$$se_s = \\sqrt{\\frac{\\textrm{var}(s)}{n}}$$\n",
    "\n",
    "This simple equation trades off variance — high variance in the data\n",
    "gives higher uncertainty about the location of the mean — and sample\n",
    "size – more data gives more certainty. So, *low variance* and *large\n",
    "datasets* have *small* standard errors; *high variance* and *small\n",
    "datasets* have *large* standard errors. Variance is calculated using\n",
    "sums of squares and so the square root is needed to give a standard\n",
    "error in the same units as the mean.\n",
    "\n",
    "So, all we need are three values calculated from the data: mean,\n",
    "variance and the number of data points and we can calculate $t$. R can\n",
    "do this for us:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate the three values from the data:\n",
    "> mean.gs <- mean(genome$GenomeSize)\n",
    "> print(mean.gs)\n",
    "[1] 1.014\n",
    "> var.gs <- var(genome$GenomeSize)\n",
    "> print(var.gs)\n",
    "[1] 0.1397\n",
    "> n.gs <- length(genome$GenomeSize)\n",
    "> print(n.gs)\n",
    "[1] 100\n",
    "# get the difference\n",
    "> diff <- mean.gs - 1.25\n",
    "> print(diff)\n",
    "[1] -0.2357\n",
    "# get the standard error\n",
    "> se.gs <- sqrt(var.gs/n.gs)\n",
    "> print(se.gs)\n",
    "[1] 0.03738\n",
    "# get the t value\n",
    "> t.gs <- diff/se.gs\n",
    "> print(t.gs)\n",
    "[1] -6.306"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Copy and paste the code above into your script in R and run it. Read\n",
    "through the code and make sure you understand the steps.\n",
    "\n",
    "This is a big $t$ value — values this extreme don’t even appear on the\n",
    "graph above — so we would conclude that the mean genome size for Odonata\n",
    "is different from the average for insects.\n",
    "\n",
    "We can do this more easily and get some more information using the\n",
    "function <span>t.test</span>. The null hypothesis can be set using the\n",
    "option (sometimes called a function *argument*) <span>mu</span> — the\n",
    "Greek letter $\\mu$ is often used to refer to a mean:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> t.test(genome$GenomeSize, mu = 1.25)\n",
    " \n",
    "\tOne Sample t-test\n",
    " \n",
    " data:  genome$GenomeSize \n",
    " t = -6.306, df = 99, p-value = 8.034e-09\n",
    " alternative hypothesis: true mean is not equal to 1.25 \n",
    " 95 percent confidence interval:\n",
    "  0.9401 1.0885 \n",
    " sample estimates:\n",
    " mean of x \n",
    "     1.014 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This confirms the values we calculated by hand and adds a $p$ value. The\n",
    "output also gives the degrees of freedom. This is something we will come\n",
    "back to later, but the degrees of freedom are basically the number of\n",
    "data points minus the number of estimated parameters, which in this case\n",
    "is one mean.\n",
    "\n",
    "The output also gives a confidence interval for the observed mean. The\n",
    "mean is the best estimate of the population mean given our sample of\n",
    "species of Odonata, but the actual mean for the order could be bigger or\n",
    "smaller. The confidence interval tells us the region in which we are 95%\n",
    "confident that this actual mean lies.\n",
    "\n",
    "It is calculated using the $t$ distribution. Remember that $t$ is a\n",
    "difference divided by a standard error; if we multiply $t$ by a standard\n",
    "error, we get back to a difference. If we pick a pair of $t$ values that\n",
    "contain the middle 95% of the $t$ distribution, as in the plot on page\n",
    "2, then we can multiply that by the standard error from the data to get\n",
    "a range above and below the mean. If we sampled lots of sets of 100\n",
    "species of Odonata, we expect 95% of the observed means to lie inside\n",
    "this range. The code below shows the calculation of the confidence\n",
    "interval for the test above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find the edges of the middle 95% of a t distribution with 99 df\n",
    "# (quantiles of the t distribution, so qt)\n",
    "> tlim <- qt(c(0.025,0.975), df = 99)\n",
    "> print(tlim)\n",
    "[1] -1.984 1.984\n",
    "# use the mean and standard error from above to get a confidence interval\n",
    "> mean.gs + tlim * se.gs\n",
    "[1] 0.9401 1.0885"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Using the <span>t.test</span> code above as an template, test whether\n",
    "the body weight (in grams) of Odonata is different from the average[^3]\n",
    "for arthropods of 0.045 grams.\n",
    "\n",
    "Two sample $t$ tests\n",
    "--------------------\n",
    "\n",
    "It is more common to use a $t$ test to compare the means of two samples.\n",
    "This includes questions like:\n",
    "\n",
    "Do two rivers have the same concentration of a pollutant?\n",
    "\n",
    "Do chemical A and chemical B cause different rates of mutation?\n",
    "\n",
    "Do damselflies and dragonflies have different genome sizes?\n",
    "\n",
    "The main difference here is that with a one sample $t$ test, we assume\n",
    "that one of the means is known exactly: the only error is in the single\n",
    "sample. With a two sample test, we are comparing two means estimated\n",
    "from samples and both contain error. The graph below illustrates this:\n",
    "\n",
    "![image](twosampleIllustration.pdf){width=\"90.00000%\"}\n",
    "\n",
    "The vertical lines show the mean (solid lines) and one standard error to\n",
    "each side (dashed lines). The red mean is the same in both cases, but\n",
    "the second graph shows that this is also estimated from a sample with\n",
    "error: the difference in the means looks less convincing and we’d expect\n",
    "a smaller $t$ value. The $t$ tests in for these two graphs confirm this:\n",
    "\n",
    "The mean for blue is significantly different from 16.74 (mean=14.98,\n",
    "se=0.38, df=59, $t$=-4.65, $p$=0.00002).\n",
    "\n",
    "The means of blue and red are significantly different (blue: mean=14.98,\n",
    "se=0.38; red: mean=16.74, se=0.42; df=118, $t$=-3.13, $p$=0.002)\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Have a close look at the previous two statements. This shows the kind of\n",
    "detail needed when reporting the results of $t$ tests. The following is\n",
    "*not* acceptable: The means of blue and red are significantly different\n",
    "($p=0.002$).\n",
    "\n",
    "So, with two samples, we shouldn’t be so confident about the difference\n",
    "between the values — it should have a higher standard error. We can do\n",
    "this simply by combining the variance and sample size for the two\n",
    "samples ($a$ and $b$) into the calculation:\n",
    "\n",
    "$$se_{a-b}= \\sqrt{\\frac{\\textrm{var}(a)}{n_a} + \\frac{\\textrm{var}(b)}{n_b}}$$\n",
    "\n",
    "We’ll use a $t$ test to address that last question: are the genome sizes\n",
    "of Anisoptera and Zygoptera different? First, we’ll do this by hand.\n",
    "We’ll use a really handy function <span>tapply(X, INDEX, FUN)</span> to\n",
    "quickly find the values for the two groups: it takes some values (X),\n",
    "splits those values into groups based on a factor (INDEX) and runs each\n",
    "group through another function (FUN)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate the three values from the data\n",
    "> mean.gs <- tapply(X = genome$GenomeSize, INDEX = genome$Suborder, FUN = mean)\n",
    "> print(mean.gs)\n",
    "Anisoptera \tZygoptera\n",
    "1.018\t\t\t\t1.012\n",
    "> var.gs <- tapply(X = genome$GenomeSize, INDEX = genome$Suborder, FUN = var)\n",
    "> print(var.gs)\n",
    "Anisoptera \tZygoptera\n",
    "0.18458 \t\t0.06946\n",
    "> n.gs <- tapply(X = genome$GenomeSize, INDEX = genome$Suborder, FUN = length)\n",
    "> print(n.gs)\n",
    "Anisoptera \tZygoptera\n",
    "38\t\t\t\t\t62\n",
    "\n",
    "# get the difference\n",
    "> diff <- mean.gs[1] - mean.gs[2]\n",
    "> print(diff)\n",
    "Anisoptera\n",
    "-0.006647\n",
    "\n",
    "# get the standard error of the difference\n",
    "> se.gs <- sqrt((var.gs[1]/n.gs[1]) + (var.gs[2]/n.gs[2]))\n",
    "> print(se.gs)\n",
    "Anisoptera\n",
    "0.06932\n",
    "\n",
    "# get the t value\n",
    "> t.gs <- diff/se.gs\n",
    "> print(t.gs)\n",
    "Anisoptera\n",
    "\n",
    "-0.09589"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Type the code above into your script in R and run it. Again, read\n",
    "through the code and make sure you understand the steps.\n",
    "\n",
    "The <span>t.test</span> function automates this all for us, and we can\n",
    "use a formula (see Chapter \\[ch:ExpDesign\\]) to get a test between the\n",
    "two suborders."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ">  t.test(GenomeSize ~ Suborder, data = genome)\n",
    " \n",
    " \tWelch Two Sample t-test\n",
    " \n",
    " data:  GenomeSize by Suborder \n",
    " t = -0.0959, df = 98, p-value = 0.9238\n",
    " alternative hypothesis: true difference in means is not equal to 0 \n",
    " 95 percent confidence interval:\n",
    "  -0.1442  0.1309 \n",
    " sample estimates:\n",
    " mean in group Anisoptera  mean in group Zygoptera \n",
    "                    1.012                    1.018 \n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output looks very similar to the one sample test, except that the\n",
    "output now gives two estimated means, rather than one and it reports the\n",
    "$p$ value for the calculated $t$ value.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Add this to your script and run it.\n",
    "\n",
    "Copy and modify this in your script to test whether the body weight of\n",
    "the two suborders are different.\n",
    "\n",
    "$F$ tests for equal variance\n",
    "----------------------------\n",
    "\n",
    "$F$ tests are used to compare the variances of two samples or\n",
    "populations. You will most prominently see them in analysis of variance\n",
    "(ANOVA), to test the hypothesis that the means of a given set of\n",
    "normally distributed populations all have the same variance.\n",
    "\n",
    "Let’s use our genomesize datast to have a look at F-tests as well.\n",
    "Ideally, the $t$ test should be used with data that are: i) *relatively\n",
    "normally distributed*, so that means can be estimated sensibly; and ii)\n",
    "have *similar variances*. We’ll deal with the similar variance question\n",
    "here using an $F$ test for equal variances.\n",
    "\n",
    "First, let’s visualize the data. As I hope you’ve already noticed, this\n",
    "session has been neglecting one very important part of analysis —\n",
    "*plotting the data*. We are going to compare two plots, so it helps to\n",
    "have them side by side in the same window. We can use the function\n",
    "<span>par</span> to change a set of options called graphics parameters\n",
    "to get R to do this. The option to change is <span>mfrow</span>. This\n",
    "sets a graphics window to include *m*ultiple *f*igures and we need to\n",
    "tell R the number of rows and columns to divide the window into:\n",
    "<span>par(mfrow=c(1,2))</span>.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Copy <span>par(mfrow=c(1,2))</span> into your script, add a comment and\n",
    "run it.\n",
    "\n",
    "Using your (rapidly improving!) R skills, create a boxplot comparing the\n",
    "genome sizes of the two suborders.\n",
    "\n",
    "Add another boxplot beside it comparing the body weight of the two\n",
    "suborders.\n",
    "\n",
    "It should look like this:\n",
    "\n",
    "![image](bxp.pdf){width=\"90.00000%\"}\n",
    "\n",
    "The distribution of the test statistic $F$ is simply the ratio of the\n",
    "variances for sample $a$ and $b$: $\\textrm{var}(a)/\\textrm{var}(b)$. If\n",
    "the two variances are the same then $F=1$; if $\\textrm{var}(a) > \n",
    "\\textrm{var}(b)$ then $F > 1$; and if $\\textrm{var}(a) < \n",
    "\\textrm{var}(b)$ then $F < 1$ (\\[fig:fdist\\]).\n",
    "\n",
    "![The F distribution. The two vertical blue lines show the edges of the\n",
    "central 95% of the area of the curve: if the two samples are drawn at\n",
    "random from a population with the same variance then values of\n",
    "$F < 0.482$ or $> 2.074$ are observed fewer than 1 time in 20\n",
    "($p\\le0.05$ ). The shape of the $F$ distribution changes depending on\n",
    "the amount of data in each of the two samples but will always be\n",
    "centered near 1 and with a tail to the right (right-skewed). Note that\n",
    "the F-distribution arises as the ratio of two appropriately scaled\n",
    "<span> *chi-square distributed variates*</span>, because, as we saw\n",
    "above, variances should be chi-square distributed.<span\n",
    "data-label=\"fig:fdist\"></span>](fvals.pdf){width=\"70.00000%\"}\n",
    "\n",
    "[^4]\n",
    "\n",
    "We can use R to calculate $F$ for the variance in genome size in each of\n",
    "the two suborders. We calculated the variance for the $t$ test above, so\n",
    "we can just do this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> var.gs[1]/var.gs[2]\n",
    "Anisoptera\n",
    "\t\t2.657"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That’s quite a big $F$ value and we can use the function\n",
    "<span>var.test</span> to do all the calculations for us and give us the\n",
    "actual $p$ value:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> var.test(GenomeSize ~ Suborder, data = genome) \n",
    " \tF test to compare two variances\n",
    " \n",
    " data:  GenomeSize by Suborder \n",
    " F = 2.657, num df = 61, denom df = 37, p-value = 0.001946\n",
    " alternative hypothesis: true ratio of variances is not equal to 1 \n",
    " 95 percent confidence interval:\n",
    "  1.449 4.671 \n",
    " sample estimates:\n",
    " ratio of variances \n",
    "              2.657 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It produces the same value that we calculated by hand and shows that, if\n",
    "the two samples are drawn from populations with the same variance, an\n",
    "$F$ value this extreme will only be observed roughly 1 time in 500\n",
    "($1/0.00195 \\approx 500$) .\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Open a new empty script called <span>FTests.R</span>.\n",
    "\n",
    "In this write your script to test whether the variances in the body\n",
    "weight of the two suborders from the <span>GenomSize</span> dataset are\n",
    "different.\n",
    "\n",
    "There are clearly problems with the variance in both examples. The next\n",
    "two sections present ways to address these kinds of problems.\n",
    "\n",
    "$t$ tests revisited\n",
    "-------------------\n",
    "\n",
    "The first thing to say is that R is aware of the problem with the\n",
    "variance. If you look back at the output from the previous $t$ tests,\n",
    "you will see that the degrees of freedom vary a lot. We have 100\n",
    "observations and – after subtracting one for each mean we calculate —\n",
    "our degrees of freedom should be either 99 (one sample test) or 98 (two\n",
    "sample test). What we actually see are smaller numbers, with the\n",
    "smallest being <span>df = 60.503</span> for the two sample test of body\n",
    "weight.\n",
    "\n",
    "The explanation is that R is applying a *penalty* to the degrees of\n",
    "freedom to account for differences in variance. With fewer degrees for\n",
    "freedom, more extreme $t$ values are more likely and so it is harder to\n",
    "find significant results. This doesn’t mean we can forget about checking\n",
    "the variances or plotting the data!\n",
    "\n",
    "In this case, we can also apply a transformation to the data in order to\n",
    "make the variances more equal. Forgetting the wings and assuming Odonata\n",
    "are shaped like a box, the model in the graph below shows how volume\n",
    "changes with length: equal changes in length do not lead to equal\n",
    "changes in volume and longer species will have a disproportionately\n",
    "large volume. This is a classic feature of morphological data known as\n",
    "allometric scaling and we’ll look at it again in Chapter \\[ch:ANOVA\\].\n",
    "In the meantime, a log transformation will turn body weight from a\n",
    "skewed distribution to a more normal distribution.\n",
    "\n",
    "![image](transform.pdf){width=\"\\textwidth\"}\n",
    "\n",
    "$\\log_e$ body weight as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> genome$logBodyWeight <- log(genome$BodyWeight)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Copy the line into your script and run it.\n",
    "\n",
    "Now write three lines of code to get a boxplot of $\\log_e$ body weight\n",
    "and then run a variance test and $t$ test on the differences in $\\log_e$\n",
    "body weight between suborders.\n",
    "\n",
    "This gives a much clearer result — the variances are almost identical\n",
    "and the differences between the suborders are much more cleanly tested.\n",
    "\n",
    "Non-parametric tests\n",
    "--------------------\n",
    "\n",
    "What happens if there isn’t a convenient transformation for the variable\n",
    "that gives roughly constant variation and equal variance? In a\n",
    "parametric test, like the $t$ and $F$ test above, we use parameters\n",
    "(mean and variance) to describe the data, assume these describe the data\n",
    "well and then just use these parameters to run the test. If these\n",
    "assumptions don’t seem very sound, the non-parametric tests provide a\n",
    "way of using the ranks of the data to test for differences. They aren’t\n",
    "as powerful — they are less likely to reveal significant differences —\n",
    "but they are more robust. The most commonly used alternative is the\n",
    "Wilcoxon test, which uses the function <span>wilcox.test</span> in R.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Using <span>wilcox.test</span> as a replacement for <span>t.test</span>,\n",
    "repeat the one and two sample $t$ test for genome size and body weight.\n",
    "\n",
    "Compare the two results.\n",
    "\n",
    "Repeat the same with the predator and prey body mass data from the\n",
    "plotting and visualization chapter in SilBioComp.pdf – check how\n",
    "different the results are when using $t$ vs. Wilcoxon test.\n",
    "\n",
    "Linear Models: Regression {#ch:regress}\n",
    "=========================\n",
    "\n",
    "Aims of this chapter[^5]:\n",
    "\n",
    "More functions for plotting data and models.\n",
    "\n",
    "Calculating correlation coefficients.\n",
    "\n",
    "Fitting a regression model and significance testing.\n",
    "\n",
    "Using diagnostic plots to assess model suitability.\n",
    "\n",
    "As with the previous chapter, we’ll start with creating a new blank\n",
    "script for you to fill in during the practical. We’ll also be using the\n",
    "genome size data again, so:\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Open R and change to the <span>code</span> directory.\n",
    "\n",
    "Create a new blank script called ‘Regression.R’ and add some\n",
    "introductory comments.\n",
    "\n",
    "Add code to your script to load the genome size data into R and check\n",
    "it.\n",
    "\n",
    "Exploring the data\n",
    "------------------\n",
    "\n",
    "In previous chapters we used <span>plot</span> to create a scatterplot\n",
    "between two variables. If you have a set of variables to explore,\n",
    "writing code for each plot is tiresome, so R provides a the function\n",
    "<span>pairs</span>, which creates a grid of scatterplots between each\n",
    "pair of variables. All it needs is a dataset.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Add <span>pairs(genome, col=genome\\$Suborder)</span> into your script\n",
    "and run the code.\n",
    "\n",
    "The result is messy! There are far too many variables in\n",
    "<span>genome</span> for this to be useful. We need to cut down the data\n",
    "to fewer variables. In Chapter \\[ch:ExpDesign\\], we used indices to\n",
    "select colours; here, we can use indices to select columns from the data\n",
    "frame. This again uses square brackets (<span>x\\[ \\]</span>), but a data\n",
    "frame has two dimensions, rows and columns, so you need to provide an\n",
    "index for each dimension, separated by commas. If an index is left\n",
    "blank, then all of that dimension (i.e. all rows or columns) are\n",
    "selected. Try the following to re-acquaint yourself to access data frame\n",
    "content using indices:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a small data frame:\n",
    "> dat <- data.frame(A = c(\"a\", \"b\", \"c\", \"d\", \"e\"), B = c(1, 2, 3, 4, 5))\n",
    "> dat[1, ] # select row 1 (all columns selected)\n",
    "\tA B\n",
    "1 a 1\n",
    "\n",
    "> dat[, 2] # select column 2 (all rows selected)\n",
    "[1] 1 2 3 4 5\n",
    "> dat[2, 1] # select row 2, column 1\n",
    "[1] \"b\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let’s get started with the actual analysis. We will look at five key\n",
    "variables: genome size, body weight, total length, forewing length and\n",
    "forewing area. If you look at the output of <span>str(genome)</span>,\n",
    "you’ll see that these are in columns 4, 7, 8, 12 and 14. We can record\n",
    "the indices of these columns and use this to select the data in the\n",
    "pairs plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "morpho <- c(4, 7, 8, 12, 14)\n",
    "pairs(genome[, morpho], col = genome$Suborder)\t"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Add the code above to your script and run it\n",
    "\n",
    "The <span>pairs</span> plot should give you something like the plot\n",
    "below:\n",
    "\n",
    "![image](pairs.pdf){width=\"\\textwidth\"}\n",
    "\n",
    "Each scatterplot is shown twice, with the variables swapping between the\n",
    "$x$ and $y$ axes. You can see immediately that the relationships between\n",
    "the four morphological measurements and genome size are fairly scattered\n",
    "but that the plots comparing morphology show much clearer relationships.\n",
    "\n",
    "Correlations\n",
    "------------\n",
    "\n",
    "One way of summarising how close strong the relationship between these\n",
    "variables are is to calculate a correlation coefficient. Pearson\n",
    "correlations look at the difference of each point from the mean of each\n",
    "variable (and since it uses means, it is a parametric statistic).\n",
    "\n",
    "It is calculated using of the differences from the mean on each axis.\n",
    "The key calculation is — for each point – to get the product of the\n",
    "differences on each axis and add them up. If the points are mostly top\n",
    "left ($-x$, $y$) or bottom right ($x$, $-y$) then these products are\n",
    "mostly negative ($-xy$); if the points are mostly top right ($x$, $y$)\n",
    "or bottom left ($-x$, $-y$) then the products are mostly positive\n",
    "($xy$).\n",
    "\n",
    "![image](corr.pdf){width=\"\\textwidth\"}\n",
    "\n",
    "The plots above show three clear cases where all the values of $xy$ are\n",
    "negative or positive or where both are present and sum to zero. The\n",
    "Pearson correlation coefficient simply scales these sums of $xy$ to be\n",
    "between -1 (perfectly negatively correlated) and 1 (perfectly positively\n",
    "correlated) via zero (no correlation).\n",
    "\n",
    "We will use two functions to look at correlations. The first is <span>\n",
    "cor</span>, which can calculate correlations between pairs of variables,\n",
    "so is a good partner for <span>pairs</span> plots. The second is\n",
    "<span>cor.test</span>, which can only compare a single pair of\n",
    "variables, but uses a $t$ test to assess whether the correlation is\n",
    "significant.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Try the following (and include it in your R script file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> cor(genome[, morpho], use = \"pairwise\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You should see a correlation matrix. Then,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> cor.test(genome$GenomeSize, genome$TotalLength, use = \"pairwise\")\t\n",
    "\n",
    "Pearson's product-moment correlation\n",
    "\n",
    "data: genome$GenomeSize and genome$TotalLength\n",
    "t = 3.551, df = 96, p-value = 0.0005972\n",
    "\n",
    "alternative hypothesis: true correlation is not equal to 0 \n",
    "95 percent confidence interval:\n",
    "  0.1526 0.5050 \n",
    "sample estimates:\n",
    "    cor \n",
    " 0.3407   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The <span>use=’pairwise’</span> in the above tells R to omit\n",
    "observations with missing data and use complete pairs of observations.\n",
    "The first function confirms our impressions from the graphs: the\n",
    "correlations between genome size and morphology are positive but\n",
    "comparatively weak and the correlations between morphological\n",
    "measurements are positive and very strong (i.e. close to 1). The\n",
    "correlation test tells us that genome size and body length are\n",
    "positively correlated (r=0.34, $t$ = 3.5507, df = 96, $p$ = 0.0006).\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Again, remember this example when reporting correlations!\n",
    "\n",
    "Transformations and allometric scaling\n",
    "--------------------------------------\n",
    "\n",
    "There is one problem with the correlations above: <span>*correlations\n",
    "assume a straight line relationship*</span>. Some of the scatterplots\n",
    "above are fairly straight but there are some strongly curved\n",
    "relationships. This is due to the allometric scaling mentioned in\n",
    "Chapter \\[ch:ExpDesign\\]: two of the variables are in linear units\n",
    "(total and forewing length), one is in squared units (forewing area) and\n",
    "one in cubic units (body weight, which is approximately volume).\n",
    "\n",
    "The relationships between these variables can be described using a power\n",
    "law: $y = ax^b$. Fortunately, if we log transform this equation, we get\n",
    "$\\log(y) = \\log(a) + b \\log(x)$. This is the equation of a straight line\n",
    "($y=a+bx$), so we should be able to make these plots straighter by\n",
    "logging both axes. We saw in Chapter \\[ch:ExpDesign\\] that we can create\n",
    "a new logged variable in the data frame like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> genome$logGS <- log(genome$GenomeSize)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Using this line as a template, create a new logged version of the five\n",
    "variables listed above.\n",
    "\n",
    "Using <span>str</span>, work out which column numbers the logged\n",
    "variables are and create a new variable called <span>logmorpho</span>\n",
    "containing these numbers.\n",
    "\n",
    "Copy the <span>pairs</span> and <span>cor</span> test from earlier in\n",
    "your script and modify them to run these functions for the columns given\n",
    "in <span>logmorpho</span>.\n",
    "\n",
    "The correlations should give the following output:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> cor(genome[, logmorpho], use = \"pairwise\")\n",
    "\n",
    "         logGS   logBW  logTL  logFL   logFA\n",
    " logGS 1.00000 0.08406 0.2224 0.1150 0.06808\n",
    " logBW 0.08406 1.00000 0.8892 0.9456 0.94996\n",
    " logTL 0.22244 0.88919 1.0000 0.9158 0.86207\n",
    " logFL 0.11500 0.94565 0.9158 1.0000 0.97916\n",
    " logFA 0.06808 0.94996 0.8621 0.9792 1.00000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The scatterplots should look like this and show that logging the data\n",
    "has very successfully removed allometric scaling effects in the data:\n",
    "\n",
    "![image](pairsLog.pdf){width=\"\\textwidth\"}\n",
    "\n",
    "Regression\n",
    "----------\n",
    "\n",
    "We’ll now look at fitting the first linear model of this course, to\n",
    "explore whether log genome size explain log body weight. The first thing\n",
    "to do is to plot the data:\n",
    "\n",
    "![image](gsVbw.pdf){width=\"50.00000%\"}\n",
    "\n",
    "It is clear that the two suborders have very different relationships: to\n",
    "begin with we will look at dragonflies (Anisoptera). We will calculate\n",
    "two linear models:\n",
    "\n",
    "This is the simplest linear model: nothing is going on and the response\n",
    "variable just has variation around the mean: $y = \\beta_1$. This is\n",
    "written as an R formula as <span>y \\~ 1</span>.\n",
    "\n",
    "This models a straight line relationship between the response variable\n",
    "and a continuous explanatory variable: $y= \\beta_1 + \\beta_{2}x$.\n",
    "\n",
    "The code below fits these two models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> nullModelDragon <- lm(logBW ~ 1, data = genome, subset = Suborder == \n",
    "\"Anisoptera\")\n",
    "> genomeSizeModelDragon <- lm(logBW ~ logGS, data = genome, subset = \n",
    "Suborder == \"Anisoptera\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Note the long names for the models. Short names are easier to type but\n",
    "calling R objects names like <span>mod1</span>, <span>mod2</span>,\n",
    "<span> xxx</span> swiftly get confusing!\n",
    "\n",
    "Enter these models into your script and run them.\n",
    "\n",
    "Now we want to look at the output of the model. Remember from the\n",
    "lecture that a model has <span>*coefficients*</span> (the $\\beta$ values\n",
    "in the equation of the model) and <span>*terms*</span> which are the\n",
    "explanatory variables in the model. We’ll look at the\n",
    "<span>*coefficients*</span> first:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> summary(genomeSizeModelDragon) \n",
    " Call:\n",
    " lm(formula = logBW ~ logGS, data = genome, subset = Suborder == \n",
    "     \"Anisoptera\")\n",
    " \n",
    " Residuals:\n",
    "    Min     1Q Median     3Q    Max \n",
    " -1.324 -0.612  0.097  0.519  1.324 \n",
    " \n",
    " Coefficients:\n",
    "             Estimate Std. Error t value Pr(>|t|)    \n",
    " (Intercept)  -2.3995     0.0908  -26.41  < 2e-16 ***\n",
    " logGS         1.0052     0.2398    4.19  9.5e-05 ***\n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 \n",
    " \n",
    " Residual standard error: 0.697 on 58 degrees of freedom\n",
    "   (2 observations deleted due to missingness)\n",
    " Multiple R-squared: 0.233,\tAdjusted R-squared: 0.219 \n",
    " F-statistic: 17.6 on 1 and 58 DF,  p-value: 9.54e-05  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is a lot of information there: the model description (‘<span>\n",
    "Call</span>’), a summary of the residuals, a table of coefficients and\n",
    "then information on residual standard error, r squared and an $F$ test.\n",
    "All of these will become clearer during this course — for the moment,\n",
    "concentrate on the coefficients table.\n",
    "\n",
    "There are two rows in the coefficient table, one for each coefficient in\n",
    "$y=\\beta_1 + \\beta_2x$ — these are the intercept and the slope of the\n",
    "line. The rest the details on each row are a $t$ test of whether the\n",
    "slope and intercept are significantly different from zero.\n",
    "\n",
    "Now we will look at the <span>*terms*</span> of the model using the\n",
    "<span>anova</span> function. We will have a proper look at ANOVA\n",
    "(Analysis of Variance) in chapter \\[ch:ANOVA\\]. Meanwhile, for our\n",
    "current purposes, all you need to know is that ANOVA tests how much\n",
    "variation in the response variable is explained by each explanatory\n",
    "variable. We only have one variable and so there is only one row in the\n",
    "output:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> anova(genomeSizeModelDragon)\n",
    "\n",
    " Analysis of Variance Table\n",
    " \n",
    " Response: logBW\n",
    "           Df Sum Sq Mean Sq F value  Pr(>F)    \n",
    " logGS      1   8.53    8.53    17.6 9.5e-05 ***\n",
    " Residuals 58  28.14    0.49                    \n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This table is comparing the variation in log body weight explained by\n",
    "log genome size to the total variation in log body weight. We are\n",
    "interested in how much smaller the residuals are for the genome size\n",
    "model than the null model. Graphically, how much shorter are the red\n",
    "residuals than the blue residuals:\n",
    "\n",
    "![image](regResid.pdf){width=\"80.00000%\"}\n",
    "\n",
    "We can get the sums of the squares of these residuals from the two\n",
    "models using the function <span>resid</span>, and then square them and\n",
    "add them up:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> sum(resid(nullModelDragon) ^ 2)\n",
    " [1] 36.67\n",
    " \n",
    "> sum(resid(genomeSizeModelDragon) ^ 2)\n",
    " [1] 28.14"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we have five columns in the table:\n",
    "\n",
    "This shows the degrees of freedom. Each fitted parameter/coefficient\n",
    "takes up a degree of freedom from the total sample size, and the left\n",
    "over are the residuals degree of freedom. In this case, genome size adds\n",
    "a slope (compare the null model $y=\\beta_1$ and this model\n",
    "$y=\\beta_1 + \\beta_2x$ — there is one more $\\beta$).\n",
    "\n",
    "This shows sums of squares. The bottom line is the residual sum of\n",
    "squares for the model and the one above is the variation explained by\n",
    "genome size. Using the two values from above, the sums of square\n",
    "residuals for the null model are 36.67. In the genome size model, the\n",
    "sum of square residuals are 28.14 and so $36.67-28.14=8.53$ units of\n",
    "variance have been explained by this model.\n",
    "\n",
    "These are just the Sum Sq (Sum of Squares) values divided by the degrees\n",
    "of freedom. The idea behind this is simple: if we explain lots of\n",
    "variation with one coefficient, that is good (the null model), and if we\n",
    "explain a small amount of variation with a loss of degree of freedom (by\n",
    "adding and then estimating more parameters), then that is bad.\n",
    "\n",
    "This is the ratio of the Mean Sq for the variable and the residual Mean\n",
    "Sq. This is used to test whether the explained variation is large or\n",
    "small.\n",
    "\n",
    "This is a $p$ value — the probability of the variable explaining this\n",
    "much variance by chance.\n",
    "\n",
    "In this case, it is clear that genome size explains a significant\n",
    "variation in body weight.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Include the <span>summary</span> and <span>anova</span> commands for\n",
    "<span> genomeSizeModelDragon</span> in your script, run them and check\n",
    "you are happy with the output.\n",
    "\n",
    "Using this code as a template, create a new model called <span>\n",
    "genomeSizeModelDamsel</span> that fits log body weight as a function of\n",
    "log genome size for damselflies.\n",
    "\n",
    "Write and run code to get the <span>summary</span> and\n",
    "<span>anova</span> tables for this new model.\n",
    "\n",
    "Plotting the model\n",
    "------------------\n",
    "\n",
    "Now we can plot the data and add lines to show the models. For simple\n",
    "regression models, we can use the function\n",
    "<span>abline(modelName)</span> to add a line based on the model.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "You already know how to create and customise scatterplots from previous\n",
    "chapters. Create a plot of log body weight as a function of log genome\n",
    "size, picking your favourite colours for the points.\n",
    "\n",
    "Use <span>abline</span> to add a line for each model and use the <span>\n",
    "col</span> option in the function to colour each line to match the\n",
    "points. For example: <span>abline(genomeSizeModelDragon,\n",
    "col=’red’)</span>.\n",
    "\n",
    "You should get something like Figure \\[fig:GenoRegModels\\].\n",
    "\n",
    "![Linear regression models fitted to the body weight vs. genome size to\n",
    "the Dragonfly (red) and Damselfly (blue) subsets of the data.<span\n",
    "data-label=\"fig:GenoRegModels\"></span>](GenoRegModels.pdf){width=\"50.00000%\"}\n",
    "\n",
    "Model diagnostics\n",
    "-----------------\n",
    "\n",
    "Now that we have our models, we need to check that they are appropriate\n",
    "for the data. For this, we will inspect “diagnostic plots”. Producing\n",
    "diagnostic plots is easy in R — if you <span>plot</span> a model, then R\n",
    "produces a set of diagnostic plots!\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Try the following code (and include in the R script file):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> par(mfrow = c(2, 2), mar = c(5, 5, 1.5, 1.5))\n",
    "> plot(genomeSizeModelDragon)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This should give the plots shown in figure \\[fig:DiagModDragon\\].\n",
    "\n",
    "![Diagnostics for the <span>lm</span> fit to the Dragonfly data\n",
    "subset.<span\n",
    "data-label=\"fig:DiagModDragon\"></span>](DiagModDragon.pdf){width=\"70.00000%\"}\n",
    "\n",
    "And,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> par(mfrow = c(2, 2), mar = c(5, 5, 1.5, 1.5))\n",
    "> plot(genomeSizeModelDamsel)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This should give the plots shown in figure \\[fig:DiagModDamsel\\].\n",
    "\n",
    "![Diagnostics for the <span>lm</span> fit to the Damselfly data\n",
    "subset.<span\n",
    "data-label=\"fig:DiagModDamsel\"></span>](DiagModDamsel.pdf){width=\"70.00000%\"}\n",
    "\n",
    "The diagnostic plots are:\n",
    "\n",
    "This plot is used to spot if the distribution of the residuals (the\n",
    "vertical distance from a point to the regression line) has\n",
    "<span>*similar variance*</span> for different predicted values (the\n",
    "y-value on the line corresponding to each x-value). There should be no\n",
    "obvious patterns (such as curves) or big gaps. If there was no scatter,\n",
    "if all the points fell exactly on the line, then all of the dots on this\n",
    "plot would lie on the gray horizontal dashed line. The red line is a\n",
    "smoothed curve to make it easier to see trends in the residuals. It is\n",
    "flat in the Dragonfly model fit (Figure \\[fig:DiagModDragon\\]), and a\n",
    "bit more wavy than we would like in the in the Damselfly model fit\n",
    "(Figure \\[fig:DiagModDamsel\\]), but there are no clear trends in either,\n",
    "which is what you hope to see.\n",
    "\n",
    "This plot is to check whether the residuals are <span>*normally\n",
    "distributed* </span> — are the values of the observed residuals similar\n",
    "to those expected under a normal distribution? Ideally, the points\n",
    "should form a perfectly straight line, indicating that the observed\n",
    "residuals exactly match the expected. Here, note that the points lie\n",
    "pretty close to the dashed line in both Figures \\[fig:DiagModDragon\\] &\n",
    "\\[fig:DiagModDamsel\\], but deviate at the ends, especially for\n",
    "Damselflies. However, some deviation is to be expected near the ends —\n",
    "here these deviations are just about acceptable.\n",
    "\n",
    "The x-axis on this plot is identical to the Residuals vs Fitted plot –\n",
    "these are the fitted values. The y-axis is the square root of the\n",
    "<span>*standardized residuals*</span>, which are residuals rescaled so\n",
    "that they have a mean of zero and a variance of one. As a result, all\n",
    "y-axis values are positive. Thus large residuals (both positive and\n",
    "negative) plot at the top, and small residuals plot at the bottom (so\n",
    "only their <span>*scale*</span> is retained). Thus, all of the numbered\n",
    "points (which will be the same in all plots) plot at the top here. The\n",
    "red line here shows the trend, just like the Residuals vs Fitted plot.\n",
    "The regression analysis has assumed homoscedasticity, that the variance\n",
    "in the residuals doesn’t change as a function of the predictor. If that\n",
    "assumption is correct, the red line should be relatively flat. It is not\n",
    "quite as flat as we would like, especially for the Dragonfly analysis\n",
    "(Figure \\[fig:DiagModDragon\\]).\n",
    "\n",
    "This plot shows the standardized residuals against leverage. “Leverage”\n",
    "is a measure of how much each data point influences the linear model’s\n",
    "coefficient estimates. Because the regression line must pass through the\n",
    "centroid (“pivot point”) of the data (Figure \\[fig:Leverage\\]), points\n",
    "that lie far from the centroid have greater leverage, and their leverage\n",
    "increases if there are fewer points nearby. There are two key things to\n",
    "note about this plot:\n",
    "\n",
    "![Leverage of data points on slope of a regression. The points further\n",
    "away from the centroid in the x-axis direction have more leverage, and\n",
    "can therefore move the regression line up or down (dashed red\n",
    "lines).<span\n",
    "data-label=\"fig:Leverage\"></span>](Leverage.pdf){width=\"40.00000%\"}\n",
    "\n",
    "1.  The standardized residuals (y-axis) are centered around zero and\n",
    "    reach 2-3 standard deviations away from zero. They should also lie\n",
    "    symmetrically about zero, as would be expected for a\n",
    "    normal distribution. This is the case for the Damselfly plot\n",
    "    (Figure \\[fig:DiagModDamsel\\]) , but not so much for the Dragonfly\n",
    "    plot Figure \\[fig:DiagModDragon\\].\n",
    "\n",
    "2.  The contours values show <span>*Cook’s distance*</span> (only\n",
    "    visible in the Damsefly plot), which measures how much the\n",
    "    regression would change if a point was deleted. Cook’s distance is\n",
    "    increased by leverage and by large residuals: a point far from the\n",
    "    centroid with a large residual can severely distort the coefficient\n",
    "    estimates from the regression. On this plot, you want to see that\n",
    "    the red smoothed line stays close to the horizontal gray dashed line\n",
    "    and that no points have a large Cook’s distance (i.e, &gt;0.5). Both\n",
    "    are true here.\n",
    "\n",
    "This is an important diagnostic plot in regression analyses in\n",
    "particular because it tells you whether your estimate of the slope\n",
    "coefficient in particular is strongly affected by certain data points.\n",
    "\n",
    "Note that certain points are numbered in all the plots — these are\n",
    "points to pay special attention to because they are\n",
    "<span>*potential*</span> outliers. The numbers correspnd to the row\n",
    "number for that dataset in your data frame. You can easily identify\n",
    "these points in your data plot (Figure \\[fig:GenoRegModels\\]) because\n",
    "the order of the points along the fitted values axis (y-axis) in the\n",
    "diagnostic plot matches the order along the x-axis in the data plot. So,\n",
    "fo example here, in Figure \\[fig:DiagModDragon\\], the two numbered\n",
    "points (46, 10) near the bottom correspond in the data plot (Figure\n",
    "\\[fig:GenoRegModels\\]) to the two red points near the center-left that\n",
    "lie farthest below the red line.\n",
    "\n",
    "Thus, neither the Drangonfly nor the Damselfly diagnostic plots look\n",
    "perfect, but this level of deviation from assumptions of linear models\n",
    "is acceptable. The main worrying factors are that the QQ plot for\n",
    "Damselflies indicates the observed residuals are a bit more extreme than\n",
    "expected, and the Scale–Location plot for Dragonflies suggests some\n",
    "pattern in the standardized residuals wrt location of the fitted values.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Copy the code to create the diagnostic plots into your script to keep a\n",
    "record of the code and run it.\n",
    "\n",
    "Reporting the model\n",
    "-------------------\n",
    "\n",
    "Now we know that the models are appropriate and we have a plot, the last\n",
    "thing is to report the statistics. For the damselfly model, here is one\n",
    "summary that would do: log genome size explains significant variation in\n",
    "log body weight in dameselflies (F=10.5, df=1,36, p=0.0025) and shows\n",
    "that body weight decreases with genome size (intercept: -4.65, se=0.09;\n",
    "slope: -1.14, se=0.35).\n",
    "\n",
    "Linear Models: Analysis of variance {#ch:ANOVA}\n",
    "===================================\n",
    "\n",
    "Aims of this chapter [^6]:\n",
    "\n",
    "Plotting boxplots and barplots using factors\n",
    "\n",
    "Fitting factors in linear models using analysis of variance\n",
    "\n",
    "Diagnostic plots for explanatory factors\n",
    "\n",
    "Exploring differences between levels of a factor\n",
    "\n",
    "What is ANOVA?\n",
    "--------------\n",
    "\n",
    "![A dataset where an ANOVA would be appropriate. performing an ANOVA\n",
    "test on this dataset is the same as fitting the linear model\n",
    "$y  = \\beta_1  + \\beta_2 x_s + \\beta_3 x_a$, where $x_s$ and $x_a$ are\n",
    "two levels. There are three “treatments” here with the first treatment,\n",
    "the control, captured by the baseline value $\\beta_1$ (the sample with\n",
    "the lowest value, on the far left)<span\n",
    "data-label=\"fig:anova1\"></span>](ANOVA_is_LM.pdf){width=\".4\\textwidth\"}\n",
    "\n",
    "A <span>*One-way analysis of variance*</span> (one-way ANOVA) is a\n",
    "technique used to compare means of two or more samples representing\n",
    "numerical, continuous data.\n",
    "\n",
    "ANOVA tests the null hypothesis that samples from two or more groups are\n",
    "drawn from populations with the <span>*same mean value*</span>. To do\n",
    "this, ANOVA uses the <span>*F*</span>-statistic — the ratio of the\n",
    "variance calculated across the samples (groups) (the null hypothesis) to\n",
    "the variance within the samples (groups). If the null hypothesis that\n",
    "the group means are drawn from populations with the same mean is indeed\n",
    "true, the between-group variance (numerator in the F-ratio) should be\n",
    "lower than the within-group variance (denominator). A higher ratio (and\n",
    "<span>*F*</span> value) therefore implies that the samples were drawn\n",
    "from populations with different mean values.\n",
    "\n",
    "This is same as asking whether a linear model with a predictor (or\n",
    "explanatory variable) with at least two categorical levels (or factors),\n",
    "better accounts for the variance (Explained Sum of Squares, ESS) than a\n",
    "null model of the form $y  = \\beta_1$ (Figure \\[fig:anova1\\]). Thus,\n",
    "ANOVA is just a type of linear model.\n",
    "\n",
    "By the end of this chapter, it will make more sense to you how/why\n",
    "linear regression models that we covered in Chapter \\[ch:regress\\], of\n",
    "the form $y = \\beta_1  + \\beta_2 x$ (where $x$ is a continuous predictor\n",
    "variable), require ANOVA to determine if the the model better fits than\n",
    "a null model of the form $y  = \\beta_1$.\n",
    "\n",
    "Typically, one-way ANOVA is used to test for differences among at least\n",
    "three groups, since the two-group (or levels or factors) case can be\n",
    "covered by a $t$-test (see Chapter \\[ch:t\\_F\\_tests\\]). When there are\n",
    "only two means to compare, the $t$-test and the F-test are equivalent;\n",
    "the relation between ANOVA and t is given by $F = t^2$.\n",
    "\n",
    "An extension of one-way ANOVA is two-way analysis of variance that\n",
    "examines the influence of two different categorical independent\n",
    "variables on one dependent variable — we will look at multiple predictor\n",
    "variables in Chapter \\[ch:MulExpl\\] onwards.\n",
    "\n",
    "Calculating the ANOVA test statistic\n",
    "------------------------------------\n",
    "\n",
    "ANOVA “partitions” variability in your data as follows:\n",
    "\n",
    "Total sum of squares (TSS)\n",
    "\n",
    ":   This is sum of the squared difference between the observed dependent\n",
    "    variable ($y$) and the mean of the response variable $y$ (denoted by\n",
    "    $\\bar{y}$), i.e., $$\\text{TSS} = \\sum_{i=1}^{n}(y_i - \\bar{y})^2$$\n",
    "    TSS tells us how much variation there is in the dependent variable\n",
    "    without having any other information (your null model). You might\n",
    "    notice that TSS is the numerator of the sample variance you learned\n",
    "    about in Chapter \\[ch:ExpDesign\\].\n",
    "\n",
    "Explained sum of squares (ESS)\n",
    "\n",
    ":   Sum of the squared differences between the predicted $y$’s (denoted\n",
    "    $\\hat{y}$’s) and $\\bar{y}$, or,\n",
    "    $$\\text{ESS} = \\sum_{i=1}^{n} (\\hat{y}_i - \\bar{y})^2$$ ESS tells us\n",
    "    how much of the variation in the dependent variable our\n",
    "    alternative (linear) model was able to explain. That is, it’s the\n",
    "    reduction in uncertainty that occurs when the linear model is used\n",
    "    to predict the responses.\n",
    "\n",
    "Residual sum of squares (RSS)\n",
    "\n",
    ":   Sum of the squared differences between the observed $y$’s (denoted\n",
    "    by $y_i$) and the predicted $\\hat{y}$, or,\n",
    "    $$\\text{RSS} = \\sum_{i=1}^{n} (\\hat{y}_i - y_i)^2$$ RSS tells us how\n",
    "    much of the variation in the dependent variable our model could\n",
    "    not explain. That is, it’s the uncertainty that remains even after\n",
    "    the linear model is used. The linear model is considered to be\n",
    "    statistically significant if it can account for a large amount of\n",
    "    variability in the response.\n",
    "\n",
    "And of course, TSS = ESS + RSS; the OLS method “decomposes” the total\n",
    "variation in the dependent variable into an explained component (ESS;\n",
    "explained by the predictor) and an unexplained or residual component\n",
    "(the RSS).\n",
    "\n",
    "These sums of squares can then be used to calculate the statistical\n",
    "significance of the linear model (Regression, ANOVA, etc) through the\n",
    "F-Value (or F-Ratio), as follows:\n",
    "\n",
    "<span>|&gt;m<span>2cm</span>|&gt;m<span>2.1cm</span>|&gt;m<span>2cm</span>|&gt;m<span>2cm</span>|p<span>1.5cm</span>|</span>\n",
    "Type of Sum of Squares (SS) & Calculation & Degrees of Freedom (DF) &\n",
    "Mean Sum of Squares (MSS) & F-Value\\\n",
    "\n",
    "TSS & $\\sum_{i=1}^{n}(y_i - \\bar{y})^2$ & $n-1$ &$\\frac{TSS}{n-1}$&\\\n",
    "\n",
    "ESS & $\\sum_{i=1}^{n} (\\hat{y}_i - \\bar{y})^2$ & $n_c-1$ &\n",
    "$\\frac{ESS}{n_c-1}$ &\\\n",
    "\n",
    "RSS & $\\sum_{i=1}^{n} (\\hat{y}_i - y_i)^2$ & $n-n_c$ &\n",
    "$\\frac{RSS}{n-n_c}$ &\\\n",
    "\n",
    "### Degrees of freedom\n",
    "\n",
    "Thus each sum of squares has a corresponding degrees of freedom (DF)\n",
    "associated with it that gives the Mean Sum of Squares (MSS) — the Sums\n",
    "of Squares divided by the corresponding degrees of freedom.\n",
    "\n",
    "The TSS DF is one less than the number of observations $n-1$. This is\n",
    "because calculating TSS, needs $\\bar y$ , which imposes loss of one\n",
    "degree of freedom. Note that MSS is thus nothing but the sample\n",
    "variance.\n",
    "\n",
    "The ESS DF is one less than the number of coefficients ($n_c$)\n",
    "(estimated parameters) in the model: $n_c-1$. Note that in the case\n",
    "where the linear model is an ANOVA, it the number of coefficients equals\n",
    "the number of “treatments” (the categories or levels in the predictor).\n",
    "So for example, in Fig. \\[fig:anova1\\], there are three treatments\n",
    "(predictors) and therefore three coefficients ($\\beta_1$, $\\beta_2$,\n",
    "$\\beta_3$), which means that the ESS degrees of freedom there is\n",
    "$n_c-1 = 2$.\n",
    "\n",
    "The RSS DF is the sample size $n$ minus the number of coefficients\n",
    "$n_c$, that is, $n - n_c$, because each estimated coefficient is an\n",
    "unknown parameter.\n",
    "\n",
    "### The F-Value (or Ratio)\n",
    "\n",
    "Finally, The F-Value or F-Ratio, the test statistic used to decide\n",
    "whether the linear model fit is statistically significant, is the ratio\n",
    "of the Mean ESS to the Mean RSS. The null hypothesis is rejected if the\n",
    "F-ratio is large — the model explains a significant amount of variance.\n",
    "The p-value is then calculated from the F-distribution as you learned\n",
    "before, in Chapter \\[ch:t\\_F\\_tests\\] (see Fig. \\[fig:fdist\\]).\n",
    "\n",
    "Also note that the Root Mean Square Error (RMSE), also known as the\n",
    "standard error of the estimate, is the square root of the Mean RSS. It\n",
    "is the standard deviation of the data about the Linear model, rather\n",
    "than about the sample mean.\n",
    "\n",
    "### The $R^{2}$\n",
    "\n",
    "Finally, $R^{2}$, also called the Coefficient of Determination, is the\n",
    "proportion of total error (TSS) explained by the model (ESS), so the\n",
    "ratio ESS/TSS. That is it is the proportion of the variability in the\n",
    "response that is explained by by the fitted model. Since TSS = ESS +\n",
    "RSS, $R^{2}$ can be rewritten as (TSS-RSS)/TSS = 1 - RSS/TSS. If a model\n",
    "has perfectly fits the data, $R^{2}=1$, and if it has no predictive\n",
    "capability $R^{2}=0$. In reality, $R^{2}$ will never be exactly 0\n",
    "because even a null model will explain some variance just by chance due\n",
    "to sampling error. Note that $R$, the square root of $R^2$, is the\n",
    "multiple correlation coefficient: the correlation between the observed\n",
    "values ($y$), and the predicted values ($\\hat{y}$).\n",
    "\n",
    "As additional predictors (end therefore linear model coefficients) are\n",
    "added to a linear model, $R^2$ increases even when the new predictors\n",
    "add no real predictive capability. The adjusted-$R^2$ tries to addresses\n",
    "this problem of over-specification or over-fitting by including the\n",
    "degrees of freedom: Adjusted $R^2$ = 1 - (RSS/$n-n_c-2$)/(TSS/$n-1$)\n",
    "[^7]. Thus additional predictors with little explanatory capability will\n",
    "increase the ESS (and reduce the RSS), but they will also have lower RSS\n",
    "degrees of freedom (because of the additional number of fitted\n",
    "coefficients, $n_c$’s)[^8]. Thus if the additional predictors have poor\n",
    "predictive capability, these two reductions will cancel each other out.\n",
    "In other words, the Adjusted $R^2$ penalizes the addition of new\n",
    "predictors to the linear model, so you should always have a look at the\n",
    "Adjusted $R^2$ as a corrected measure of $R^2$.\n",
    "\n",
    "A new dataset\n",
    "-------------\n",
    "\n",
    "In this Chapter, we will use a new dataset of genome size and life\n",
    "history in mammals to try out a one-way ANOVA. The dataset is a\n",
    "composite of data taken from an online database of genome sizes and a\n",
    "published database of mammalian life history:\n",
    "\n",
    "Average genome sizes for available mammal species are taken from the\n",
    "online database [www.genomesize.com](www.genomesize.com).\n",
    "\n",
    "Trait data for these species are taken from: [ Jones, K. E. <span>*et\n",
    "al.*</span> (2009) PanTHERIA: a species-level database of life history,\n",
    "ecology, and geography of extant and recently extinct mammals. Ecology\n",
    "90, 2648–2648](http://esapubs.org/archive/ecol/e090/184/metadata.htm).\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Download the file <span>MammalData.csv</span> from bitbucket and save to\n",
    "your <span>Data</span> directory.\n",
    "\n",
    "Create a new blank script called <span>ANOVA\\_Prac.R</span> and add some\n",
    "introductory comments.\n",
    "\n",
    "Use <span>read.csv</span> to load the data in the data frame\n",
    "<span>mammals</span> and then <span>str</span> and <span>summary</span>\n",
    "to examine the data.\n",
    "\n",
    "There are nine variables. The first two are the latin binomial and\n",
    "taxonomic order of each species, followed by the species mean genome\n",
    "size (‘C value’, picograms), adult body mass (g), diet breadth, habitat\n",
    "breadth, litter size and then two factors showing whether the species\n",
    "are ground dwelling and their trophic level. For more information, see\n",
    "the link above.\n",
    "\n",
    "You will see from the output of <span>summary</span> that there is lots\n",
    "of missing data for the life history traits.\n",
    "\n",
    "Exploring the data with boxplots\n",
    "--------------------------------\n",
    "\n",
    "We are interested in finding out whether the mean C value for species\n",
    "varies predictably for different levels of life history traits (a\n",
    "typical one-way ANOVA question). For example:\n",
    "\n",
    "Do carnivores or herbivores have larger genome sizes?\n",
    "\n",
    "Do ground dwelling mammals have larger or smaller genome sizes?\n",
    "\n",
    "Before we fit any models, we want to plot the data to see if the means\n",
    "within these groupings look different. We also want to check whether the\n",
    "variance looks similar for each group: <span>*constant normal\n",
    "variance*</span>! A simple way is to look at box and whisker plots,\n",
    "showing the median and range of the data:\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Use <span>plot(meanCvalue \\~ TrophicLevel, data= mammals)</span> to\n",
    "generate a boxplot of the differences in genome sizes between trophic\n",
    "levels.\n",
    "\n",
    "Looking at the plots, it is clear that there is more spread in the data\n",
    "above the median than below. Create a new variable <span>\n",
    "logCvalue</span> in the <span>mammals</span> data frame containing\n",
    "logged C values.\n",
    "\n",
    "Create a boxplot of log C values within trophic groups.\n",
    "\n",
    "Repeat the two plot commands to look at differences between ground\n",
    "dwelling and other species.\n",
    "\n",
    "Differences in means with barplots\n",
    "----------------------------------\n",
    "\n",
    "Box and whisker plots show the median and spread in the data very\n",
    "clearly, but we want to test whether the means are different. This is\n",
    "$t$ test territory — how different are the means given the standard\n",
    "error — so it is common to use barplots and standard error bars to show\n",
    "these differences.\n",
    "\n",
    "We’re going to use some R code to construct a barplot by hand. We need\n",
    "to calculate the means and standard errors within trophic groups, but\n",
    "before we can do that, we need a new functions to calculate the standard\n",
    "error of a mean:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get standard error of the mean from a set of values (x)\n",
    "\n",
    "seMean <- function(x){\n",
    "\tx <- na.omit(x) # get rid of missing values\n",
    "\n",
    "\tse <- sqrt(var(x)/length(x)) # calculate the standard error\n",
    "\n",
    "\treturn(se) \t# tell the function to return the standard error\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can use the function <span>tapply</span>: it splits a variable up\n",
    "into groups from a factor and calculates statistics on each group using\n",
    "a function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trophMeans <- tapply(mammals$logCvalue, mammals$TrophicLevel, FUN = \n",
    "mean, na.rm = TRUE)\n",
    "\n",
    "print(trophMeans)\n",
    "\n",
    " Carnivore Herbivore  Omnivore \n",
    "     1.085     1.197     1.236 \n",
    "\n",
    "trophSE <- tapply(mammals$logCvalue, mammals$TrophicLevel, FUN = seMean)\n",
    "\n",
    "print(trophSE)\n",
    "\n",
    " Carnivore Herbivore  Omnivore \n",
    "   0.03983   0.02206   0.01844 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have to put these values together on the plot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the upper and lower limits of the error bars\n",
    "upperSE <- trophMeans + trophSE\n",
    "lowerSE <- trophMeans - trophSE\n",
    "\n",
    "# get a barplot\n",
    "# - this function can report where the middle of the bars are on the x-axis\n",
    "# - set the y axis limits to contain the error bars\n",
    "\n",
    "barMids <- barplot(trophMeans, ylim=c(0, max(upperSE)), ylab = 'log C value (pg)')\n",
    "\n",
    "# Now use the function to add error bars\n",
    "# - draws arrows between the points (x0,y0) and (x1,y1)\n",
    "# - arrow heads at each end (code=3) and at 90 degree angles\n",
    "\n",
    "arrows(barMids, upperSE, barMids, lowerSE, ang=90, code=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image](TLbarplot.pdf){width=\"70.00000%\"}\n",
    "\n",
    "Now we need to draw all these pieces together into a script and get used\n",
    "to using them.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Copy all the lines of code from this section into your script.\n",
    "\n",
    "Run it and check you get the graph above.\n",
    "\n",
    "Use the second two chunks as a model to plot a similar graph for\n",
    "<span>GroundDwelling</span>. You should get something like the plot\n",
    "below.\n",
    "\n",
    "![image](GDbarplot.pdf){width=\"60.00000%\"}\n",
    "\n",
    "An alternative to barplots\n",
    "--------------------------\n",
    "\n",
    "That is a lot of work to go through for a plot. Doing it the hard way\n",
    "uses some useful tricks, but one strength of R is that there is a huge\n",
    "list of add-on packages that you can use to get new functions that other\n",
    "people have written.\n",
    "\n",
    "We will use the <span>gplots</span> package to create plots of group\n",
    "means and confidence intervals. Rather than plotting the means $\\pm$ 1\n",
    "standard error, the option <span>p=0.95</span> uses the standard error\n",
    "and the number of data points to get 95% confidence intervals. The\n",
    "default <span>connect=TRUE</span> option adds a line connecting the\n",
    "means, which isn’t useful here.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Replicate the code below into your script and run it to get the plots\n",
    "below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load the gplots package\n",
    "> library(gplots)\n",
    "\n",
    "# Get plots of group means and standard errors\n",
    "> par(mfrow=c(1,2))\n",
    "> plotmeans(logCvalue ~ TrophicLevel, data=mammals, p=0.95, connect=FALSE)\n",
    "> plotmeans(logCvalue ~ GroundDwelling, data=mammals, p=0.95, connect=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analysis of variance\n",
    "--------------------\n",
    "\n",
    "Hopefully, those plots should convince you that there are differences in\n",
    "genome size between different trophic groups and between ground dwelling\n",
    "and other mammals. We’ll now use a linear model to test whether those\n",
    "differences are significant.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Using your code from Chapter \\[ch:regress\\] as a guide, create a linear\n",
    "model called <span>trophicLM</span> which models log C value as a\n",
    "function of trophic group.\n",
    "\n",
    "Use <span>anova</span> and <span>summary</span> to look at the analysis\n",
    "of variance table and then the coefficients of the model.\n",
    "\n",
    "The ANOVA table for the model should look like the one below: trophic\n",
    "level explains highly significant variation in genome size ($F= 7.22, \n",
    "\\textrm{df}=2 \\textrm{ and } 300, p =0.0009$). <span>*Note the style of\n",
    "reporting the result*</span> - the statistic ($F$), degrees of freedom\n",
    "and $p$ value are all provided in support. It is common to contract this\n",
    "style to this: $F_{2,300}=7.22, p=0.0009$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> anova(trophicLM)\n",
    " \n",
    " Analysis of Variance Table\n",
    " \n",
    " Response: logCvalue\n",
    "               Df Sum Sq Mean Sq F value  Pr(>F)    \n",
    " TrophicLevel   2   0.83   0.413    7.22 0.00087 ***\n",
    " Residuals    300  17.18   0.057                    \n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, look at the sum of squares column. Of a total of $17.18+0.83 = \n",
    "18.01$ units of sums of squares, only 0.83 are explained by trophic\n",
    "level: $0.83/18.01 \\approx 0.046$ or 4.6%. This ratio is called $r^2$, a\n",
    "measure of explanatory power, and shows that, although the model is very\n",
    "significant, it isn’t very explanatory. We care about explanatory power\n",
    "or effect size, <span>*not*</span> $p$ values.\n",
    "\n",
    "The coefficients table for the model looks like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> summary(trophicLM)\n",
    " \n",
    " Call:\n",
    " lm(formula = logCvalue ~ TrophicLevel, data = mammals)\n",
    " \n",
    " Residuals:\n",
    "     Min      1Q  Median      3Q     Max \n",
    " -0.5038 -0.1635 -0.0038  0.1511  0.9313 \n",
    " \n",
    " Coefficients:\n",
    "                       Estimate Std. Error t value Pr(>|t|)    \n",
    " (Intercept)             1.0851     0.0335   32.38  < 2e-16 ***\n",
    " TrophicLevelHerbivore   0.1119     0.0396    2.83  0.00503 ** \n",
    " TrophicLevelOmnivore    0.1513     0.0399    3.80  0.00018 ***\n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 \n",
    " \n",
    " Residual standard error: 0.239 on 300 degrees of freedom\n",
    "   (76 observations deleted due to missingness)\n",
    " Multiple R-squared: 0.0459,\tAdjusted R-squared: 0.0396 \n",
    " F-statistic: 7.22 on 2 and 300 DF,  p-value: 0.000866 \n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It shows the following:\n",
    "\n",
    "The reference level (or intercept) is for carnivores. Their mean genome\n",
    "size is significantly different from zero - this is not an exciting\n",
    "finding!\n",
    "\n",
    "The mean genome size for both herbivores and omnivores are both\n",
    "significantly different from carnivores. Both larger in fact: herbivore\n",
    "mean genome size = $1.085 + 0.112 = 1.197$ and omnivore mean genome size\n",
    "= $1.085 + 0.151 = 1.236$. These are the same group means we found\n",
    "above.\n",
    "\n",
    "The $r^2$ is shown and is the 4.6% we calculated above. The\n",
    "<span>*adjusted*</span> $r^2$ reduces the raw $r^2$ to account for the\n",
    "number of variables included in the model. That 4.6% would be even less\n",
    "impressive if we needed 6 explanatory variables to get it…\n",
    "\n",
    "The $F$ statistic, as in the ANOVA table above.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Repeat the analysis of variance above to look at the effects of ground\n",
    "dwelling on genome size.\n",
    "\n",
    "Model criticism\n",
    "---------------\n",
    "\n",
    "The next question must be — and actually, we should do this before we go\n",
    "anywhere near the model summaries — is the model appropriate to the\n",
    "data.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Using Chapter \\[ch:regress\\] to guide you, get the four model diagnostic\n",
    "plots for the trophic level model on a single figure.\n",
    "\n",
    "The four plots are:\n",
    "\n",
    "![image](modelDiag.pdf){width=\"80.00000%\"}\n",
    "\n",
    "Note that in regression, the predicted (or fitted) values from the model\n",
    "take a range along the relationship $y=a + bx$ (as we saw in the Figures\n",
    "\\[fig:DiagModDragon\\] & \\[fig:DiagModDamsel\\]). For ANOVA, there are\n",
    "only a few predicted values — one for each group mean. This means that\n",
    "the plots above look different but we are looking for the same things:\n",
    "is there constant variance at each fitted value and are the residuals\n",
    "normally distributed? The answer for this model looks to be yes.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Check the ground dwelling model in the same way.\n",
    "\n",
    "Testing pairwise differences between levels\n",
    "-------------------------------------------\n",
    "\n",
    "The one thing that the trophic level model does not tell us is whether\n",
    "there is a difference in genome size between omnivores and herbivores —\n",
    "both are compared to carnivores, but not to each other. This is because\n",
    "of the multiple pairwise testing problem mentioned in Chapter\n",
    "\\[ch:t\\_F\\_tests\\] — if you do lots of tests then you may find small $p$\n",
    "values by chance and say something important is going on when it is just\n",
    "random chance. This is called a false positive or Type I error.\n",
    "\n",
    "With a 95% confidence interval, there is a 5% chance of a false positive\n",
    "<span>*per test*</span> but there are ways of getting a 5% chance across\n",
    "a set (or family) of tests. For linear models, we can use Tukey’s Honest\n",
    "Significant Difference test. We have to convert the <span> lm</span>\n",
    "object into an <span>aov</span> object first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> TukeyTroph <- TukeyHSD(aov(trophicLM))\n",
    "> print(TukeyTroph)\n",
    "\t\n",
    "   Tukey multiple comparisons of means\n",
    "     95% family-wise confidence level\n",
    " \n",
    " Fit: aov(formula = trophicLM)\n",
    " \n",
    " $TrophicLevel\n",
    "                        diff      lwr    upr  p adj\n",
    " Herbivore-Carnivore 0.11186  0.01863 0.2051 0.0139\n",
    " Omnivore-Carnivore  0.15128  0.05741 0.2452 0.0005\n",
    " Omnivore-Herbivore  0.03942 -0.03161 0.1104 0.3923\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The table shows the following:\n",
    "\n",
    "The differences between the three possible pairs and then the lower and\n",
    "upper bounds of the 95% confidence interval for the difference and a $p$\n",
    "value.\n",
    "\n",
    "In each case, we want to know if the difference could be zero: does the\n",
    "95% confidence interval for each pair include zero.\n",
    "\n",
    "For the first two pairs, carnivores versus omnivores and herbivores, the\n",
    "confidence intervals do not include zero, so they are significantly\n",
    "different. For the comparison between herbivores and omnivores, the\n",
    "interval does include zero (difference = 0.039, 95% CI’s limits are\n",
    "-0.032 & 0.110), so these groups are not significantly different.\n",
    "\n",
    "The $p$ values for the top two pairs are both larger (less significant)\n",
    "than in the summary table. The test has made it harder to find\n",
    "significant results.\n",
    "\n",
    "You can visualise these confidence intervals by plotting the Tukey test.\n",
    "You have to tweak the graphics parameters to get a clean plot though."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> par(las=1, mar=c(4,10,3,1))\n",
    "# las= 1 turns labels horizontal\n",
    "# mar makes the left margin wider (bottom, left, top, right)\n",
    "> plot(TukeyTroph)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The result should be:\n",
    "\n",
    "<span>![image](TukeyPLot.pdf){width=\"80.00000%\"} </span>\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Run the Tukey test in your script for both the trophic level and ground\n",
    "dwelling models.\n",
    "\n",
    "Are the factors independent?\n",
    "----------------------------\n",
    "\n",
    "We’ve looked at two models, using trophic level and ground dwelling. It\n",
    "is worth asking whether these are independent factors. What if, for\n",
    "example, our herbivores are all big, ground dwellers? This is important\n",
    "to know because otherwise, a two-way ANOVA would not be appropriate. We\n",
    "will look at interactions in Chapter \\[ch:MulExplInter\\].\n",
    "\n",
    "OK, so we want to know whether the two factors are independent. This is\n",
    "a job for the $\\chi^2$ test!\n",
    "\n",
    "### The Chi-square test and count data\n",
    "\n",
    "The Chi-square test, also known as $\\chi^{2}$ test or chi-square test,\n",
    "is designed for scenarios where you want to statistically test how\n",
    "likely it is that an observed distribution of values is due to chance.\n",
    "It is also called a “goodness of fit” statistic, because it measures how\n",
    "well the observed distribution of data fits with the distribution that\n",
    "is expected if the variables of which measurements are made are\n",
    "independent. In our mammals example below, the two variables are trophic\n",
    "level and ground dwelling.\n",
    "\n",
    "Note that a $\\chi^{2}$ test is designed to analyze categorical data.\n",
    "That is the data have been counted (count data) and divided into\n",
    "categories. It is not meant for continuous data (such as body weight,\n",
    "genome size, or height). For example, if you want to test whether\n",
    "attending class influences how students perform on an exam, using test\n",
    "scores (from 0-100) as data would not be appropriate for a Chi-square\n",
    "test. However, arranging students into the categories “Pass” and “Fail”\n",
    "and counting up how many fall in each categories would be appropriate.\n",
    "Additionally, the data in a Chi-square table (see below) should not be\n",
    "in the form of percentages – only count data are allowed!\n",
    "\n",
    "#### The Chi-square test with the mammals data\n",
    "\n",
    "We can easily build a table for a Chi-square test on the mammals data as\n",
    "follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> factorTable <- table(mammals$GroundDwelling, mammals$TrophicLevel)\n",
    "> print(factorTable) \n",
    "\n",
    "       Carnivore Herbivore Omnivore\n",
    "   No         26        45       64\n",
    "   Yes        22        62       40"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let’s run the test:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> chisq.test(factorTable)\n",
    "  \n",
    " \tPearson's Chi-squared test\n",
    " \n",
    " data:  factorTable \n",
    " X-squared = 8.12, df = 2, p-value = 0.01725"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The “<span>X-squared</span>” value is the $\\chi^{2}$ <span>*test\n",
    "statistic*</span>, akin to the t-value of the t-test or W value in the\n",
    "Wilcox test.\n",
    "\n",
    "The $\\chi^{2}$ statistic is calculated as the sum of the quantity\n",
    "$$\\frac{(\\mathrm{Observed} - \\mathrm{Expected})^2}{\\mathrm{Expected}}$$\n",
    "across all the cells/categories in the table (so the sum would be over 6\n",
    "categories in our current mammals example).\n",
    "\n",
    "“Observed” is the observed proportion of data that fall in a certain\n",
    "category. For example, there are 26 species observed in the\n",
    "<span>Carnivore</span>, <span>No</span> category, and 22 in the\n",
    "<span>Carnivore</span>, <span> Yes</span> category.\n",
    "\n",
    "“Expected” is what count would be expected if the values in each\n",
    "category we truly independent. Each cell has its own expected value,\n",
    "which is simply calculated as the count one would expect in each\n",
    "category if the value were generated in proportion to the total number\n",
    "seen in that category. So in our example, the expected value for the\n",
    "<span>Carnivore</span>, <span>No</span> category would be\n",
    "$$26+22 \\mathrm{~(Total~number~of~carnivore~species)} \n",
    "\\times \\frac{26+45+64 \\mathrm{~(Total~number~in~the~''No''~category)}}{ \n",
    "26+22+45+62+64+40 \\mathrm{~(Total~number~of~species)}}$$\n",
    "$$= 48 \\times \\frac{135}{259} = 25.02$$\n",
    "\n",
    "The sum of all six (one for each cell in the table above) such\n",
    "calculations would be the $\\chi^{2}$ value that R gave you through the\n",
    "<span>chisq.test()</span> above — try it!\n",
    "\n",
    "Now back to the R output from the <span>chisq.test()</span> above. Why\n",
    "df = 2? This is calculated as $DF = (r - 1) * (c - 1)$ where $r$ and $c$\n",
    "are the number of rows and columns in the $\\chi^{2}$ table,\n",
    "respectively. The same principle you learned before applies here; you\n",
    "lose one degree of freedom for each new level of information you need to\n",
    "estimate: there is uncertainity about the information (number of\n",
    "categories) in both rows and columns, so you need to lose one degree of\n",
    "freedom for each.\n",
    "\n",
    "Finally, note that the p-value is significant — we can conclude that the\n",
    "factors aren’t independent. From the table, carnivores can be either\n",
    "ground dwelling or not, but herbivores tend to be ground dwelling and\n",
    "omnivores tend not to be. Ah well... it’s OK. We will look at a better\n",
    "way to analyze these data using “interactions” in Chapter\n",
    "\\[ch:MulExplInter\\].\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Include and run the $\\chi^2$ test in your script.\n",
    "\n",
    "Saving data\n",
    "-----------\n",
    "\n",
    "The last thing to do is to save a copy of the mammal data, including our\n",
    "new column of log data, for use in later chapters.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Use this code in your script to create the saved data in you\n",
    "<span>Data</span> directory :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "save(mammals, file='../Data/mammals.Rdata')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Linear Models: Multiple explanatory variables {#ch:MulExpl}\n",
    "=============================================\n",
    "\n",
    "Aims of this chapter[^9]:\n",
    "\n",
    "Including several explanatory variables in a model\n",
    "\n",
    "Interpreting summary tables for more complex models\n",
    "\n",
    "Loading the data\n",
    "----------------\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Create a new blank script called <span>MulExpl.R</span> in your\n",
    "<span>Code</span> directory and add some introductory comments.\n",
    "\n",
    "Use <span>load(’../Data/mammals.Rdata’)</span> to load the data saved at\n",
    "the end of Chapter \\[ch:ANOVA\\]. Look back at the end of Chapter\n",
    "\\[ch:ANOVA\\] to see how you saved the RData file. If <span>\n",
    "mammals.Rdata</span> is missing, just import the data again using <span>\n",
    "read.csv(“../Data/MammalData.csv”)</span> and add the log C Value column\n",
    "to the imported data frame again (got back to Chapter \\[ch:ANOVA\\] and\n",
    "have a look if you have forgotten how).\n",
    "\n",
    "Use <span>ls</span> and <span>str</span> to check that the data has\n",
    "loaded correctly.\n",
    "\n",
    "The models we looked at in Chapter \\[ch:ANOVA\\] explored whether the log\n",
    "genome size (C value, in picograms) of terrestrial mammals varied with\n",
    "trophic level and whether or not the species is ground dwelling. We will\n",
    "now look at a single model that includes both explanatory variables.\n",
    "\n",
    "The first thing to do is look at the data again. In Chapter\n",
    "\\[ch:regress\\], we asked if carnivores or herbivores had larger genomes.\n",
    "Now we want to ask questions like: do ground-dwelling carnivores have\n",
    "larger genomes than arboreal or flying omnivores? We need to look at\n",
    "plots within groups.\n",
    "\n",
    "Before we do that, there is a lot of missing data in the data frame and\n",
    "we should make sure that we are using the same data for our plots and\n",
    "models. We will subset the data down to the complete data for the three\n",
    "variables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> mammals <- subset(mammals, select = c(GroundDwelling, TrophicLevel, \n",
    "logCvalue))\n",
    "> mammals <- na.omit(mammals)\n",
    "> str(mammals)\n",
    "\n",
    "'data.frame':\t259 obs. of  3 variables:\n",
    " $GroundDwelling: Factor w/ 2 levels \"No\",\"Yes\": 2 2 2 2 2 1 2 1 1 1 ...\n",
    " $TrophicLevel  : Factor w/ 3 levels \"Carnivore\",\"Herbivore\",..: 1 2 2 2 3 3 3 2 2 3 ...\n",
    " $logCvalue     : num  0.94 1.322 1.381 1.545 0.888 ...\n",
    "  - attr(*, \"na.action\")=Class 'omit'  Named int [1:120] 2 4 7 9 10 11 14 15 20 21 ...\n",
    "   .. ..- attr(*, \"names\")= chr [1:120] \"2\" \"4\" \"7\" \"9\" ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Boxplots within groups\n",
    "----------------------\n",
    "\n",
    "In Chapter \\[ch:regress\\], we used the <span>subset</span> option to fit\n",
    "a model just to dragonflies. You can use <span>subset</span> with plots\n",
    "too.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Add <span>par(mfrow=c(1,2))</span> to your script to split the graphics\n",
    "into two panels.\n",
    "\n",
    "Copy the code from Chapter \\[ch:ANOVA\\] to create a boxplot of genome\n",
    "size by trophic level into your script.\n",
    "\n",
    "Using this, and adding a <span>subset</span> option to the code,\n",
    "generate the plots shown in the figure below.\n",
    "\n",
    "You can use the option <span>main</span> to add titles to a plot.\n",
    "\n",
    "![image](boxplots.pdf){width=\"100.00000%\"}\n",
    "\n",
    "<span>lattice</span> again\n",
    "--------------------------\n",
    "\n",
    "Recall that the <span>lattice</span> package provides some very neat\n",
    "extra ways to plot data in groups. They look pretty but the downside is\n",
    "that they don’t use the same graphics system — all those\n",
    "<span>par</span> commands are useless for these graphs. The defaults\n",
    "look good though!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> library(lattice)\n",
    "> bwplot(logCvalue ~ TrophicLevel | GroundDwelling, data= mammals)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image](bwplots.pdf){width=\"100.00000%\"}\n",
    "\n",
    "The code <span>logCvalue \\~ TrophicLevel | GroundDwelling</span> means\n",
    "plot the relationship between genome size and trophic level, but group\n",
    "within levels of ground dwelling. We are using the function\n",
    "<span>bwplot</span>, which is provided by <span>lattice</span> to create\n",
    "box and whisker plots.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Create the lattice plots above from within your script.\n",
    "\n",
    "Rearrange this code to have three plots, showing the box and whisker\n",
    "plots for <span>GroundDwelling</span>, grouped within the levels of\n",
    "<span>TrophicLevel</span>.\n",
    "\n",
    "Try reshaping the R plot window and running the command again. Lattice\n",
    "tries to make good use of the available space when creating lattice\n",
    "plots.\n",
    "\n",
    "Barplots again\n",
    "--------------\n",
    "\n",
    "We’re going to make the barplot code from Chapter \\[ch:regress\\] even\n",
    "more complicated! This time we want to know the mean log genome size\n",
    "within combinations of <span>TrophicLevel</span> and\n",
    "<span>GroundDwelling</span>. We can still use <span>tapply</span>,\n",
    "providing more than one grouping factor. We create a set of grouping\n",
    "factors like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> groups <- list(mammals$GroundDwelling, mammals$TrophicLevel)\n",
    "> groupMeans <- tapply(mammals$logCvalue, groups, FUN = mean)\n",
    "> print(groupMeans)\n",
    "\n",
    "     Carnivore Herbivore Omnivore\n",
    " No     0.9589     1.012    1.192\n",
    " Yes    1.2138     1.298    1.299"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Copy this code into your script and run it.\n",
    "\n",
    "Use this code and the script from Chapter \\[ch:ANOVA\\] to get the set of\n",
    "standard errors for the groups(<span>groupSE</span>). You should get\n",
    "this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Carnivore Herbivore Omnivore\n",
    "     No    0.04842   0.03419  0.02410\n",
    "     Yes   0.05976   0.02787  0.03587"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can use <span>barplot</span>. The default option for a barplot of\n",
    "a table is to create a stacked barplot, which is not what we want. The\n",
    "option <span>beside=TRUE</span> makes the bars for each column appear\n",
    "side by side. Once again, we save the midpoints of the bars to add the\n",
    "error bars. The other options in the code below change the colours of\n",
    "the bars and the length of error bar caps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\t# get upper and lower standard error height\n",
    "> upperSE <- groupMeans + groupSE\n",
    "> lowerSE <- groupMeans - groupSE\n",
    "\n",
    "# create barplot\n",
    "> barMids <- barplot(groupMeans, ylim=c(0, max(upperSE)), beside=TRUE, \n",
    "ylab= ' log C value (pg) ' , col=c( ' white ' , ' grey70 '))\n",
    "\n",
    "> arrows(barMids, upperSE, barMids, lowerSE, ang=90, code=3, len=0.05)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image](barplot_1.pdf){width=\"60.00000%\"}\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Generate the barplot above and then edit your script to change the\n",
    "colours and error bar lengths to your taste.\n",
    "\n",
    "Plotting means and confidence intervals\n",
    "---------------------------------------\n",
    "\n",
    "We’ll use the <span>plotmeans</span> function again as an exercise to\n",
    "change graph settings and to prepare figures for write ups.\n",
    "\n",
    "![Means and 95% confidence intervals for log genome size (picograms) in\n",
    "mammals for different trophic levels for a) ground dwelling species and\n",
    "b) other species.<span\n",
    "data-label=\"fig:plotmeans\"></span>](plotmeans.pdf){width=\"90.00000%\"}\n",
    "\n",
    "The default options in R use wide margins and spaced out axes and take\n",
    "up a lot of space that could be used for plotting data. You’ve already\n",
    "seen the <span>par</span> function and the options <span>mfrow</span>\n",
    "for multiple plots and <span>mar</span> to adjust margin size. The\n",
    "option <span>mgp</span> adjusts the placement of the axis label, tick\n",
    "labels and tick locations. See <span>?par</span> for help on the these\n",
    "options.\n",
    "\n",
    "Adding large titles to graphs is also a bad idea — it uses lots of space\n",
    "to explain something that should be in the figure legend. With multiple\n",
    "plots in a figure, you have to label graphs so that the figure legend\n",
    "can refer to them. You can add labels using\n",
    "<span>text(x,y,’label’)</span>.\n",
    "\n",
    "A figure legend should give a clear stand-alone description of the whole\n",
    "figure.\n",
    "\n",
    "You <span>*must*</span> link from your text to your figures — a reader\n",
    "has to know which figures refer to which results. So: ‘There are clear\n",
    "differences in mean genome size between species at different trophic\n",
    "levels and between ground dwelling and other species (Figure\n",
    "\\[fig:plotmeans\\])’.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Use <span>plotmeans</span> from Chapter \\[ch:ANOVA\\] and the <span>\n",
    "subset</span> option to generate the two plots below. You will need to\n",
    "set the <span>ylim</span> option for the two plots to make them use the\n",
    "same $y$ axis.\n",
    "\n",
    "Use <span>text</span> to add labels — the command\n",
    "<span>par(’usr’)</span> will show you the limits of the plot\n",
    "($x_{min}, x_{max}, y_{min}, \n",
    "\ty_{max}$) and help pick a location for the labels.\n",
    "\n",
    "Change the <span>par</span> settings in your code and redraw the plots\n",
    "to try and make better use of the space. In the example below, the box\n",
    "shows the edges of the R graphics window.\n",
    "\n",
    "Multiple explanatory variables\n",
    "------------------------------\n",
    "\n",
    "All those graphs suggest:\n",
    "\n",
    "Carnivores have smaller genome size; omnivores have larger genome size.\n",
    "\n",
    "Herbivores are somewhere in between, but not consistently.\n",
    "\n",
    "All ground dwelling mammals typically have larger genome sizes.\n",
    "\n",
    "We suspected these things from Chapter \\[ch:ANOVA\\], but now we can see\n",
    "that they might have separate effects. We’ll fit a linear model to\n",
    "explore this and add the two explanatory variables together.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "This is an important section — read it through carefully and ask\n",
    "questions if you are unsure. Copy the code into your script and add\n",
    "comments. <span>*Do not just jump to the next action item*</span>!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> model <- lm(logCvalue ~ TrophicLevel + GroundDwelling, data = mammals)\t"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We’re going to do things right this time and check the model diagnostics\n",
    "before we rush into interpretation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> par(mfrow=c(2,2))\n",
    "> plot(model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image](diagnostics1.pdf){width=\"\\textwidth\"}\n",
    "\n",
    "There are six predicted values now - three trophic levels for each of\n",
    "the two levels of ground dwelling. Those plots look ok so now we can\n",
    "look at the analysis of variance table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> anova(model)\n",
    "\n",
    " Analysis of Variance Table\n",
    " \n",
    " Response: logCvalue\n",
    "                 Df Sum Sq Mean Sq F value  Pr(>F)    \n",
    " TrophicLevel     2   0.81   0.407    7.86 0.00049 ***\n",
    " GroundDwelling   1   2.75   2.747   53.04 4.1e-12 ***\n",
    " Residuals      255  13.21   0.052                    \n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span>*Ignore the $p$ values*</span>! Yes, they’re highly significant\n",
    "but we want to understand the model, not rubber stamp it with\n",
    "‘significant’.\n",
    "\n",
    "The sums of squares for the variables are both small compared to the\n",
    "residual sums of squares — there is lots of unexplained variation. We\n",
    "can calculate the $r^2$ as explained sums of squares over total sums of\n",
    "squares:\n",
    "$$\\frac{0.81 + 2.75}{0.81 + 2.75 + 13.21} = \\frac{3.56}{16.77} = 0.212$$\n",
    "\n",
    "Trophic level explain much less variation than ground dwelling — this\n",
    "makes intuitive sense from the plots since there are big differences\n",
    "between Figure \\[fig:plotmeans\\]a and \\[fig:plotmeans\\]b, but small\n",
    "differences within.\n",
    "\n",
    "We could also calculate a significance for the whole model by merging\n",
    "the terms. The total explained sums of squares of $0.81 + 2.75 \n",
    "=3.56$ uses $2+1 =3$ degrees of freedom, so the mean sums of squares for\n",
    "all the terms together is $3.56/3=1.187$. Dividing this by the residual\n",
    "mean square of 0.052 gives an F of $1.187 / 0.052 = 22.83$.\n",
    "\n",
    "Now we can look at the summary table to see the coefficients."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> summary(model) \n",
    "\n",
    " Call:\n",
    " lm(formula = logCvalue ~ TrophicLevel + GroundDwelling, data = mammals)\n",
    " \n",
    " Residuals:\n",
    "     Min      1Q  Median      3Q     Max \n",
    " -0.4990 -0.1784 -0.0146  0.1250  0.8624 \n",
    " \n",
    " Coefficients:\n",
    "                       Estimate Std. Error t value Pr(>|t|)    \n",
    " (Intercept)             0.9798     0.0354   27.68  < 2e-16 ***\n",
    " TrophicLevelHerbivore   0.0766     0.0397    1.93    0.055 .  \n",
    " TrophicLevelOmnivore    0.1727     0.0398    4.34  2.0e-05 ***\n",
    " GroundDwellingYes       0.2095     0.0288    7.28  4.1e-12 ***\n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 \n",
    " \n",
    " Residual standard error: 0.228 on 255 degrees of freedom\n",
    " Multiple R-squared: 0.212,\tAdjusted R-squared: 0.203 \n",
    " F-statistic: 22.9 on 3 and 255 DF,  p-value: 3.6e-13 \n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Starting at the bottom, <span>summary</span> has again calculated $r^2$\n",
    "for us and also an $F$ statistic for the whole model, which matches the\n",
    "calculation above.\n",
    "\n",
    "The other important bits are the four coefficients. The intercept is now\n",
    "the reference level for two variables: it is the mean for carnivores\n",
    "that are not ground dwelling. We then have differences from this value\n",
    "for being an omnivore or herbivore and for being ground dwelling. There\n",
    "is a big change in genome size associated with ground dwelling and\n",
    "omnivory and both of these have large effects sizes, each introducing\n",
    "about a 20% difference in genome size from the non-ground dwelling\n",
    "carnivores. In contrast, herbivory makes a small difference — about 8%.\n",
    "Because the difference is small and the standard error is large, the $t$\n",
    "value suggests that this difference might arise just be chance. Put\n",
    "another way, it isn’t significant.\n",
    "\n",
    "The table below shows how these four coefficients combine to give the\n",
    "predicted values for each of the group means.\n",
    "\n",
    "$$\\begin{array}{|r|rrr|}\n",
    "\\hline\n",
    "\t       & \\textrm{Carnivore} & \\textrm{Herbivore} & \\textrm{Omnivore}\\\\\n",
    "\\hline\n",
    "\\textrm{Not ground} & {\\it 0.98} = 0.98    & {\\it 0.98  + 0.08} =1.06    & {\\it 0.98 + 0.17} =1.15 \\\\\n",
    "\\textrm{Ground}     & {\\it 0.98 + 0.21} = 1.19    & {\\it 0.98  + 0.08   + 0.21} =1.27   & {\\it 0.98 + 0.17  + 0.21} =1.36\\\\\n",
    "\\hline\n",
    "\\end{array}$$\n",
    "\n",
    "Predicted values\n",
    "----------------\n",
    "\n",
    "Getting the model predictions by hand in this way is tedious and error\n",
    "prone. There is handy function called <span>predict</span> which uses\n",
    "the model directly to calculate values. The default is to give you the\n",
    "prediction for each point in the original data, but you can also ask for\n",
    "specific predictions.\n",
    "\n",
    "The first thing to do is to set up a small data frame containing the\n",
    "explanatory values we want to use. The variable names and the level name\n",
    "have to match <span>*exactly*</span>, so we’ll use the\n",
    "<span>levels</span> function to get the names. We want to look at all\n",
    "six combinations, so we’ll use the <span>rep</span> function to set this\n",
    "up. The <span>each=2</span> option repeats each value twice in\n",
    "succession; the <span>times=3</span> options repeats the whole set of\n",
    "values three times."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data frame of combinations of variables\n",
    "> gd <- rep(levels(mammals$GroundDwelling), times = 3)\n",
    "> print(gd)\n",
    "\n",
    " [1] \"No\"  \"Yes\" \"No\"  \"Yes\" \"No\"  \"Yes\"\n",
    " \n",
    "> tl <- rep(levels(mammals$TrophicLevel), each = 2)\n",
    "> print(tl)\n",
    "\n",
    "  [1] \"Carnivore\" \"Carnivore\" \"Herbivore\" \"Herbivore\" \"Omnivore\"  \"Omnivore\" \n",
    "\n",
    "> predVals <- data.frame(GroundDwelling = gd, TrophicLevel = tl)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have the data frame of values we want, we can use predict. As\n",
    "when we created log values, we can save the output back into a new\n",
    "column in the data frame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> predVals$predict <- predict(model, newdata = predVals)\n",
    "> print(predVals)\n",
    "\n",
    "   GroundDwelling TrophicLevel predict\n",
    " 1             No    Carnivore  0.9798\n",
    " 2            Yes    Carnivore  1.1892\n",
    " 3             No    Herbivore  1.0563\n",
    " 4            Yes    Herbivore  1.2658\n",
    " 5             No     Omnivore  1.1524\n",
    " 6            Yes     Omnivore  1.3619"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "These are in the same order as the bars from your barplot. Make a copy\n",
    "of the barplot and arrows code and then add the code below to generate\n",
    "the plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> points(barMids, predVals$predict, col= ' red ' , pch=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image](barplotpred.pdf){width=\"60.00000%\"}\n",
    "\n",
    "The red points do not match to the calculated means. This is because the\n",
    "model only includes a single difference between ground and non-ground\n",
    "species, which has to be the same for each trophic group. That is, there\n",
    "is no interaction between trophic level and ground / non-ground identity\n",
    "of each species in the current model.\n",
    "\n",
    "The Chapter \\[ch:MulExplInter\\] will look at interactions, which allows\n",
    "these values to differ using an interaction term in the model.\n",
    "\n",
    "Linear Models: Multiple variables and interactions {#ch:MulExplInter}\n",
    "==================================================\n",
    "\n",
    "Aims of this chapter[^10]:\n",
    "\n",
    "Creating more complex models, including ANCOVA\n",
    "\n",
    "Looking at interactions between variables\n",
    "\n",
    "Plotting predictions from models\n",
    "\n",
    "We will look at two models in this chapter:\n",
    "\n",
    "Model 1: Is mammalian genome size predicted by interactions between\n",
    "trophic level and whether species are ground dwelling?\n",
    "\n",
    "ANCOVA: Is body size in Odonata predicted by interactions between genome\n",
    "size and taxonomic suborder?\n",
    "\n",
    "So far, we have only looked at the independent effects of variables. For\n",
    "example, in the trophic level and ground dwelling model from Chapter\n",
    "\\[ch:MulExpl\\], we only looked for specific differences for being a\n",
    "omnivore <span>*or*</span> being ground dwelling, not for being\n",
    "specifically a <span> *ground dwelling omnivore*</span>. These\n",
    "independent effects of a variable are known as <span>*main\n",
    "effects*</span> and the effects of combinations of variables acting\n",
    "together are known as <span>*interactions*</span> — they describe how\n",
    "the variables <span>*interact*</span>.\n",
    "\n",
    "Formulae with interactions in R\n",
    "-------------------------------\n",
    "\n",
    "We’ve already seen a number of different model formulae in R. They all\n",
    "use this syntax:\\\n",
    "<span>response variable \\~ explanatory variable(s)</span>\\\n",
    "but we are now going to add two extra pieces of syntax:\n",
    "\n",
    "The <span>a:b</span> means the interaction between <span>a</span> and\n",
    "<span>b</span> — do combinations of these variables lead to different\n",
    "outcomes?\n",
    "\n",
    "This a shorthand for the model above. The means fit <span>a</span> and\n",
    "<span>b</span> as main effects and their interaction <span>a:b</span>.\n",
    "\n",
    "Model 1: Mammalian genome size\n",
    "------------------------------\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Make sure you have changed the working directory to <span>Code</span> in\n",
    "your stats coursework directory.\n",
    "\n",
    "Create a new blank script called ‘Interactions.R’ and add some\n",
    "introductory comments.\n",
    "\n",
    "Use <span>load(’mammals.Rdata’)</span> to load the data.\n",
    "\n",
    "If <span>mammals.Rdata</span> is missing, just import the data again\n",
    "using <span>read.csv(“../Data/MammalData.csv”)</span>. You will then\n",
    "have to add the log C Value column to the imported data frame again.\n",
    "\n",
    "Let’s refit the model from Chapter \\[ch:MulExpl\\], but including the\n",
    "interaction between trophic level and ground dwelling. We’ll immediately\n",
    "check the model is appropriate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> model <- lm(logCvalue ~ TrophicLevel * GroundDwelling, data= mammals)\n",
    "> par(mfrow=c(2,2), mar=c(3,3,1,1), mgp=c(2, 0.8,0))\n",
    "> plot(model)\t"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This gives:\n",
    "\n",
    "![image](mamMod.pdf){width=\"\\textwidth\"}\n",
    "\n",
    "Now, we’ll examine the <span>anova</span> and <span>summary</span>\n",
    "outputs for the model:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> anova(model)\n",
    "\n",
    " Analysis of Variance Table\n",
    " \n",
    " Response: logCvalue\n",
    "                              Df Sum Sq Mean Sq F value  Pr(>F)    \n",
    " TrophicLevel                  2   0.81   0.407    8.06  0.0004 ***\n",
    " GroundDwelling                1   2.75   2.747   54.40 2.3e-12 ***\n",
    " TrophicLevel:GroundDwelling   2   0.43   0.216    4.27  0.0150 *  \n",
    " Residuals                   253  12.77   0.050                    \n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compared to the model from Chapter \\[ch:MulExpl\\], there is an extra\n",
    "line at the bottom. The top two are the same and show that trophic level\n",
    "and ground dwelling both have independent main effects. The extra line\n",
    "shows that there is also an interaction between the two. It doesn’t\n",
    "explain a huge amount of variation, about half as much as trophic level,\n",
    "but it is significant.\n",
    "\n",
    "Again, we can calculate the $r^2$ for the model: $$\\frac{0.81 + 2.75 + \n",
    "0.43}{0.81+2.75+0.43+12.77} = 0.238$$ The model from Chapter\n",
    "\\[ch:MulExpl\\] without the interaction had an $r^2 = 0.212$ — our new\n",
    "model explains 2.6% more of the variation in the data.\n",
    "\n",
    "The summary table is as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ">summary(model)\n",
    " \n",
    " Call:\n",
    " lm(formula = logCvalue ~ TrophicLevel * GroundDwelling, data = mammals)\n",
    " \n",
    " Residuals:\n",
    "    Min     1Q Median     3Q    Max \n",
    " -0.523 -0.171 -0.010  0.119  0.831 \n",
    " \n",
    " Coefficients:\n",
    "                                         Estimate Std. Error t value Pr(>|t|)    \n",
    " (Intercept)                               0.9589     0.0441   21.76  < 2e-16 ***\n",
    " TrophicLevelHerbivore                     0.0535     0.0554    0.97  0.33460    \n",
    " TrophicLevelOmnivore                      0.2328     0.0523    4.45  1.3e-05 ***\n",
    " GroundDwellingYes                         0.2549     0.0651    3.92  0.00012 ***\n",
    " TrophicLevelHerbivore:GroundDwellingYes   0.0303     0.0786    0.39  0.69979    \n",
    " TrophicLevelOmnivore:GroundDwellingYes   -0.1476     0.0793   -1.86  0.06384 .  \n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 \n",
    " \n",
    " Residual standard error: 0.225 on 253 degrees of freedom\n",
    "   (120 observations deleted due to missingness)\n",
    " Multiple R-squared: 0.238,\tAdjusted R-squared: 0.223 \n",
    " F-statistic: 15.8 on 5 and 253 DF,  p-value: 1.5e-13 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The lines in this are:\n",
    "\n",
    "The reference level (intercept) for non ground dwelling carnivores. (The\n",
    "reference level is decided just by the alphabetic order of the levels)\n",
    "\n",
    "Two differences for being in different trophic levels.\n",
    "\n",
    "One difference for being ground dwelling\n",
    "\n",
    "Two new differences that give specific differences for ground dwelling\n",
    "herbivores and omnivores.\n",
    "\n",
    "The first four lines, as in the model from Chapter \\[ch:ANOVA\\], which\n",
    "would allow us to find the predicted values for each group <span>*if the\n",
    "size of the differences did not vary between levels because of the\n",
    "interactions*</span>. That is, this part of the model only includes a\n",
    "single difference ground and non-ground species, which has to be the\n",
    "same for each trophic group because it ignores interactions between\n",
    "trophic level and ground / non-ground identity of each species. The last\n",
    "two lines then give the estimated coefficients associated with the\n",
    "interaction terms, and allow cause the size of differences to vary\n",
    "between levels because of the further effects of interactions.\n",
    "\n",
    "The table below show how these combine to give the predictions for each\n",
    "group combination, with those two new lines show in red:\n",
    "\n",
    "$$\\begin{array}{|r|r|r|}\n",
    "\\hline\n",
    " & \\textrm{Not ground} &  \\textrm{Ground} \\\\\n",
    "\\hline\n",
    "\\textrm{Carnivore} & 0.96 = 0.96 &  0.96+0.25=1.21 \\\\\n",
    "\\textrm{Herbivore} & 0.96 + 0.05 = 1.01 & 0.96+0.05+0.25{\\color{red}+0.03}=1.29\\\\\n",
    "\\textrm{Omnivore} & 0.96 + 0.23 = 1.19 & 0.96+0.23+0.25{\\color{red}-0.15}=1.29\\\\\n",
    "\\hline\n",
    "\\end{array}$$\n",
    "\n",
    "So why are there two new coefficients? For interactions between two\n",
    "factors, there are always $(n-1)\\times(m-1)$ new coefficients, where $n$\n",
    "and $m$ are the number of levels in the two factors (Ground dwelling or\n",
    "not: 2 levels and trophic level: 3 levels, in our current example). So\n",
    "in this model, $(3-1) \\times (2-1) =2$. It is easier to understand why\n",
    "graphically: the prediction for the white boxes below can be found by\n",
    "adding the main effects together but for the grey boxes we need to find\n",
    "specific differences and so there are $(n-1)\\times(m-1)$ interaction\n",
    "coefficients to add.\n",
    "\n",
    "$$n=4,m=4\\quad\n",
    "\\begin{array}{|c|c|c|c|}\n",
    "\\hline\n",
    " & & & \\\\ \n",
    "\\hline\n",
    " & {\\cellcolor[gray]{0.8}}& {\\cellcolor[gray]{0.8}}&  {\\cellcolor[gray]{0.8}}\\\\ \n",
    "\\hline\n",
    " & {\\cellcolor[gray]{0.8}}& {\\cellcolor[gray]{0.8}}& {\\cellcolor[gray]{0.8}}\\\\ \n",
    "\\hline\n",
    " &  {\\cellcolor[gray]{0.8}}&  {\\cellcolor[gray]{0.8}}& {\\cellcolor[gray]{0.8}}\\\\ \n",
    "\\hline\n",
    "\\end{array}\n",
    "\\quad n=3,m=6 \\quad\n",
    "\\begin{array}{|c|c|c|c|c|c|}\n",
    "\\hline\n",
    " & & & & &\\\\ \n",
    "\\hline\n",
    " & {\\cellcolor[gray]{0.8}}& {\\cellcolor[gray]{0.8}}&  {\\cellcolor[gray]{0.8}}&  {\\cellcolor[gray]{0.8}}&  {\\cellcolor[gray]{0.8}}\\\\ \n",
    "\\hline\n",
    " & {\\cellcolor[gray]{0.8}}& {\\cellcolor[gray]{0.8}}& {\\cellcolor[gray]{0.8}}&  {\\cellcolor[gray]{0.8}}&  {\\cellcolor[gray]{0.8}}\\\\ \n",
    "\\hline\n",
    "\\end{array}$$\n",
    "\n",
    "If we put this together, what is the model telling us?\n",
    "\n",
    "Herbivores have the same genome sizes as carnivores, but omnivores have\n",
    "larger genomes.\n",
    "\n",
    "Ground dwelling mammals have larger genomes.\n",
    "\n",
    "These two findings suggest that ground dwelling omnivores should have\n",
    "extra big genomes. However, the interaction shows they are smaller than\n",
    "expected and are, in fact, similar to ground dwelling herbivores.\n",
    "\n",
    "Note that although the interaction term in the <span>anova</span> output\n",
    "is significant, neither of the two coefficients in the\n",
    "<span>summary</span> has a $p<0.05$. There are two weak differences (one\n",
    "very weak, one nearly significant) that together explain significant\n",
    "variance in the data.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Copy the code above into your script and run the model.\n",
    "\n",
    "Make sure you understand the output!\n",
    "\n",
    "Just to make sure the sums above are correct, we’ll use the same code as\n",
    "in \\[ch:MulExpl\\] to get R to calculate predictions for us:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# a data frame of combinations of variables\n",
    "> gd <- rep(levels(mammals$GroundDwelling), times = 3)\n",
    "> print(gd)\n",
    " [1] \"No\"  \"Yes\" \"No\"  \"Yes\" \"No\"  \"Yes\"\n",
    "\n",
    "> tl <- rep(levels(mammals$TrophicLevel), each = 2)\n",
    "> print(tl)\n",
    "\n",
    " [1] \"Carnivore\" \"Carnivore\" \"Herbivore\" \"Herbivore\" \"Omnivore\"  \"Omnivore\" \n",
    "\n",
    "# New data frame\n",
    "> predVals <- data.frame(GroundDwelling = gd, TrophicLevel = tl)\n",
    "\n",
    "# predict using the new data frame\n",
    "> predVals$predict <- predict(model, newdata = predVals)\n",
    "> print(predVals)\n",
    " \n",
    "    GroundDwelling TrophicLevel predict\n",
    " 1             No    Carnivore  0.9589\n",
    " 2            Yes    Carnivore  1.2138\n",
    " 3             No    Herbivore  1.0125\n",
    " 4            Yes    Herbivore  1.2977\n",
    " 5             No     Omnivore  1.1918\n",
    " 6            Yes     Omnivore  1.2990"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Run these predictions in your script.\n",
    "\n",
    "If we plot these data points onto the barplot from Chapter\n",
    "\\[ch:MulExpl\\], they now lie exactly on the mean values, because we’ve\n",
    "allowed for interactions. The triangle on this plot shows the\n",
    "predictions for ground dwelling omnivores from the main effects\n",
    "($0.96 + 0.23  + 0.25 = 1.44$), the interaction of $-0.15$ pushes the\n",
    "prediction back down.\n",
    "\n",
    "![image](predPlot.pdf){width=\"80.00000%\"}\n",
    "\n",
    "ANCOVA: Body Weight in Odonata\n",
    "------------------------------\n",
    "\n",
    "We’ll go all the way back to the regression analyses from\n",
    "Chapter \\[ch:regress\\]. Remember that we fitted two separate regression\n",
    "lines to the data for damselflies and dragonflies. We’ll now use an\n",
    "interaction to fit these in a single model. This kind of linear model —\n",
    "with a mixture of continuous variables and factors — is often called an\n",
    "<span>*analysis of covariance*</span>, or ANCOVA. That is, ANCOVA is a\n",
    "type of linear model that blends ANOVA and regression. ANCOVA evaluates\n",
    "whether population means of a dependent variable are equal across levels\n",
    "of a categorical independent variable, while statistically controlling\n",
    "for the effects of other continuous variables that are not of primary\n",
    "interest, known as covariates.\n",
    "\n",
    "<span>*That is, this is still a linear model, but with one categorical\n",
    "and one or more continuous predictors*</span>.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Load the data: <span>odonata &lt;-\n",
    "read.csv(’../Data/GenomeSize.csv’)</span>.\n",
    "\n",
    "Create two new variables in the <span>odonata</span> data set called\n",
    "<span>logGS</span> and <span>logBW</span> containing log genome size and\n",
    "log body weight.\n",
    "\n",
    "The models we fitted before looked like this:\n",
    "\n",
    "![image](dragonData.pdf){width=\"70.00000%\"}\n",
    "\n",
    "We can now fit the model of body weight as a function of both genome\n",
    "size and suborder:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> odonModel <- lm(logBW ~ logGS * Suborder, data = odonata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Again, we’ll look at the <span>anova</span> table first:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> anova(odonModel)\n",
    " \n",
    " Analysis of Variance Table\n",
    " \n",
    " Response: logBW\n",
    "                Df Sum Sq Mean Sq F value  Pr(>F)    \n",
    " logGS           1    1.1     1.1    2.71     0.1    \n",
    " Suborder        1  112.0   112.0  265.13 < 2e-16 ***\n",
    " logGS:Suborder  1    9.1     9.1   21.65 1.1e-05 ***\n",
    " Residuals      94   39.7     0.4                    \n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Interpreting this gives the following:\n",
    "\n",
    "There is no significant main effect of log genome size. The\n",
    "<span>*main*</span> effect is the important thing here — genome size is\n",
    "hugely important but does very different things for the two different\n",
    "suborders. If we ignored <span>Suborder</span>, there isn’t an overall\n",
    "relationship: the average of those two lines is pretty much flat.\n",
    "\n",
    "There is a very strong main effect of Suborder: the mean body weight in\n",
    "the two groups are very different.\n",
    "\n",
    "There is a strong interaction between suborder and genome size. This is\n",
    "an interaction between a factor and a continuous variable and shows that\n",
    "the <span>*slopes*</span> are different for the different factor levels.\n",
    "\n",
    "The summary table looks like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> summary(odonModel)\n",
    " \n",
    " Call:\n",
    " lm(formula = logBW ~ logGS * Suborder, data = odonata)\n",
    " \n",
    " Residuals:\n",
    "     Min      1Q  Median      3Q     Max \n",
    " -1.3243 -0.3225  0.0073  0.3962  1.4976 \n",
    " \n",
    " Coefficients:\n",
    "                         Estimate Std. Error t value Pr(>|t|)    \n",
    " (Intercept)              -2.3995     0.0848  -28.31  < 2e-16 ***\n",
    " logGS                     1.0052     0.2237    4.49  2.0e-05 ***\n",
    " SuborderZygoptera        -2.2489     0.1354  -16.61  < 2e-16 ***\n",
    " logGS:SuborderZygoptera  -2.1492     0.4619   -4.65  1.1e-05 ***\n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 \n",
    " \n",
    " Residual standard error: 0.65 on 94 degrees of freedom\n",
    "   (2 observations deleted due to missingness)\n",
    " Multiple R-squared: 0.755,\tAdjusted R-squared: 0.747 \n",
    " F-statistic: 96.5 on 3 and 94 DF,  p-value: <2e-16 \n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first thing to note is that the $r^2$ value is really high. The\n",
    "model explains three quarters (0.752) of the variation in the data.\n",
    "Next, there are four coefficients:\n",
    "\n",
    "The intercept is for the first level of <span>Suborder</span>, which is\n",
    "Anisoptera (dragonflies).\n",
    "\n",
    "The next line, for log genome size, is the slope for Anisoptera.\n",
    "\n",
    "We then have a coefficient for the second level of <span>\n",
    "Suborder</span>, which is Zygoptera (damselflies). As with the first\n",
    "model, this difference in factor levels is a difference in mean values\n",
    "and shows the difference in the intercept for Zygoptera.\n",
    "\n",
    "The last line is the interaction between <span>Suborder</span> and\n",
    "<span>logGS</span>. This shows how the slope for Zygoptera differs from\n",
    "the slope for Anisoptera.\n",
    "\n",
    "How do these hang together to give the two lines shown in the model? We\n",
    "can calculate these by hand: $$\\begin{aligned}\n",
    "\t\\textrm{Body Weight} &= -2.40 + 1.01 \\times \\textrm{logGS} & \\textrm{[Anisoptera]}\\\\\n",
    "\t\\textrm{Body Weight} &= (-2.40 -2.25) + (1.01 - 2.15) \\times \\textrm{logGS} & \\textrm{[Zygoptera]}\\\\\n",
    "\t                     &= -4.65 - 1.14 \\times \\textrm{logGS} \\\\\\end{aligned}$$\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Add the code into your script and check that you understand the outputs.\n",
    "\n",
    "We’ll use the <span>predict</span> function to get the predicted values\n",
    "from the model and add lines to the plot above.\n",
    "\n",
    "First, we’ll create a set of numbers spanning the range of genome size:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the range of the data\n",
    "> rng <- range(odonata$logGS)\n",
    "#get a sequence from the min to the max with 100 equally spaced values\n",
    "> LogGSForFitting <- seq(rng[1], rng[2], length = 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Have a look at these numbers:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\tprint(LogGSForFitting)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now use the model to predict the values of body weight at each of\n",
    "those points for each of the two suborders. We’ve added <span>\n",
    "se.fit=TRUE</span> to the function to get the standard error around the\n",
    "regression lines. Note that we are now using"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get a data frame of new data for the order\n",
    "> ZygoVals <- data.frame(logGS = LogGSForFitting, Suborder = \"Zygoptera\")\n",
    "\n",
    "#get the predictions and standard error\n",
    "> ZygoPred <- predict(odonModel, newdata = ZygoVals, se.fit = TRUE)\n",
    "\n",
    "#repeat for anisoptera\n",
    "AnisoVals <- data.frame(logGS = LogGSForFitting, Suborder = \"Anisoptera\")\n",
    "\n",
    "AnisoPred <- predict(odonModel, newdata = AnisoVals, se.fit = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Both <span>AnisoPred</span> and <span>ZygoPred</span> contain predicted\n",
    "values (called <span>fit</span>) and standard error values (called\n",
    "<span>se.fit</span>) for each of the values in our generated values in\n",
    "<span>LogGSForFitting</span> for each of the two suborders.\n",
    "\n",
    "We can add the predictions onto a plot like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the scatterplot of the data\n",
    "> plot(logBW ~ logGS, data = odonata, col = Suborder)\n",
    "# add the predicted lines\n",
    "> lines(AnisoPred$fit ~ LogGSForFitting, col = \"black\")\n",
    "> lines(AnisoPred$fit + AnisoPred$se.fit ~ LogGSForFitting, col = \"black\", lty = 2)\n",
    "> lines(AnisoPred$fit - AnisoPred$se.fit ~ LogGSForFitting, col = \"black\", lty = 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Copy the prediction code into your script and run the plot above. Copy\n",
    "and modify the last three lines to add the lines for the Zygoptera. Your\n",
    "final plot should look like this.\n",
    "\n",
    "![image](odonPlot.pdf){width=\"70.00000%\"}\n",
    "\n",
    "Linear Models: Model simplification {#chap:ModelSimp}\n",
    "===================================\n",
    "\n",
    "Aims of this chapter[^11]:\n",
    "\n",
    "Simplifying complex models by removing non-explanatory terms\n",
    "\n",
    "In biology, we often use statistics to compare competing hypotheses in\n",
    "order to work out the simplest explanation for some data. This often\n",
    "involves collecting several explanatory variables that describe\n",
    "different hypotheses and then fitting them together in a single model,\n",
    "and often including interactions between those variables.\n",
    "\n",
    "In all likelihood, not all of these model <span>*terms*</span> will be\n",
    "important. If we remove unimportant terms, then the explanatory power of\n",
    "the model will get worse, but might not get significantly worse.\n",
    "\n",
    "> *“It can scarcely be denied that the supreme goal of all theory is to\n",
    "> make the irreducible basic elements as simple and as few as possible\n",
    "> without having to surrender the adequate representation of a single\n",
    "> datum of experience.”*\n",
    ">\n",
    "> *Albert Einstein*\n",
    "\n",
    "Or to paraphrase:\n",
    "\n",
    "> *“Everything should be made as simple as possible, but no simpler.”*\n",
    "\n",
    "The approach we will look at is to start with a <span>*maximal\n",
    "model*</span> — the model that contains everything that might be\n",
    "important — and simplify it towards the <span>*null model*</span> — the\n",
    "model that says that none of your variables are important. Hopefully,\n",
    "there is a point somewhere in between where you can’t remove any further\n",
    "terms without making the model significantly worse: this is called the\n",
    "<span>*minimum adequate model*</span>.\n",
    "\n",
    "\\[every node/.style=<span>draw=black!70, very thick, rectangle, rounded\n",
    "corners</span>\\] (a) at (0,0) <span>Maximal model</span>; (b) at (5,0)\n",
    "<span>Minimum adequate model</span>; (c) at (10,0) <span>Null\n",
    "model</span>; (a.east) – (b.west); (b.east) – (c.west);\n",
    "\n",
    "A maximal model\n",
    "---------------\n",
    "\n",
    "We’ll be using the mammal dataset for this practical, so once again:\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Make sure you have changed the working directory to your stats module\n",
    "<span>Code</span> folder.\n",
    "\n",
    "Create a new blank script called ‘MyModelSimp.R’.\n",
    "\n",
    "Load the mammals data into a data frame called <span>mammals</span>.\n",
    "\n",
    "In Chapters 5 & \\[ch:MulExplInter\\], we looked at how the categorical\n",
    "variables <span>GroundDwelling</span> and <span>TrophicLevel</span>\n",
    "predicted genome size in mammals. In this chapter, we will add in two\n",
    "more continuous variables: litter size and body mass. The first thing we\n",
    "will do is to log both variables and reduce the dataset to the rows for\n",
    "which all of these data are available:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get logs of continuous variables\n",
    "> mammals$logLS <- log(mammals$LitterSize)\n",
    "> mammals$logCvalue <- log(mammals$meanCvalue)\n",
    "> mammals$logBM <- log(mammals$AdultBodyMass_g)\n",
    "\n",
    "# reduce dataset to five key variables\n",
    "> mammals <- subset(mammals, select = c(logCvalue, logLS, logBM, \n",
    "TrophicLevel, GroundDwelling))\n",
    "\n",
    "# remove the row with missing data\n",
    "> mammals <- na.omit(mammals)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Copy the code above into your script and run it\n",
    "\n",
    "Check that the data you end up with has this structure:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'data.frame':\t240 obs. of  5 variables:\n",
    "$logCvalue     : num  0.94 1.322 1.381 1.545 0.888 ...\n",
    "$logLS         : num  1.1 1.12 0 0 1.52 ...\n",
    "$logBM         : num  10.83 4.87 11.46 10.86 3.23 ...\n",
    "$TrophicLevel  : Factor w/ 3 levels \"Carnivore\",\"Herbivore\",..: 1 2 2 2 3 3 3 2 2 3 ...\n",
    "$GroundDwelling: Factor w/ 2 levels \"No\",\"Yes\": 2 2 2 2 2 1 2 1 1 1 ...\n",
    " - attr(*, \"na.action\")=Class 'omit'  Named int [1:139] 2 4 7 9 10 11 14 15 20 21 ...\n",
    "  .. ..- attr(*, \"names\")= chr [1:139] \"2\" \"4\" \"7\" \"9\" ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we’ll fit a model including all of these variables and all of the\n",
    "interactions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> model <- lm(formula = logCvalue ~ logLS * logBM * TrophicLevel * \n",
    "GroundDwelling, data = mammals)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Run this model in your script.\n",
    "\n",
    "Look at the output of <span>anova(model)</span> and <span>\n",
    "summary(model)</span>.\n",
    "\n",
    "Scared? Don’t be! There are a number of points to this exercise:\n",
    "\n",
    "1.  These tables show exactly the kind of output you’ve seen before.\n",
    "    Sure, there are lots of rows but each row is just asking whether a\n",
    "    model term (<span>anova</span>) or a model coefficient (<span>\n",
    "    summary</span>) is significant.\n",
    "\n",
    "2.  Some of the rows are significant, others aren’t: some of the model\n",
    "    terms are not explanatory.\n",
    "\n",
    "3.  The two tables show slightly different things - lots of stars for\n",
    "    the <span>anova</span> table and only a few for the\n",
    "    <span>summary</span> table.\n",
    "\n",
    "4.  That last line in the <span>anova</span> table: <span>\n",
    "    logLS:logBM:TrophicLevel:GroundDwelling</span>. This is an\n",
    "    interaction of four variables capturing how the slope for litter\n",
    "    size changes for different body masses for species in different\n",
    "    trophic groups and which are arboreal or ground dwelling. Does this\n",
    "    seem easy to understand?\n",
    "\n",
    "The real lesson here is that it is easy to fit complicated models in R.\n",
    "<span>*Understanding and explaining them is a different matter*</span>.\n",
    "The temptation is always to start with the most complex possible model\n",
    "but this is rarely a good idea.\n",
    "\n",
    "A better maximal model\n",
    "----------------------\n",
    "\n",
    "Instead of all possible interactions, we’ll consider two-way\n",
    "interactions: how do pairs of variables affect each other? There is a\n",
    "shortcut for this: <span>y \\~ (a + b + c)\\^2</span> gets all two way\n",
    "combinations of the variables in the brackets, so is a quicker way of\n",
    "getting this model:\n",
    "\n",
    "<span>y \\~ a + b + c + a:b + a:c + b:c</span>.\n",
    "\n",
    "So let’s use this to fit a simpler maximal model:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> model <- lm(logCvalue ~ (logLS + logBM + TrophicLevel + GroundDwelling)^2, data = mammals)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The <span>anova</span> table for this model looks like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> anova(model)\n",
    "\n",
    " Analysis of Variance Table\n",
    " \n",
    " Response: logCvalue\n",
    "                              Df Sum Sq Mean Sq F value  Pr(>F)    \n",
    " logLS                         1   0.99   0.989   25.72 8.2e-07 ***\n",
    " logBM                         1   3.03   3.032   78.83 < 2e-16 ***\n",
    " TrophicLevel                  2   0.48   0.239    6.21  0.0024 ** \n",
    " GroundDwelling                1   0.11   0.110    2.87  0.0915 .  \n",
    " logLS:logBM                   1   0.27   0.275    7.15  0.0081 ** \n",
    " logLS:TrophicLevel            2   0.19   0.095    2.48  0.0862 .  \n",
    " logLS:GroundDwelling          1   0.14   0.136    3.55  0.0609 .  \n",
    " logBM:TrophicLevel            2   0.09   0.044    1.14  0.3230    \n",
    " logBM:GroundDwelling          1   0.88   0.883   22.96 3.0e-06 ***\n",
    " TrophicLevel:GroundDwelling   2   0.04   0.022    0.58  0.5607    \n",
    " Residuals                   225   8.65   0.038                    \n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first lines are the <span>*main effects*</span>, which are all\n",
    "significant or near significant. Then there are the six interactions.\n",
    "One of these is very significant: <span>logBM:GroundDwelling</span>,\n",
    "which suggests that the slope of log C value with body mass differs\n",
    "between ground dwelling and non-ground dwelling species. The other\n",
    "interactions are non-significant although some are close.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Run this model in your script.\n",
    "\n",
    "Look at the output of <span>anova(model)</span> and <span>\n",
    "summary(model)</span>.\n",
    "\n",
    "Check the model diagnostic plots.\n",
    "\n",
    "Model simplification\n",
    "--------------------\n",
    "\n",
    "Model simplification is not a simple process. Each time you remove a\n",
    "term from a model, the model will change: the model will get worse,\n",
    "since some of the sums of squares are no longer explained, but the\n",
    "remaining variables may take over.\n",
    "\n",
    "The first question is: <span>*what terms can you remove from a\n",
    "model*</span>? Obviously, you only want to remove non-significant terms,\n",
    "but there is another rule – you cannot remove a main effect or an\n",
    "interaction while those main effects or interactions are present in a\n",
    "more complex interaction. For example, in the model <span>y \\~ a + b + c\n",
    "+ a:b + a:c + b:c</span>, you cannot drop <span>c</span> without\n",
    "dropping both <span>a:c</span> and <span>b:c</span>.\n",
    "\n",
    "The R function <span>drop.scope</span> tells you what you can drop from\n",
    "a model. Some examples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> drop.scope(model)\n",
    " [1] \"logLS:logBM\"                 \"logLS:TrophicLevel\"         \n",
    " [3] \"logLS:GroundDwelling\"        \"logBM:TrophicLevel\"         \n",
    " [5] \"logBM:GroundDwelling\"        \"TrophicLevel:GroundDwelling\"\n",
    "\n",
    "> drop.scope(y ~ a + b + c + a:b)\n",
    " [1] \"c\"   \"a:b\"\n",
    "> drop.scope(y ~ a + b + c + a:b + b:c + a:b:c)\n",
    " [1] \"a:b:c\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Model simplification is an iterative process. The flow diagram below\n",
    "shows how it works: at each stage you try and find an acceptable\n",
    "simplification. If successful, then you start again with the new simpler\n",
    "model and try and find a way to simplify this, until eventually, you\n",
    "can’t find anything more to remove.\n",
    "\n",
    "\\[bx/.style=<span>draw=black!70, very thick, rectangle, align=center,\n",
    "rounded corners, fill=black!10, minimum width=4 cm</span>, every\n",
    "path/.style=<span>draw=black!70, very thick, -stealth, shorten &gt;=\n",
    "1mm</span>, node distance=0.8cm\\] (a) at (0,0) <span>**Maximal\n",
    "model**</span>; (b) <span>Current model</span>; (b2) <span>Make list of\n",
    "valid\\\n",
    "terms to drop</span>; (c) <span>Remove the least-\\\n",
    "significant valid term</span>; (d) <span>New model</span>; (e)\n",
    "<span>Compare current\\\n",
    "and new models</span>; (f) <span>Is the new\\\n",
    "model significantly\\\n",
    "worse?</span>; (g) <span>Remove current\\\n",
    "term from list\\\n",
    "of valid terms</span>; (h) <span>Are there\\\n",
    "other valid terms\\\n",
    "to drop?</span>; (i) <span>**Minimum adequate model**</span>; (j)\n",
    "<span>New model becomes\\\n",
    "current model</span>;\n",
    "\n",
    "(a.south) – (b.north); (b.south) – (b2.north); (b2.south) – (c.north);\n",
    "(c.south) – (d.north); (d.south) – (e.north); (e.south) – (f.north);\n",
    "(f.west) -| node\\[below\\]<span>No</span> (j.south) ; (j.north) |-\n",
    "(b.west) ; (f.east) – node\\[below\\]<span>Yes</span> (g.west); (g.north)\n",
    "– (h.south); (h.north) |- node\\[right,\n",
    "pos=0.5\\]<span>Yes</span>(c.east); (h.east) –\n",
    "node\\[below\\]<span>No</span> ++(1,0) |- (i.east);\n",
    "\n",
    "As always, we can use an $F$ test to compare two models and see if they\n",
    "have significantly different explanatory power. Here, significance is a\n",
    "bad thing — it means that we’ve removed a term that makes the model\n",
    "significantly worse.\n",
    "\n",
    "The last thing we need to do is work out how to remove a term from a\n",
    "model. We could type out the model again, but there is a shortcut using\n",
    "the function <span>update</span>:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# a simple model\n",
    "> f <- y ~ a + b + c + b:c\n",
    "\n",
    "# remove b:c from the current model\n",
    "> update(f, . ~ . - b:c)\n",
    " y ~ a + b + c\n",
    " \n",
    "# model g as a response using the same explanatory variables.\n",
    "> update(f, g ~ .)\n",
    " g ~ a + b + c + b:c"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Yes, the syntax is a little odd. The function uses a model or a formula\n",
    "and then allows you to alter the current formula. The dots in the code\n",
    "<span>. \\~ . </span> mean ‘use whatever is currently in the response or\n",
    "explanatory variables’. It gives a simple way of changing a model.\n",
    "\n",
    "Putting this together, let’s try a simplification. From the previous\n",
    "<span>anova</span> and <span>drop.scope</span> output, we know that the\n",
    "interaction <span>TrophicLevel:GroundDwelling</span> is not significant\n",
    "and a valid term."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove TrophicLevel:GroundDwelling\n",
    "> model2 <- update(model, . ~ . - TrophicLevel:GroundDwelling)\n",
    "\n",
    "# use anova to compare the two models\n",
    "> anova(model, model2)\n",
    "\n",
    " Analysis of Variance Table\n",
    " \n",
    " Model 1: logCvalue ~ (logLS + logBM + TrophicLevel + GroundDwelling)^2\n",
    " Model 2: logCvalue ~ logLS + logBM + TrophicLevel + GroundDwelling + \n",
    "\t\t\tlogLS:logBM + logLS:TrophicLevel + logLS:GroundDwelling + \n",
    "\t\t\tlogBM:TrophicLevel + logBM:GroundDwelling\n",
    "   Res.Df  RSS Df Sum of Sq    F Pr(>F)\n",
    " 1    225 8.65                         \n",
    " 2    227 8.70 -2   -0.0446 0.58   0.56"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This tells us that <span>model2</span> is not significantly worse than\n",
    "<span> model</span>. We can now look at this model and see what else can\n",
    "be removed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> anova(model2)\n",
    " \n",
    " Analysis of Variance Table\n",
    " \n",
    " Response: logCvalue\n",
    "                       Df Sum Sq Mean Sq F value  Pr(>F)    \n",
    " logLS                  1   0.99   0.989   25.82 7.8e-07 ***\n",
    " logBM                  1   3.03   3.032   79.12 < 2e-16 ***\n",
    " TrophicLevel           2   0.48   0.239    6.24  0.0023 ** \n",
    " GroundDwelling         1   0.11   0.110    2.88  0.0909 .  \n",
    " logLS:logBM            1   0.27   0.275    7.17  0.0079 ** \n",
    " logLS:TrophicLevel     2   0.19   0.095    2.49  0.0854 .  \n",
    " logLS:GroundDwelling   1   0.14   0.136    3.56  0.0604 .  \n",
    " logBM:TrophicLevel     2   0.09   0.044    1.14  0.3216    \n",
    " logBM:GroundDwelling   1   0.88   0.883   23.05 2.9e-06 ***\n",
    " Residuals            227   8.70   0.038                    \n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 \n",
    " \n",
    "> drop.scope(model2)\n",
    " \n",
    " [1] \"logLS:logBM\"          \"logLS:TrophicLevel\"   \"logLS:GroundDwelling\"\n",
    " [4] \"logBM:TrophicLevel\"   \"logBM:GroundDwelling\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Run this first simplification in your script.\n",
    "\n",
    "Look at the output above and decide what is the next possible term to\n",
    "delete\n",
    "\n",
    "Using the code above as a model, create <span>model3</span> as the next\n",
    "simplification! (remember to use <span>model2</span> in your update call\n",
    "and not <span>model</span>)\n",
    "\n",
    "Now for a difficult exercise:\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Using the code above to guide you, try and find a minimal adequate model\n",
    "that you are happy with. In each step, the output of <span>anova(model,\n",
    "modelN)</span> should be non-significant (where $N$ is the current\n",
    "step).\n",
    "\n",
    "It can be important to consider both <span>anova</span> and <span>\n",
    "summary</span> tables. It can be worth trying to remove things that look\n",
    "significant in one table but not the other — some terms can explain\n",
    "significant variation on the <span>anova</span> table but the\n",
    "coefficients are not significant.\n",
    "\n",
    "Remember to remove <span>*terms*</span>: with categorical variables,\n",
    "several coefficients in the <span>summary</span> table may come from one\n",
    "term in the model and have to be removed together.\n",
    "\n",
    "When you have got your final model, save the model as an R data file:\\\n",
    "<span>save(modelN, file=’myFinalModel.Rda’)</span>.\n",
    "\n",
    "Generalised Linear Models {#chap:GLM}\n",
    "=========================\n",
    "\n",
    "Aims of this chapter[^12]:\n",
    "\n",
    "Use generalised linear models (GLMs) to handle count data.\n",
    "\n",
    "Analyse some genetics practical data.\n",
    "\n",
    "This chapter will step through the analysis carefully. These are not\n",
    "simple analyses so you should concentrate on understanding the process\n",
    "and the biology and think about how to present your results.\n",
    "\n",
    "What is a GLM?\n",
    "--------------\n",
    "\n",
    "the generalized linear model (GLM) is a generalization of ordinary\n",
    "linear regression analyses to accommodate response variables to have\n",
    "non-normal error distributions (e.g., count data, as in the genetics\n",
    "practical data — see below).\n",
    "\n",
    "The data\n",
    "--------\n",
    "\n",
    "We will use mutation data collected by a previous year’s batch in the\n",
    "Genetics Practical. So let’s actually use some of the skills you’ve\n",
    "learned to do some statistical modelling on data you might collect. That\n",
    "is, you can aim to repeat these analyses with similar data you collect.\n",
    "\n",
    "The students were basically counting colonies looking for mutations.\n",
    "There were a number of bacterial strains which were different mutants of\n",
    "<span>*Salmonella*</span>. Each group applied a mutagen Nitroguanisine\n",
    "(NG) as well as histidine and streptomycine. A control plate was also\n",
    "tested.\n",
    "\n",
    "The data file in CSV format is available from the bitbucket site, as\n",
    "usual in the <span>Data</span> directory. It is called\n",
    "<span>PracData.csv</span>.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Save the <span>PracData.csv</span> dataset into your <span>Data</span>\n",
    "directory.\n",
    "\n",
    "Create a new script called MyGLM in your <span>Code</span> directory.\n",
    "Use the code below to load and check your data.\n",
    "\n",
    "Start R and change the working directory to <span>Code</span>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> colonies <- read.csv(\"../Data/PracData.csv\")\n",
    "> str(colonies)\n",
    "\n",
    " 'data.frame':\t680 obs. of  5 variables:\n",
    " $Student.ID : Factor w/ 34 levels \"A1\",\"A10\",\"A11\",..: 1 1 1 2 2 2 4 4 4 4 ...\n",
    " $Strain     : Factor w/ 5 levels \"421\",\"712\",\"881\",..: 4 3 5 1 2 3 4 2 3 5 ...\n",
    " $Treatment  : Factor w/ 4 levels \"Control\",\"His\",..: 1 1 1 3 3 1 1 3 3 1 ...\n",
    " $ColonyCount: int  0 0 0 0 0 0 0 0 0 0 ...\n",
    " $HaloLawn   : Factor w/ 2 levels \"N\",\"Y\": NA NA NA NA NA NA NA NA NA NA ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we’ve got the data loaded, we need to look at it and try and\n",
    "see what is going on.\n",
    "\n",
    "Plotting the data\n",
    "-----------------\n",
    "\n",
    "We have a continuous response variable (<span>ColonyCount</span>) and\n",
    "two categorical explanatory variables (<span>Strain</span> and\n",
    "<span>Treatment</span>). We also have observations of halos and\n",
    "bacterial lawns around the treated areas (<span>HaloLawn</span>), which\n",
    "we will come back to at the end of this chapter.\n",
    "\n",
    "So, with two factors as the explanatory variables, we will use box and\n",
    "whisker plots and boxplots to explore the data. First, we’ll look at the\n",
    "effects of the four treatments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> boxplot(ColonyCount ~ Treatment, data=colonies)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image](Bxp.pdf){width=\"60.00000%\"}\n",
    "\n",
    "There are two immediate things to note.\n",
    "\n",
    "1.  The distributions of colony counts are very <span>*skewed*</span> —\n",
    "    many small counts and a few large counts. We’ve already seen that\n",
    "    taking a log of data sometimes works in these cases. However, as the\n",
    "    tables above show, we have zero counts for all treatments and\n",
    "    $\\log(0)$ is undefined. A common trick is therefore to use\n",
    "    $\\log(n+1)$ (add 1 and take a log) when dealing with count data like\n",
    "    this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> colonies$logCC <- log(colonies$ColonyCount + 1)\n",
    "> boxplot(logCC ~ Treatment, data=colonies)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image](logBxp.pdf){width=\"60.00000%\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "I hope you’ll agree that this still doesn’t look very convincingly\n",
    "like normal data, but we’ll come back to this point."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2.  The colony counts are vastly different between the\n",
    "    different treatments. It is hard to say for sure from the two plots,\n",
    "    but it looks like colonies never grow under the histidine and\n",
    "    streptomycine treatments. We can check that:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> tapply(colonies$ColonyCount, colonies$Treatment, min, na.rm = TRUE)\n",
    "\n",
    " Control     His      NG   Strep \n",
    "       0       0       0       0 \n",
    "\n",
    "> tapply(colonies$ColonyCount, colonies$Treatment, max, na.rm = TRUE)\n",
    "\n",
    " Control     His      NG   Strep \n",
    "    2400       0    2367       0 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is indeed no variation at all in colony count for histidine\n",
    "    and streptomycine — colonies never grow in these treatments. We\n",
    "    don’t really need statistics for this observation and, in fact,\n",
    "    variation is needed for statistics to work. So, for the rest of this\n",
    "    analysis, we will reduce the dataset to the control and\n",
    "    nitroguanisine treatments.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Update your script to contain the code for these plots.\n",
    "\n",
    "We’ll use a new piece of code here to get the right subset. <span>var\n",
    "%in% c(’a’,’b’,’c’)</span> finds all entries in <span>var</span> whose\n",
    "values are equal to <span>’a’</span>, <span>’b’</span> or\n",
    "<span>’c’</span>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> coloniesCN <- subset(colonies, Treatment %in% c(\"Control\", \"NG\"), \n",
    "drop = TRUE) \n",
    "> str(coloniesCN)\n",
    "\n",
    " 'data.frame':\t340 obs. of  6 variables:\n",
    " $Student.ID : Factor w/ 34 levels \"A1\",\"A10\",\"A11\",..: 1 1 1 2 2 2 4 4 4 4 ...\n",
    " $Strain     : Factor w/ 5 levels \"421\",\"712\",\"881\",..: 4 3 5 1 2 3 4 2 3 5 ...\n",
    " $Treatment  : Factor w/ 4 levels \"Control\",\"His\",..: 1 1 1 3 3 1 1 3 3 1 ...\n",
    " $ColonyCount: int  0 0 0 0 0 0 0 0 0 0 ...\n",
    " $HaloLawn   : Factor w/ 2 levels \"N\",\"Y\": NA NA NA NA NA NA NA NA NA NA ...\n",
    " $logCC      : num  0 0 0 0 0 0 0 0 0 0 ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You’ll see that, although we have removed two treatments, their names\n",
    "still appear in the list of levels in the <span>str</span> output. R\n",
    "retains a list of all the levels that were originally in a factor, even\n",
    "when those levels aren’t used any more. This will be annoying later, so\n",
    "we’ll use the <span>droplevels</span> function to strip them out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> coloniesCN <- droplevels(coloniesCN)\n",
    "> str(coloniesCN)\n",
    "\n",
    " 'data.frame':\t340 obs. of  6 variables:\n",
    " $Student.ID : Factor w/ 34 levels \"A1\",\"A10\",\"A11\",..: 1 1 1 2 2 2 4 4 4 4 ...\n",
    " $Strain     : Factor w/ 5 levels \"421\",\"712\",\"881\",..: 4 3 5 1 2 3 4 2 3 5 ...\n",
    " $Treatment  : Factor w/ 2 levels \"Control\",\"NG\": 1 1 1 2 2 1 1 2 2 1 ...\n",
    " $ColonyCount: int  0 0 0 0 0 0 0 0 0 0 ...\n",
    " $HaloLawn   : Factor w/ 0 levels: NA NA NA NA NA NA NA NA NA NA ...\n",
    " $logCC      : num  0 0 0 0 0 0 0 0 0 0 ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Add these commands to subset your data to your code file.\n",
    "\n",
    "Looking at strains too\n",
    "----------------------\n",
    "\n",
    "Now we’ll look to see how counts differ between the strains. A simple\n",
    "way to visualise this is to use the <span>lattice</span> package again\n",
    "to get plots grouped by treatment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> library(lattice) \n",
    "> bwplot(logCC ~ Strain | Treatment, data=coloniesCN)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image](GLM_lattice.pdf){width=\"\\textwidth\"}\n",
    "\n",
    "First impressions from this figure:\n",
    "\n",
    "1.  The strains are doing <span>*very*</span> different things under the\n",
    "    two treatments. Hopefully this now leaps out at you as suggesting\n",
    "    that the two variables (Strain and Treatment) are\n",
    "    <span>*interacting*</span>.\n",
    "\n",
    "2.  The distributions are still pretty ugly — the variances differ\n",
    "    hugely between combinations and four combinations have a median\n",
    "    of zero.\n",
    "\n",
    "We could also use a barplot of means here. We’ll use the original data\n",
    "to get the means, but can use a log scale on the $y$ axis (<span>\n",
    "log=’y’</span>)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> tab <- tapply(coloniesCN$ColonyCount, list(coloniesCN$Treatment,\n",
    "coloniesCN$Strain), mean, na.rm=TRUE)\n",
    "> print(tab)\n",
    "\n",
    "            421     712    881     899 TA102\n",
    " Control 538.20 138.867  12.73   0.375 126.7\n",
    " NG       61.29   5.517 292.71 593.000 523.9"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "An then,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> barplot(tab, beside=TRUE, log= ' y ' )\t"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Which should give,\n",
    "\n",
    "![image](GLM_barplot.pdf){width=\"60.00000%\"}\n",
    "\n",
    "Lets have a look at a first model.\n",
    "\n",
    "A linear model\n",
    "--------------\n",
    "\n",
    "We’ll fit a model of colony count as the interaction between strain and\n",
    "treatment and then look at the diagnostic plots. We’d do this anyway,\n",
    "but we’re already suspicious about the variance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> modLM <- lm(logCC ~ Strain * Treatment, data=coloniesCN)\n",
    "> par(mfrow=c(2,2), mar=c(3,3,3,1), mgp=c(2,0.8,0))\n",
    "> plot(modLM)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image](lmDisaster.pdf){width=\"\\textwidth\"}\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Run this code and have a close look at the plots.\n",
    "\n",
    "That normal QQ plot is not good. Our suspicions were justified and it\n",
    "doesn’t look like we can use a simple log transformation. We’re not even\n",
    "going to look at the <span>anova</span> and <span>summary</span> tables\n",
    "— if the diagnostic plots are bad enough, then the model outputs are not\n",
    "to be trusted.\n",
    "\n",
    "Generalised linear models {#generalised-linear-models}\n",
    "-------------------------\n",
    "\n",
    "In the linear models lecture, we looked at the expectation of <span>\n",
    "*constant normal variance*</span> in linear models. Whatever the\n",
    "combination of explanatory variables for a particular prediction, the\n",
    "residuals around that prediction have similar variance and are roughly\n",
    "normally distributed. The panel on the left shows this basic idea.\n",
    "\n",
    "![image](GLMexample.pdf){width=\"\\textwidth\"}\n",
    "\n",
    "As we have seem, count data does not have this distribution, even when\n",
    "logged. The panel on the right shows the expected distribution of count\n",
    "data as the mean count increases with an explanatory variable. There are\n",
    "three key differences between the two panels:\n",
    "\n",
    "1.  Counts can <span>*never*</span> be negative but can be zero.\n",
    "\n",
    "2.  Counts are always <span>*integers*</span> — whole numbers — rather\n",
    "    than being continuous.\n",
    "\n",
    "3.  The variance of count data is <span>*not constant*</span>. As the\n",
    "    average predicted count gets larger, so does the variance. Unlike\n",
    "    the normal distribution, where variance can take any value, for\n",
    "    count data the variance is expected to be equal to the mean.\n",
    "\n",
    "So, we have data that is unsuitable for a linear model because it\n",
    "doesn’t show constant normal variance. This is where generalised linear\n",
    "models come in — we can change the model for the expected residuals to\n",
    "use a different distribution. For count data, this is the <span>\n",
    "*Poisson*</span> distribution.\n",
    "\n",
    "We need to change the function we use to fit models to <span>glm</span>,\n",
    "but otherwise the process is very similar. The whole point of the GLM is\n",
    "to model the original count data more appropriately, so we will abandon\n",
    "the logged data too. GLMs can cope with a range of different\n",
    "distributions, so we have to specify the <span>family</span> of the\n",
    "distribution we want to use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> modPois <- glm(ColonyCount ~ Strain * Treatment, data=coloniesCN, \n",
    "family= 'poisson')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we’ll look at the summary table for this model. We have 5 levels\n",
    "of strain and 2 levels of factor in the subset so we get an intercept\n",
    "($i$), 4 differences for strains($s_{2-5}$), one difference for\n",
    "treatment ($t_2$) and then four differences for the interaction\n",
    "($s_{2-5}t_2$). These combine like this:\n",
    "\n",
    "$$\\begin{array}{|l|c|c|}\n",
    "\\hline\n",
    "\t\t\t\t\t\t\t& \\textrm{Control} & \\textrm{Nitroguanisine} \\\\\n",
    "\\hline\n",
    "\\textrm{421}   & i        & i+t_2              \\\\\n",
    "\\textrm{712}   & i + s_2  & i+s_2+t_2+ s_2t_2  \\\\\n",
    "\\textrm{881}   & i + s_3  & i+s_3+t_2 + s_3t_2  \\\\\n",
    "\\textrm{889}   & i + s_4  & i+s_4+t_2 + s_4t_2  \\\\\n",
    "\\textrm{TA102} & i + s_5  & i+s_5+t_2  + s_5t_2 \\\\\n",
    "\\hline\n",
    "\\end{array}$$\n",
    "\n",
    "The summary table looks like this — very similar to the\n",
    "<span>summary</span> table for a linear model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> summary(modPois)\n",
    " \n",
    " Call:\n",
    " glm(formula = ColonyCount ~ Strain * Treatment, family = \"poisson\", \n",
    "     data = coloniesCN)\n",
    " \n",
    " Deviance Residuals: \n",
    "    Min      1Q  Median      3Q     Max  \n",
    " -32.37  -10.33   -3.32    0.84   97.84  \n",
    " \n",
    " Coefficients:\n",
    "                         Estimate Std. Error z value Pr(>|z|)    \n",
    " (Intercept)              6.28823    0.00787   799.0   <2e-16 ***\n",
    " Strain712               -1.35472    0.01738   -78.0   <2e-16 ***\n",
    " Strain881               -3.74421    0.05549   -67.5   <2e-16 ***\n",
    " Strain899               -7.26906    0.28878   -25.2   <2e-16 ***\n",
    " StrainTA102             -1.44651    0.01757   -82.3   <2e-16 ***\n",
    " TreatmentNG             -2.17268    0.02539   -85.6   <2e-16 ***\n",
    " Strain712:TreatmentNG   -1.05295    0.08446   -12.5   <2e-16 ***\n",
    " Strain881:TreatmentNG    5.30786    0.06152    86.3   <2e-16 ***\n",
    " Strain899:TreatmentNG    9.53871    0.28989    32.9   <2e-16 ***\n",
    " StrainTA102:TreatmentNG  3.59226    0.03090   116.2   <2e-16 ***\n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 \n",
    " \n",
    " (Dispersion parameter for poisson family taken to be 1)\n",
    " \n",
    "     Null deviance: 134445  on 293  degrees of freedom\n",
    " Residual deviance:  61579  on 284  degrees of freedom\n",
    "   (46 observations deleted due to missingness)\n",
    " AIC: 62910\n",
    " \n",
    " Number of Fisher Scoring iterations: 7\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, interpreting this table quickly. Under the control treatment, strain\n",
    "421 (the intercept) has the highest number of colonies and all the other\n",
    "strains have lower numbers to some degree — the differences are\n",
    "negative. The overall effect of nitrogaunasine is to decrease the number\n",
    "of colonies — again a negative coefficient — but then the positive\n",
    "interactions show big increases in colony counts for nitroguanisine for\n",
    "specific strains. Everything is hugely significant.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Copy the code in this section into your script and explore the model.\n",
    "\n",
    "Overdispersion\n",
    "--------------\n",
    "\n",
    "There’s a problem. You may have already spotted it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> par(mfrow = c(2, 2), mar = c(3, 3, 3, 1), mgp = c(2, 0.8, 0))\n",
    "> plot(modPois)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image](poisDiag.pdf){width=\"80.00000%\"}\n",
    "\n",
    "Actually, there are two problems. First, that QQ plot is still a bit\n",
    "dubious. More of the points are close to the line than in the linear\n",
    "model but there are some extreme positive residuals. Second, the\n",
    "magnitude of the residuals is enormous, and this is really clear in the\n",
    "plot in the bottom right hand corner. This plot identifies outliers and\n",
    "any points outside of the red dotted line are possible problems.\n",
    "\n",
    "The problem here is <span>*overdispersion*</span>. The Poisson\n",
    "distribution predicts that the variance at a point in the model is equal\n",
    "to the prediction — the mean count at that point. Our count data shows\n",
    "much more variance than this — particularly that there are some huge\n",
    "counts given the means.\n",
    "\n",
    "There is a simple way to check the dispersion of count data using the\n",
    "<span>summary</span> table: the ratio of the residual deviance to the\n",
    "residual degrees of freedom should be approximately 1. This expectation\n",
    "is actually given in the table:\n",
    "\n",
    "<span>(Dispersion parameter for poisson family taken to be 1)</span>\n",
    "\n",
    "In this case, the ratio is $61579/284=216.8$. That’s very strongly\n",
    "overdispersed. Fortunately, we can allow for this by using a different\n",
    "model.\n",
    "\n",
    "Generalised linear models using quasipoisson\n",
    "--------------------------------------------\n",
    "\n",
    "The quasipoisson family uses the data to estimate the dispersion of the\n",
    "model, but is otherwise very similar to using the Poisson family."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> modQPois <- glm(ColonyCount ~ Strain * Treatment, data=coloniesCN,\n",
    "family= 'quasipoisson')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The summary table now looks like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> summary(modQPois) \n",
    "\n",
    " Call:\n",
    " glm(formula = ColonyCount ~ Strain * Treatment, family = \"quasipoisson\", \n",
    "     data = coloniesCN)\n",
    " \n",
    " Deviance Residuals: \n",
    "    Min      1Q  Median      3Q     Max  \n",
    " -32.37  -10.33   -3.32    0.84   97.84  \n",
    " \n",
    " Coefficients:\n",
    "                         Estimate Std. Error t value Pr(>|t|)    \n",
    " (Intercept)                6.288      0.158   39.78  < 2e-16 ***\n",
    " Strain712                 -1.355      0.349   -3.88  0.00013 ***\n",
    " Strain881                 -3.744      1.115   -3.36  0.00089 ***\n",
    " Strain899                 -7.269      5.800   -1.25  0.21113    \n",
    " StrainTA102               -1.447      0.353   -4.10  5.4e-05 ***\n",
    " TreatmentNG               -2.173      0.510   -4.26  2.8e-05 ***\n",
    " Strain712:TreatmentNG     -1.053      1.696   -0.62  0.53529    \n",
    " Strain881:TreatmentNG      5.308      1.236    4.30  2.4e-05 ***\n",
    " Strain899:TreatmentNG      9.539      5.822    1.64  0.10246    \n",
    " StrainTA102:TreatmentNG    3.592      0.621    5.79  1.9e-08 ***\n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 \n",
    " \n",
    " (Dispersion parameter for quasipoisson family taken to be 403.4)\n",
    " \n",
    "     Null deviance: 134445  on 293  degrees of freedom\n",
    " Residual deviance:  61579  on 284  degrees of freedom\n",
    "   (46 observations deleted due to missingness)\n",
    " AIC: NA\n",
    " \n",
    " Number of Fisher Scoring iterations: 7 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is pretty similar to the previous table but there two differences.\n",
    "First, the dispersion parameter line has changed. Second, all the $p$\n",
    "values have got less significant – this is the effect of controlling for\n",
    "the overdispersion.\n",
    "\n",
    "We’ll look at the model diagnostic plots next:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> par(mfrow = c(2, 2), mar = c(3, 3, 3, 1), mgp = c(2, 0.8, 0))\n",
    "> plot(modQPois)\t"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image](qpoisDiag.pdf){width=\"80.00000%\"}\n",
    "\n",
    "The residuals and leverage plot is now ok. The QQ plot is not better,\n",
    "but is still an improvement over the original linear model. We can’t\n",
    "improve the model fit any more — it isn’t perfect but we’ll accept those\n",
    "imperfections. It is worth thinking about the imperfections though —\n",
    "what might give rise to occasional larger than expected counts of\n",
    "colonies?\n",
    "\n",
    "We’ll look at the <span>anova</span> table next. Technically, this is\n",
    "now analysis of deviance not analysis of variance but the concept is the\n",
    "same. Different tests are appropriate for different families of\n",
    "distribution, but we can use $F$ here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> anova(modQPois, test = \"F\")\n",
    "\n",
    " Analysis of Deviance Table\n",
    " \n",
    " Model: quasipoisson, link: log\n",
    " \n",
    " Response: ColonyCount\n",
    " \n",
    " Terms added sequentially (first to last)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Df Deviance Resid. Df Resid. Dev     F  Pr(>F)    \n",
    "     NULL                               293     134445                  \n",
    "     Strain            4    13923       289     120521  8.63 1.4e-06 ***\n",
    "     Treatment         1     6055       288     114467 15.01 0.00013 ***\n",
    "     Strain:Treatment  4    52888       284      61579 32.78 < 2e-16 ***\n",
    "     ---\n",
    "     Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Can we simplify the model? The interaction is the only term we can drop\n",
    "and looks highly significant, but we can check by deleting it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> drop.scope(modQPois)\n",
    "\n",
    " [1] \"Strain:Treatment\"\n",
    "\n",
    "> modQPois2 <- update(modQPois, . ~ . - Strain:Treatment)\n",
    "> anova(modQPois, modQPois2, test = \"F\")\n",
    "\n",
    " Analysis of Deviance Table\n",
    " \n",
    " Model 1: ColonyCount ~ Strain * Treatment\n",
    " Model 2: ColonyCount ~ Strain + Treatment\n",
    "   Resid. Df Resid. Dev Df Deviance    F Pr(>F)    \n",
    " 1       284      61579                            \n",
    " 2       288     114467 -4   -52888 32.8 <2e-16 ***\n",
    " ---\n",
    " Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No, that makes the model much worse, so we now have our final model.\n",
    "\n",
    "\\[$\\quad\\star$\\]\n",
    "\n",
    "Fit this new model in your script and check you’ve got the same results.\n",
    "\n",
    "Model predictions\n",
    "-----------------\n",
    "\n",
    "We can get model predictions and standard errors using the <span>\n",
    "predict</span> function. There is a difference though. GLMs use an\n",
    "internal transformation to model the data using a <span>*link\n",
    "function*</span> and the coefficients in the summary above are on the\n",
    "scale of the link transformation. For quasipoisson, this is a <span>*log\n",
    "link*</span>, which you can see in the output of <span>anova</span>. You\n",
    "can use <span>predict</span> to get predictions on the scale of the\n",
    "original <span>*response*</span>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use expand.grid to get all combinations of factors\n",
    "> df <- expand.grid(Strain = levels(coloniesCN$Strain), Treatment =\n",
    "levels(coloniesCN$Treatment))\n",
    "> predict(modQPois, newdata = df, type = \"response\")\n",
    "\n",
    "       1       2       3       4       5       6       7       8       9      10 \n",
    " 538.200 138.867  12.731   0.375 126.687  61.286   5.517 292.714 593.000 523.900 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Those are the same values as the means we calculated for the barplot.\n",
    "Adding standard errors to barplots is more difficult for GLMs and we\n",
    "won’t go into it here.\n",
    "\n",
    "Reporting the model\n",
    "-------------------\n",
    "\n",
    "Reporting complicated statistics is a difficult business. There is a lot\n",
    "of detail involved and you want the reader to understand what you have\n",
    "done well enough to repeat the analysis if needed. You also have to\n",
    "summarise and explain the results without pages of R output.\n",
    "\n",
    "Here are some pointers:\n",
    "\n",
    "-   What does the data show? Present a graph or a table to show the data\n",
    "    you are about to model. <span>*Always*</span> include a figure or\n",
    "    table legend and <span>*always*</span> refer to that figure or\n",
    "    legend from the text.\n",
    "\n",
    "-   Have you transformed the data or used a subset? If so, why?\n",
    "\n",
    "-   What kind of model or statistical test have you used?\n",
    "\n",
    "-   With linear models, what is the response variable and what are the\n",
    "    explanatory variables.\n",
    "\n",
    "-   Have you simplified the model and, if so, what was the most complex\n",
    "    model you tried?\n",
    "\n",
    "-   How did you check the suitability of the model? Are there any\n",
    "    problems with the model and, if so, what might cause them?\n",
    "\n",
    "-   If you summarise stats in text, you must include all the information\n",
    "    about the test."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "-   For $F$ tests, this is $F$, the two degrees of freedom and the\n",
    "    p value. For example: ‘There is a significant interaction\n",
    "    between treatment and strain ($F_{4,284}=32.7, p < 0.0001$)’.\n",
    "\n",
    "-   For $t$ tests, this is the coefficient, the standard error, $t$,\n",
    "    the degrees of freedom and $p$. For example, ‘Across strains,\n",
    "    the main effect of nitroguanisine is to reduce colony counts\n",
    "    relative to the control (estimate=-2.17, s.e= 0.51, $t=-4.26$,\n",
    "    df=284, $p < 0.0001$)’."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-   With more complex models, it is common to present either the anova\n",
    "    table or the coefficients table as a summary of the model output.\n",
    "    Just include the tables from R output, not the information\n",
    "    around it. See Table 1 for an example.\n",
    "\n",
    "-   <span>*Never*</span> just include chunks of raw output from R.\n",
    "\n",
    "-   Most importantly, what is the interpretation of the model. What is\n",
    "    it telling you about the data?\n",
    "\n",
    "<span>*Table 1*</span>: Coefficients from a GLM of treatment and strain\n",
    "as predictors of colony count."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Estimate   Std. Error   t value p\n",
    "  ------------------------- ---------- ------------ --------- ------------\n",
    "                (Intercept)       6.29         0.16     39.78 &lt;0.0001\n",
    "                  Strain712      -1.35         0.35     -3.88 0.0001\n",
    "                  Strain881      -3.74         1.11     -3.36 0.0009\n",
    "                  Strain899      -7.27         5.80     -1.25 0.2111\n",
    "                StrainTA102      -1.45         0.35     -4.10 &lt;0.0001\n",
    "                TreatmentNG      -2.17         0.51     -4.26 &lt;0.0001\n",
    "      Strain712:TreatmentNG      -1.05         1.70     -0.62 0.5353\n",
    "      Strain881:TreatmentNG       5.31         1.24      4.30 &lt;0.0001\n",
    "      Strain899:TreatmentNG       9.54         5.82      1.64 0.1025\n",
    "    StrainTA102:TreatmentNG       3.59         0.62      5.79 &lt;0.0001"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Halos and lawns\n",
    "---------------\n",
    "\n",
    "We’ll keep this one simple since it is harder to analyse. The response\n",
    "variable (<span>HaloLawn</span>) is binary — the plates either have a\n",
    "lawn or not. We’ll just look at a contingency table of how many plates\n",
    "have halos or lawns under each combination of treatment and strain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "> table(Halo = colonies$HaloLawn, Strain = colonies$Strain, Treatment = \n",
    "colonies$Treatment)\n",
    "\n",
    " , , Treatment = Control\n",
    " \n",
    "     Strain\n",
    " Halo 421 712 881 899 TA102\n",
    "    N   0   0   0   0     0\n",
    "    Y   0   0   0   0     0\n",
    " \n",
    " , , Treatment = His\n",
    " \n",
    "     Strain\n",
    " Halo 421 712 881 899 TA102\n",
    "    N   1   1   0   0     1\n",
    "    Y  29  29  26  32    31\n",
    " \n",
    " , , Treatment = NG\n",
    " \n",
    "     Strain\n",
    " Halo 421 712 881 899 TA102\n",
    "    N   0   0   0   0     0\n",
    "    Y   0   0   0   0     0\n",
    " \n",
    " , , Treatment = Strep\n",
    " \n",
    "     Strain\n",
    " Halo 421 712 881 899 TA102\n",
    "    N   5  30   5  21     0\n",
    "    Y  25   0  25   9    32\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, lawns and halos are never recorded from nitroguanisine or the\n",
    "control. They’re nearly always found with histidine and different\n",
    "strains have different response to streptomysin. Again, treatment and\n",
    "strain interact. Although you can use a $\\chi^2$ test with two\n",
    "dimensional contingency tables to look for independence between factors,\n",
    "you can’t with a three-way table.\n",
    "\n",
    "[^1]: Here you work with the script file <span>t\\_F\\_tests.R</span>\n",
    "\n",
    "[^2]: The number of pairs is the binomial coefficient $C(n,2)$ —\n",
    "    inevitably, R has a function for this: <span> choose(2:6,2)</span>\n",
    "\n",
    "[^3]: This slightly dodgy estimate comes from an estimated average\n",
    "    volume for arthropods of of 45.21 mm$^3$ and assuming a density of 1\n",
    "    gm per cm$^3$. The volume is from: Orme, C. D. L., Quicke, D. L. J.,\n",
    "    Cook, J. M. and Purvis, A. (2002), Body size does not predict\n",
    "    species richness among the metazoan phyla. Journal of Evolutionary\n",
    "    Biology, 15: 235–247.\n",
    "\n",
    "[^4]: Note that $1/0.482 \\approx 2.074$ and $1/2.074 \\approx 0.482$: in\n",
    "    this case, it doesn’t matter which way round you compare the two\n",
    "    variances!\n",
    "\n",
    "[^5]: Here you work with the script file <span>regress.R</span>\n",
    "\n",
    "[^6]: Here you work with the script file <span>anova.R</span>\n",
    "\n",
    "[^7]: That is, it is 1 minus the ratio of the square of the standard\n",
    "    error of the estimate to the sample variance of the response\n",
    "\n",
    "[^8]: i.e., Standard error of the estimate won’t decrease\n",
    "\n",
    "[^9]: Here you work with the script file <span>MulExpl.R</span>\n",
    "\n",
    "[^10]: Here you work with the script file <span>MulExplInter.R</span>\n",
    "\n",
    "[^11]: Here you work with the script file <span>ModelSimp.R</span>\n",
    "\n",
    "[^12]: Here you work with the script file <span>glm.R</span>"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 2
}
